{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "0a913d01-55d7-47d5-a934-9096a1056f5c",
   "metadata": {},
   "outputs": [],
   "source": [
    "rm -rf "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "536bf5aa-8434-4b45-9f98-f8a52cfdc50c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/root/miniconda3/bin/python\n"
     ]
    }
   ],
   "source": [
    "import sys\n",
    "print(sys.executable)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "f8c20021-f7b1-42b4-9aff-10a37a33fc8b",
   "metadata": {},
   "outputs": [],
   "source": [
    "import scimilarity\n",
    "from scimilarity.utils import lognorm_counts, align_dataset\n",
    "from scimilarity import CellQuery"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "88e903e0-e133-4659-9f3e-adca60850834",
   "metadata": {},
   "outputs": [],
   "source": [
    "from itertools import combinations\n",
    "import numpy as np\n",
    "import random\n",
    "import torch\n",
    "from typing import List, Union, Optional\n",
    "\n",
    "from scimilarity.ontologies import (\n",
    "    import_cell_ontology,\n",
    "    get_id_mapper,\n",
    "    get_all_ancestors,\n",
    "    get_all_descendants,\n",
    "    get_parents,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "289c7a76-a7fa-4f18-a6de-98bca9d84294",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "bf14578f-97c8-4a77-b1ca-49a78d56421d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# !pip install numpy==1.26.0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "94105c04-eb30-49d2-b777-9cb0886d2bee",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'1.26.0'"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.__version__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "02156cc8-a7dc-4d20-8fbd-be20159c8b4b",
   "metadata": {},
   "outputs": [],
   "source": [
    "import torch\n",
    "from datetime import datetime\n",
    "import json\n",
    "import os\n",
    "import pandas as pd\n",
    "import pytorch_lightning as pl\n",
    "import torch\n",
    "from torch import nn\n",
    "import torch.nn.functional as F\n",
    "from typing import Optional, List\n",
    "\n",
    "from scimilarity.triplet_selector import TripletSelector\n",
    "from scimilarity.nn_models import Encoder, Decoder"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "f737f991-3def-44bc-82f9-60ced55612cd",
   "metadata": {},
   "outputs": [],
   "source": [
    "import scanpy as sc\n",
    "import anndata as ad"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "8d28643f-4e4e-4376-b762-a6204cf80f08",
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn.model_selection import train_test_split"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "8bb8243f-9c1b-474f-93f7-22d3897859f3",
   "metadata": {},
   "outputs": [],
   "source": [
    "from torch.utils.data import TensorDataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "9ee63e6a-4030-465c-832a-bae9ef0f3343",
   "metadata": {},
   "outputs": [],
   "source": [
    "# rm -r /root/autodl-tmp/train_path_A_S/Study_A/Sample_A_zarr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "b9f40e32-39b3-4cb5-bbad-316a1240737a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# rm -r /root/autodl-tmp/train_path_A_S/Study_S/Sample_S_zarr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "43d9f220-6ba7-42f5-8ba0-caa35b6d393d",
   "metadata": {},
   "outputs": [],
   "source": [
    "import zarr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "75197c93-72bf-4feb-a9d0-ff1baddd700b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# !unzip /root/autodl-tmp/A_new_zarr.zip -d /root/autodl-tmp/train_path_A_S/Study_A"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "d2659f29-73fc-4f41-9178-06988882bae4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# !unzip /root/autodl-tmp/S_new_zarr.zip -d /root/autodl-tmp/train_path_A_S/Study_S"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "f18f599a-b86e-4573-9ae5-2ad492ea4254",
   "metadata": {},
   "outputs": [],
   "source": [
    "from collections import Counter\n",
    "import numpy as np\n",
    "import os\n",
    "import pandas as pd\n",
    "import pytorch_lightning as pl\n",
    "import torch\n",
    "from torch.utils.data import Dataset, DataLoader, WeightedRandomSampler\n",
    "from tqdm import tqdm\n",
    "from typing import Optional\n",
    "\n",
    "from scimilarity.zarr_dataset import ZarrDataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "7769cc60-9a7c-4820-b64e-36acaf5be718",
   "metadata": {},
   "outputs": [],
   "source": [
    "class scDataset(Dataset):\n",
    "    \"\"\"A class that represent a collection of single cell datasets in zarr format.\"\"\"\n",
    "\n",
    "    def __init__(self, data_list, obs_celltype=\"cell_type\", obs_study=\"orig.ident\"):\n",
    "        \"\"\"Constructor.\n",
    "\n",
    "        Parameters\n",
    "        ----------\n",
    "        data_list: list\n",
    "            List of single-cell datasets.\n",
    "        obs_celltype: str, default: \"celltype_name\"\n",
    "            Cell type name.\n",
    "        obs_study: str, default: \"study\"\n",
    "            Study name.\n",
    "        \"\"\"\n",
    "\n",
    "        self.data_list = data_list\n",
    "        self.ncells_list = [data.shape[0] for data in data_list]\n",
    "        self.ncells = sum(self.ncells_list)\n",
    "        self.obs_celltype = obs_celltype\n",
    "        self.obs_study = obs_study\n",
    "\n",
    "        self.data_idx = [\n",
    "            n for n in range(len(self.ncells_list)) for i in range(self.ncells_list[n])\n",
    "        ]\n",
    "        self.cell_idx = [\n",
    "            i for n in range(len(self.ncells_list)) for i in range(self.ncells_list[n])\n",
    "        ]\n",
    "\n",
    "    def __len__(self):\n",
    "        return self.ncells\n",
    "\n",
    "    def __getitem__(self, idx):\n",
    "        # data, label, study\n",
    "        data_idx = self.data_idx[idx]\n",
    "        cell_idx = self.cell_idx[idx]\n",
    "        return (\n",
    "            self.data_list[data_idx].get_cell(cell_idx).toarray(),\n",
    "            self.data_list[data_idx].get_obs(self.obs_celltype)[cell_idx],\n",
    "            self.data_list[data_idx].get_obs(self.obs_study)[cell_idx],\n",
    "        )\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "b5f37859-39f3-40e3-930b-7ddd3944ab86",
   "metadata": {},
   "outputs": [],
   "source": [
    "import zarr\n",
    "# 假设我们有两个单细胞数据集，分别加载\n",
    "A = ad.read_zarr(\"/root/autodl-tmp/train_path_A_S/Study_A/Sample_A_zarr\")\n",
    "S = ad.read_zarr(\"/root/autodl-tmp/train_path_A_S/Study_S/Sample_S_zarr\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "00abcb52-ab84-4350-bbe8-8cb33e3c6149",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 61000 × 34667\n",
       "    obs: 'Capbatch', 'DMGM_DL', 'DMGM_UL', 'DMWM', 'NBB_case', 'PMI', 'RIN', 'RNA_snn_res.0.3', 'SCT_snn_res.0.1', 'Seqbatch', 'UMIs', 'age', 'cell_type', 'diagnosis', 'genes', 'mito_perc', 'nCount_RNA', 'nCount_SCT', 'nFeature_RNA', 'nFeature_SCT', 'orig.ident', 'pathology', 'percent.mt', 'pseudotime', 'region', 'ribo_perc', 'sample', 'seurat_cluster', 'seurat_clusters', 'sex', 'stage'\n",
       "    layers: 'counts'"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "A"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "a0ec2fc9-3cc3-4b1e-91d6-a14fd8f799f4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "cell_type\n",
       "oligodendrocyte                   45697\n",
       "astrocyte                          8211\n",
       "neuron                             2803\n",
       "oligodendrocyte precursor cell     2175\n",
       "endothelial cell                   1708\n",
       "lymphocyte                          406\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "A.obs['cell_type'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "030abc07-c7f1-4132-b4a4-af3c2c61d3ab",
   "metadata": {},
   "outputs": [],
   "source": [
    "# A.obs['cell_type'] = A.obs['cell_type'].replace({'Oligo1':'oligodendrocyte','opc':'oligodendrocyte precursor cell','neurons':'neuron','astrocytes':'astrocyte','vascular_cells': 'endothelial cell','lymphocytes': 'lymphocyte'})\n",
    "# A.obs['cell_type'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "b0360823-a77a-4fa9-85f0-c07b179d6c69",
   "metadata": {},
   "outputs": [],
   "source": [
    "# A = A[~A.obs['cell_type'].isin(['immune']), :]\n",
    "# A.obs['cell_type'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "d232a380-55cf-4cce-8558-77c453369967",
   "metadata": {},
   "outputs": [],
   "source": [
    "# A.shape[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "144260de-7f5d-47bd-b1cd-a2447f3784f8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 45215 × 34667\n",
       "    obs: 'Capbatch', 'DMGM_DL', 'DMGM_UL', 'DMWM', 'NBB_case', 'PMI', 'RIN', 'RNA_snn_res.0.3', 'SCT_snn_res.0.1', 'Seqbatch', 'UMIs', 'age', 'cell_type', 'diagnosis', 'genes', 'mito_perc', 'nCount_RNA', 'nCount_SCT', 'nFeature_RNA', 'nFeature_SCT', 'orig.ident', 'pathology', 'percent.mt', 'pseudotime', 'region', 'ribo_perc', 'sample', 'seurat_cluster', 'seurat_clusters', 'sex', 'stage'\n",
       "    layers: 'counts'"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "S"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "c66e3372-34c3-403a-a6d2-1e7a634a890a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "cell_type\n",
       "neuron                            21539\n",
       "oligodendrocyte                   12394\n",
       "astrocyte                          5381\n",
       "oligodendrocyte precursor cell     3844\n",
       "microglial cell                    1425\n",
       "endothelial cell                    374\n",
       "phagocyte                           258\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "S.obs['cell_type'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "a7310d55-2f4a-4299-932e-69dc96e102d4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# S.obs['cell_type'] = S.obs['cell_type'].replace({'OL-A':'oligodendrocyte','OL-B':'oligodendrocyte','OL-C':'oligodendrocyte','OPC':'oligodendrocyte precursor cell','IN-VIP':'neuron','EN-L4':'neuron','EN-L2-3-A':'neuron','EN-L2-3-B':'neuron','EN-L5-6':'neuron','EN-MIX':'neuron','EN-PYR':'neuron','IN-PVALB':'neuron','IN-SST':'neuron','IN-SV2C':'neuron','Astrocytes':'astrocyte','Endo cells': 'endothelial cell','Phagocytes': 'phagocyte','Microglia': 'micrglial cell'})\n",
    "# S.obs['cell_type'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "50c06220-7f48-4979-8e29-90f5fb97d636",
   "metadata": {},
   "outputs": [],
   "source": [
    "# S = S[~S.obs['cell_type'].isin(['Stromal cells']), :]\n",
    "# S = S[~S.obs['cell_type'].isin(['T cells']), :]\n",
    "# S = S[~S.obs['cell_type'].isin(['B cells']), :]\n",
    "# S = S[~S.obs['cell_type'].isin(['Glia-MIX']), :]\n",
    "# S.obs['cell_type'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "a44036dc-2c92-4b61-8f29-2d58bb8ecd4a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 将这两个数据集传递给 scDataset\n",
    "data_list = [A, S]\n",
    "sc_dataset = scDataset(data_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "78947452-55db-4ef2-be6e-53c306806b78",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[61000, 45215]"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sc_dataset.ncells_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "cc3765d8-37f1-481b-a761-96da3a70446e",
   "metadata": {},
   "outputs": [],
   "source": [
    "class MetricLearningDataModule(pl.LightningDataModule):\n",
    "    \"\"\"A class to encapsulate a collection of zarr datasets to train the model.\"\"\"\n",
    "\n",
    "    def __init__(\n",
    "        self,\n",
    "        train_path: str,\n",
    "        gene_order: str,\n",
    "        val_path: Optional[str] = None,\n",
    "        obs_field: str = \"celltype_name\",\n",
    "        batch_size: int = 1000,\n",
    "        num_workers: int = 1,\n",
    "    ):\n",
    "        \"\"\"Constructor.\n",
    "\n",
    "        Parameters\n",
    "        ----------\n",
    "        train_path: str\n",
    "            Path to folder containing all training datasets.\n",
    "            All datasets should be in zarr format, aligned to a known gene space, and\n",
    "            cleaned to only contain valid cell ontology terms.\n",
    "        gene_order: str\n",
    "            Use a given gene order as described in the specified file. One gene\n",
    "            symbol per line.\n",
    "            IMPORTANT: the zarr datasets should already be in this gene order\n",
    "            after preprocessing.\n",
    "        val_path: str, optional, default: None\n",
    "            Path to folder containing all validation datasets.\n",
    "        obs_field: str, default: \"celltype_name\"\n",
    "            The obs key name containing celltype labels.\n",
    "        batch_size: int, default: 1000\n",
    "            Batch size.\n",
    "        num_workers: int, default: 1\n",
    "            The number of worker threads for dataloaders\n",
    "\n",
    "        Examples\n",
    "        --------\n",
    "        >>> datamodule = MetricLearningZarrDataModule(\n",
    "                batch_size=1000,\n",
    "                num_workers=1,\n",
    "                obs_field=\"celltype_name\",\n",
    "                train_path=\"train\",\n",
    "                gene_order=\"gene_order.tsv\",\n",
    "            )\n",
    "        \"\"\"\n",
    "\n",
    "        super().__init__()\n",
    "        self.train_path = train_path\n",
    "        self.val_path = val_path\n",
    "        self.batch_size = batch_size\n",
    "        self.num_workers = num_workers\n",
    "\n",
    "        # gene space needs be aligned to the given gene order\n",
    "        with open(gene_order, \"r\") as fh:\n",
    "            self.gene_order = [line.strip() for line in fh]\n",
    "\n",
    "        self.n_genes = len(self.gene_order)  # used when creating training model\n",
    "\n",
    "        train_data_list = []\n",
    "        self.train_Y = []  # text labels\n",
    "        self.train_study = []  # text studies\n",
    "\n",
    "        if self.train_path[-1] != os.sep:\n",
    "            self.train_path += os.sep\n",
    "\n",
    "        self.train_file_list = [\n",
    "    (\n",
    "        root.replace(self.train_path, \"\").split(os.sep)[0],  # 获取根目录名\n",
    "        dir_name.replace(\"_zarr\", \"\")  # 去掉'_zarr'后缀\n",
    "    )\n",
    "    for root, dirs, files in os.walk(self.train_path)\n",
    "    for dir_name in dirs  # 遍历所有子目录\n",
    "    if dir_name.endswith(\"_zarr\")  # 检查是否以'_zarr'结尾\n",
    "]\n",
    "\n",
    "        for study, sample in tqdm(self.train_file_list):\n",
    "            data_path = os.path.join(\n",
    "                self.train_path, study, sample + \"_zarr\"\n",
    "            )\n",
    "            if os.path.isdir(data_path):\n",
    "                zarr_data = ZarrDataset(data_path)\n",
    "                train_data_list.append(zarr_data)\n",
    "                self.train_Y.extend(zarr_data.get_obs(obs_field).astype(str).tolist())\n",
    "                self.train_study.extend(zarr_data.get_obs(\"orig.ident\").astype(str).tolist())\n",
    "\n",
    "        # Lazy load training data from list of zarr datasets\n",
    "        self.train_dataset = scDataset(train_data_list)\n",
    "\n",
    "        self.class_names = set(self.train_Y)\n",
    "        self.label2int = {label: i for i, label in enumerate(self.class_names)}\n",
    "        self.int2label = {value: key for key, value in self.label2int.items()}\n",
    "\n",
    "        self.val_dataset = None\n",
    "        if self.val_path is not None:\n",
    "            val_data_list = []\n",
    "            self.val_Y = []\n",
    "            self.val_study = []\n",
    "\n",
    "            if self.val_path[-1] != os.sep:\n",
    "                self.val_path += os.sep\n",
    "\n",
    "            self.val_file_list = [\n",
    "                (\n",
    "                    root.replace(self.val_path, \"\").split(os.sep)[0],\n",
    "                    dirs[0].replace(\"zarr\", \"\"),\n",
    "                )\n",
    "                for root, dirs, files in os.walk(self.val_path)\n",
    "                if dirs and dirs[0].endswith(\"zarr\")\n",
    "            ]\n",
    "\n",
    "            for study, sample in tqdm(self.val_file_list):\n",
    "                data_path = os.path.join(\n",
    "                    self.val_path, study, sample, sample + \"zarr\"\n",
    "                )\n",
    "                if os.path.isdir(data_path):\n",
    "                    zarr_data = ZarrDataset(data_path)\n",
    "                    val_data_list.append(zarr_data)\n",
    "                    self.val_Y.extend(zarr_data.get_obs(obs_field).astype(str).tolist())\n",
    "                    self.val_study.extend(\n",
    "                        zarr_data.get_obs(\"study\").astype(str).tolist()\n",
    "                    )\n",
    "\n",
    "            # Lazy load val data from list of zarr datasets\n",
    "            self.val_dataset = scDataset(val_data_list)\n",
    "\n",
    "    def get_sampler_weights(\n",
    "        self, labels: list, studies: Optional[list] = None\n",
    "    ) -> WeightedRandomSampler:\n",
    "        \"\"\"Get weighted random sampler.\n",
    "\n",
    "        Parameters\n",
    "        ----------\n",
    "        dataset: scDataset\n",
    "            Single cell dataset.\n",
    "\n",
    "        Returns\n",
    "        -------\n",
    "        WeightedRandomSampler\n",
    "            A WeightedRandomSampler object.\n",
    "        \"\"\"\n",
    "\n",
    "        if studies is None:\n",
    "            class_sample_count = Counter(labels)\n",
    "            sample_weights = torch.Tensor([1.0 / class_sample_count[t] for t in labels])\n",
    "        else:\n",
    "            class_sample_count = Counter(labels)\n",
    "            study_sample_count = Counter(studies)\n",
    "            class_sample_count = {\n",
    "                x: np.log1p(class_sample_count[x] / 1e4) for x in class_sample_count\n",
    "            }\n",
    "            study_sample_count = {\n",
    "                x: np.log1p(study_sample_count[x] / 1e5) for x in study_sample_count\n",
    "            }\n",
    "            sample_weights = torch.Tensor(\n",
    "                [\n",
    "                    1.0 / class_sample_count[labels[i]] / study_sample_count[studies[i]]\n",
    "                    for i in range(len(labels))\n",
    "                ]\n",
    "            )\n",
    "        return WeightedRandomSampler(sample_weights, len(sample_weights))\n",
    "\n",
    "\n",
    "    def collate(self, batch):\n",
    "        \"\"\"Collate tensors.\n",
    "\n",
    "        Parameters\n",
    "        ----------\n",
    "        batch:\n",
    "            Batch to collate.\n",
    "\n",
    "        Returns\n",
    "        -------\n",
    "        tuple\n",
    "            A Tuple[torch.Tensor, torch.Tensor, list] containing information\n",
    "            on the collated tensors.\n",
    "        \"\"\"\n",
    "\n",
    "        profiles, labels, studies = tuple(\n",
    "            map(list, zip(*batch))\n",
    "        )  # tuple([list(t) for t in zip(*batch)])\n",
    "        return (\n",
    "            torch.squeeze(torch.Tensor(np.vstack(profiles))),\n",
    "            torch.Tensor([self.label2int[l] for l in labels]),  # text to int labels\n",
    "            studies,\n",
    "        )\n",
    "\n",
    "\n",
    "    def train_dataloader(self) -> DataLoader:\n",
    "        \"\"\"Load the training dataset.\n",
    "\n",
    "        Returns\n",
    "        -------\n",
    "        DataLoader\n",
    "            A DataLoader object containing the training dataset.\n",
    "        \"\"\"\n",
    "\n",
    "        return DataLoader(\n",
    "            self.train_dataset,\n",
    "            batch_size=self.batch_size,\n",
    "            num_workers=self.num_workers,\n",
    "            pin_memory=True,\n",
    "            drop_last=True,\n",
    "            sampler=self.get_sampler_weights(self.train_Y, self.train_study),\n",
    "            collate_fn=self.collate,\n",
    "        )\n",
    "\n",
    "\n",
    "    def val_dataloader(self) -> DataLoader:\n",
    "        \"\"\"Load the validation dataset.\n",
    "\n",
    "        Returns\n",
    "        -------\n",
    "        DataLoader\n",
    "            A DataLoader object containing the validation dataset.\n",
    "        \"\"\"\n",
    "\n",
    "        if self.val_dataset is None:\n",
    "            return None\n",
    "        return DataLoader(\n",
    "            self.val_dataset,\n",
    "            batch_size=self.batch_size,\n",
    "            num_workers=self.num_workers,\n",
    "            pin_memory=True,\n",
    "            drop_last=True,\n",
    "            sampler=self.get_sampler_weights(self.val_Y, self.val_study),\n",
    "            collate_fn=self.collate,\n",
    "        )\n",
    "\n",
    "\n",
    "    def test_dataloader(self) -> DataLoader:\n",
    "        \"\"\"Load the test dataset.\n",
    "\n",
    "        Returns\n",
    "        -------\n",
    "        DataLoader\n",
    "            A DataLoader object containing the test dataset.\n",
    "        \"\"\"\n",
    "\n",
    "        return self.val_dataloader()\n",
    "    \n",
    "    def __getitem__(self, idx):\n",
    "        # 获取稀疏矩阵数据\n",
    "        csr_matrix_data = self.data_list[idx].get_cell(cell_idx)\n",
    "\n",
    "        # 将稀疏矩阵转为密集矩阵\n",
    "        dense_array = csr_matrix_data.toarray()  # 替换 .A 为 .toarray()\n",
    "\n",
    "        # 返回转为密集矩阵的结果\n",
    "        return dense_array"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "f2b28939-1bc2-44a7-8c68-592244f15613",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 2/2 [00:00<00:00, 44.09it/s]\n"
     ]
    }
   ],
   "source": [
    "# 准备数据和文件\n",
    "train_path = \"/root/autodl-tmp/train_path_A_S/\"\n",
    "# Val_path = \"/root/autodl-tmp/Val_path/S_new_zarr\"\n",
    "gene_order_file = \"/root/autodl-tmp/genes.tsv\"\n",
    "\n",
    "# 创建数据模块实例\n",
    "datamodule = MetricLearningDataModule(\n",
    "    train_path=train_path,          # 第一个数据集作为训练集\n",
    "    val_path=None,            # 第二个数据集作为验证集\n",
    "    gene_order=gene_order_file,\n",
    "    obs_field=\"cell_type\",         # 细胞类型标签字段\n",
    "    batch_size=1000,\n",
    "    num_workers=1\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "0b92bb08-7df4-41a1-ba7d-73458937ef3a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{0: 'oligodendrocyte',\n",
       " 1: 'lymphocyte',\n",
       " 2: 'oligodendrocyte precursor cell',\n",
       " 3: 'phagocyte',\n",
       " 4: 'endothelial cell',\n",
       " 5: 'microglial cell',\n",
       " 6: 'neuron',\n",
       " 7: 'astrocyte'}"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "int2label = datamodule.int2label\n",
    "int2label"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "id": "21909cf1-b626-4033-8c5a-2a6f3f713f4e",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.DataFrame(list(int2label.items()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "id": "ec381e37-c403-4dfa-aa8c-a041245c4df8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 保存为 CSV 文件\n",
    "df.to_csv(\"label_ints.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "1e554224-3622-425c-a6b2-a761e4e75b3a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'oligodendrocyte': 0,\n",
       " 'lymphocyte': 1,\n",
       " 'oligodendrocyte precursor cell': 2,\n",
       " 'phagocyte': 3,\n",
       " 'endothelial cell': 4,\n",
       " 'microglial cell': 5,\n",
       " 'neuron': 6,\n",
       " 'astrocyte': 7}"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "label2int = datamodule.label2int\n",
    "label2int"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "57e07ea0-12c0-4c7e-91b2-3398934c919a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'astrocyte',\n",
       " 'endothelial cell',\n",
       " 'lymphocyte',\n",
       " 'microglial cell',\n",
       " 'neuron',\n",
       " 'oligodendrocyte',\n",
       " 'oligodendrocyte precursor cell',\n",
       " 'phagocyte'}"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "datamodule.class_names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "c23f3ed4-f9f6-4837-a788-5dacba616834",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dict_keys([0, 1, 2, 3, 4, 5, 6, 7])"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "int2label.keys()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "8bcdb8b5-8ddd-4edd-9bb0-6533e99173bd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([0, 1, 2, 3, 4, 5, 6, 7])"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "labels = np.array([0, 1, 2, 3, 4, 5, 6, 7])\n",
    "labels"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "25147286-8cf7-4524-967d-382c8a7e2938",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "106215"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(datamodule.train_Y)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "986c6004-0d04-4af3-b265-c52ac197eda7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('Study_A', 'Sample_A'), ('Study_S', 'Sample_S')]"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "datamodule.train_file_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "149ca30d-b330-4796-8ce9-ebd4576e35ee",
   "metadata": {},
   "outputs": [],
   "source": [
    "train_dataloader = datamodule.train_dataloader()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "1c06d0cc-990f-4308-a6b3-c4cd90c2fd5f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<torch.utils.data.dataloader.DataLoader at 0x7efeb5b3ddb0>"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "train_dataloader"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "e5f8f932-c2c8-4af9-95e1-b20d35a69d2c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<__main__.scDataset object at 0x7eff726f6d10>\n"
     ]
    }
   ],
   "source": [
    "print(train_dataloader.dataset)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "9ee39c72-dc05-4e9e-a29f-cd4cb7d02963",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[tensor([[0., 0., 0.,  ..., 0., 0., 0.],\n",
      "        [0., 0., 0.,  ..., 0., 0., 0.],\n",
      "        [0., 0., 0.,  ..., 0., 0., 0.],\n",
      "        ...,\n",
      "        [0., 0., 0.,  ..., 0., 0., 0.],\n",
      "        [0., 0., 0.,  ..., 0., 0., 0.],\n",
      "        [0., 0., 0.,  ..., 0., 0., 0.]]), tensor([1., 3., 6., 2., 1., 6., 1., 5., 6., 1., 3., 1., 7., 2., 6., 1., 0., 6.,\n",
      "        4., 5., 7., 1., 6., 6., 0., 6., 1., 4., 3., 5., 0., 6., 6., 7., 7., 5.,\n",
      "        4., 3., 6., 5., 2., 0., 3., 0., 0., 3., 1., 1., 1., 0., 5., 1., 0., 6.,\n",
      "        0., 0., 0., 0., 1., 0., 2., 6., 0., 0., 0., 6., 1., 6., 2., 0., 0., 7.,\n",
      "        6., 6., 4., 7., 1., 0., 2., 0., 7., 7., 7., 0., 7., 6., 0., 7., 6., 0.,\n",
      "        0., 2., 1., 3., 5., 6., 2., 1., 1., 7., 6., 0., 2., 2., 0., 7., 4., 3.,\n",
      "        7., 7., 5., 0., 7., 6., 2., 4., 1., 0., 4., 0., 6., 3., 4., 0., 4., 5.,\n",
      "        5., 7., 6., 7., 3., 7., 6., 6., 6., 0., 0., 7., 6., 0., 6., 6., 1., 7.,\n",
      "        4., 5., 3., 0., 7., 3., 0., 0., 0., 0., 6., 6., 2., 6., 4., 3., 0., 2.,\n",
      "        1., 3., 5., 0., 0., 1., 3., 0., 6., 1., 1., 7., 6., 7., 0., 6., 6., 0.,\n",
      "        6., 1., 6., 1., 5., 5., 0., 3., 2., 2., 5., 1., 0., 2., 2., 0., 6., 0.,\n",
      "        1., 0., 1., 0., 2., 3., 1., 3., 0., 7., 0., 3., 3., 1., 0., 1., 1., 2.,\n",
      "        5., 4., 6., 0., 4., 2., 5., 6., 7., 6., 7., 0., 4., 6., 2., 0., 2., 7.,\n",
      "        2., 4., 0., 1., 2., 4., 0., 2., 7., 0., 0., 6., 6., 7., 5., 0., 5., 6.,\n",
      "        4., 2., 0., 6., 7., 0., 6., 0., 5., 0., 1., 5., 3., 6., 6., 1., 0., 7.,\n",
      "        2., 6., 7., 2., 2., 0., 0., 6., 6., 0., 5., 2., 1., 4., 3., 0., 0., 3.,\n",
      "        3., 2., 5., 5., 6., 2., 6., 0., 2., 5., 4., 5., 1., 0., 0., 7., 5., 2.,\n",
      "        5., 0., 7., 3., 0., 6., 0., 2., 6., 0., 7., 0., 6., 6., 5., 5., 6., 6.,\n",
      "        6., 1., 0., 7., 6., 7., 2., 0., 6., 0., 4., 6., 3., 0., 1., 5., 2., 3.,\n",
      "        7., 4., 2., 0., 1., 2., 3., 6., 4., 3., 0., 0., 3., 5., 5., 1., 1., 0.,\n",
      "        6., 7., 0., 1., 0., 1., 4., 6., 6., 0., 0., 4., 7., 7., 5., 5., 1., 0.,\n",
      "        5., 2., 0., 0., 1., 1., 5., 3., 4., 6., 2., 6., 4., 0., 6., 6., 0., 4.,\n",
      "        6., 7., 6., 6., 6., 3., 7., 7., 2., 0., 0., 0., 0., 3., 1., 6., 6., 4.,\n",
      "        4., 2., 0., 7., 7., 6., 5., 5., 7., 4., 6., 5., 5., 3., 2., 6., 6., 0.,\n",
      "        2., 0., 0., 6., 3., 1., 1., 4., 6., 6., 1., 7., 0., 0., 7., 0., 6., 3.,\n",
      "        1., 0., 1., 0., 6., 6., 5., 7., 2., 1., 0., 0., 7., 1., 6., 1., 7., 7.,\n",
      "        2., 4., 7., 7., 3., 2., 5., 0., 0., 6., 6., 1., 6., 7., 0., 1., 6., 4.,\n",
      "        0., 2., 5., 0., 0., 6., 0., 7., 0., 4., 6., 0., 0., 3., 5., 0., 1., 1.,\n",
      "        5., 0., 6., 2., 6., 1., 6., 0., 0., 0., 0., 1., 6., 4., 0., 6., 5., 2.,\n",
      "        5., 3., 3., 4., 6., 5., 4., 7., 6., 2., 6., 4., 6., 0., 2., 6., 7., 3.,\n",
      "        1., 5., 4., 6., 6., 0., 0., 1., 6., 5., 5., 6., 5., 2., 3., 7., 0., 0.,\n",
      "        7., 7., 2., 7., 0., 1., 6., 0., 5., 0., 5., 0., 3., 6., 0., 7., 7., 6.,\n",
      "        4., 6., 2., 7., 5., 4., 3., 5., 7., 2., 6., 2., 0., 6., 2., 6., 3., 0.,\n",
      "        0., 7., 3., 7., 7., 5., 7., 4., 7., 2., 5., 6., 7., 6., 0., 7., 1., 6.,\n",
      "        7., 7., 6., 5., 0., 0., 0., 0., 2., 3., 3., 5., 6., 4., 7., 0., 0., 1.,\n",
      "        4., 7., 2., 1., 4., 5., 0., 6., 3., 5., 3., 3., 3., 0., 1., 6., 7., 6.,\n",
      "        3., 0., 0., 0., 2., 6., 7., 2., 5., 0., 0., 1., 2., 0., 6., 6., 0., 2.,\n",
      "        6., 3., 5., 2., 0., 5., 5., 0., 6., 0., 7., 2., 1., 2., 4., 0., 0., 0.,\n",
      "        3., 3., 2., 2., 6., 1., 0., 0., 0., 5., 2., 5., 0., 5., 6., 4., 2., 7.,\n",
      "        3., 6., 7., 0., 1., 3., 2., 5., 5., 5., 7., 0., 6., 2., 0., 6., 6., 2.,\n",
      "        4., 6., 7., 5., 5., 6., 0., 6., 6., 7., 0., 7., 7., 2., 2., 5., 0., 2.,\n",
      "        4., 0., 1., 6., 3., 5., 2., 1., 3., 0., 4., 4., 5., 6., 0., 0., 7., 4.,\n",
      "        2., 5., 1., 7., 3., 6., 7., 5., 2., 6., 5., 6., 6., 7., 5., 5., 2., 5.,\n",
      "        5., 0., 6., 0., 0., 0., 2., 5., 0., 2., 7., 0., 1., 4., 0., 1., 2., 2.,\n",
      "        1., 6., 3., 0., 3., 0., 6., 7., 0., 2., 2., 3., 0., 0., 6., 6., 3., 4.,\n",
      "        0., 2., 0., 4., 7., 5., 6., 4., 6., 6., 2., 6., 1., 3., 5., 4., 3., 7.,\n",
      "        1., 2., 2., 3., 5., 6., 4., 2., 0., 6., 6., 5., 2., 1., 0., 5., 3., 2.,\n",
      "        7., 1., 6., 5., 5., 3., 5., 4., 1., 6., 0., 2., 0., 6., 3., 0., 6., 0.,\n",
      "        0., 1., 3., 5., 6., 5., 2., 6., 5., 5., 7., 6., 4., 1., 6., 1., 0., 1.,\n",
      "        7., 5., 6., 2., 2., 7., 4., 6., 5., 3., 7., 7., 3., 0., 2., 2., 2., 7.,\n",
      "        4., 2., 7., 6., 6., 6., 6., 6., 1., 6., 3., 0., 0., 7., 4., 3., 7., 4.,\n",
      "        4., 4., 2., 5., 3., 6., 6., 6., 6., 0., 2., 0., 6., 1., 5., 0., 2., 7.,\n",
      "        6., 7., 3., 3., 3., 7., 3., 7., 5., 5., 4., 4., 4., 0., 0., 0., 0., 7.,\n",
      "        7., 4., 0., 7., 7., 2., 2., 3., 6., 6., 0., 5., 0., 3., 6., 2., 5., 5.,\n",
      "        0., 5., 5., 7., 7., 2., 7., 7., 6., 7., 3., 6., 0., 0., 7., 4., 4., 0.,\n",
      "        5., 2., 4., 6., 6., 4., 0., 7., 6., 0.]), ['GSE180759', 'Schirmer', 'Schirmer', 'Schirmer', 'GSE180759', 'Schirmer', 'GSE180759', 'Schirmer', 'Schirmer', 'GSE180759', 'Schirmer', 'GSE180759', 'GSE180759', 'Schirmer', 'Schirmer', 'GSE180759', 'Schirmer', 'Schirmer', 'GSE180759', 'Schirmer', 'GSE180759', 'GSE180759', 'GSE180759', 'GSE180759', 'GSE180759', 'Schirmer', 'GSE180759', 'GSE180759', 'Schirmer', 'Schirmer', 'GSE180759', 'Schirmer', 'Schirmer', 'GSE180759', 'GSE180759', 'Schirmer', 'GSE180759', 'Schirmer', 'Schirmer', 'Schirmer', 'Schirmer', 'GSE180759', 'Schirmer', 'GSE180759', 'GSE180759', 'Schirmer', 'GSE180759', 'GSE180759', 'GSE180759', 'GSE180759', 'Schirmer', 'GSE180759', 'GSE180759', 'Schirmer', 'GSE180759', 'GSE180759', 'GSE180759', 'GSE180759', 'GSE180759', 'GSE180759', 'Schirmer', 'Schirmer', 'GSE180759', 'GSE180759', 'GSE180759', 'Schirmer', 'GSE180759', 'Schirmer', 'GSE180759', 'GSE180759', 'GSE180759', 'GSE180759', 'GSE180759', 'GSE180759', 'GSE180759', 'Schirmer', 'GSE180759', 'GSE180759', 'Schirmer', 'Schirmer', 'Schirmer', 'GSE180759', 'Schirmer', 'GSE180759', 'GSE180759', 'Schirmer', 'GSE180759', 'Schirmer', 'GSE180759', 'GSE180759', 'GSE180759', 'Schirmer', 'GSE180759', 'Schirmer', 'Schirmer', 'Schirmer', 'Schirmer', 'GSE180759', 'GSE180759', 'Schirmer', 'GSE180759', 'GSE180759', 'GSE180759', 'Schirmer', 'GSE180759', 'GSE180759', 'GSE180759', 'Schirmer', 'GSE180759', 'Schirmer', 'Schirmer', 'Schirmer', 'Schirmer', 'Schirmer', 'Schirmer', 'GSE180759', 'GSE180759', 'GSE180759', 'Schirmer', 'GSE180759', 'Schirmer', 'Schirmer', 'GSE180759', 'GSE180759', 'GSE180759', 'Schirmer', 'Schirmer', 'Schirmer', 'Schirmer', 'Schirmer', 'Schirmer', 'Schirmer', 'Schirmer', 'Schirmer', 'Schirmer', 'Schirmer', 'GSE180759', 'GSE180759', 'Schirmer', 'GSE180759', 'GSE180759', 'Schirmer', 'GSE180759', 'GSE180759', 'GSE180759', 'Schirmer', 'Schirmer', 'GSE180759', 'Schirmer', 'Schirmer', 'GSE180759', 'Schirmer', 'GSE180759', 'Schirmer', 'Schirmer', 'Schirmer', 'Schirmer', 'GSE180759', 'GSE180759', 'Schirmer', 'GSE180759', 'Schirmer', 'GSE180759', 'Schirmer', 'Schirmer', 'GSE180759', 'GSE180759', 'GSE180759', 'Schirmer', 'GSE180759', 'Schirmer', 'GSE180759', 'GSE180759', 'Schirmer', 'Schirmer', 'Schirmer', 'GSE180759', 'Schirmer', 'Schirmer', 'GSE180759', 'Schirmer', 'GSE180759', 'Schirmer', 'GSE180759', 'Schirmer', 'Schirmer', 'Schirmer', 'Schirmer', 'Schirmer', 'GSE180759', 'Schirmer', 'GSE180759', 'GSE180759', 'Schirmer', 'Schirmer', 'GSE180759', 'Schirmer', 'GSE180759', 'GSE180759', 'GSE180759', 'GSE180759', 'GSE180759', 'Schirmer', 'Schirmer', 'GSE180759', 'Schirmer', 'Schirmer', 'GSE180759', 'GSE180759', 'Schirmer', 'Schirmer', 'GSE180759', 'GSE180759', 'GSE180759', 'GSE180759', 'Schirmer', 'Schirmer', 'GSE180759', 'Schirmer', 'Schirmer', 'GSE180759', 'GSE180759', 'Schirmer', 'Schirmer', 'Schirmer', 'Schirmer', 'Schirmer', 'Schirmer', 'GSE180759', 'Schirmer', 'Schirmer', 'GSE180759', 'Schirmer', 'Schirmer', 'Schirmer', 'GSE180759', 'Schirmer', 'GSE180759', 'Schirmer', 'Schirmer', 'GSE180759', 'Schirmer', 'GSE180759', 'GSE180759', 'GSE180759', 'Schirmer', 'Schirmer', 'Schirmer', 'Schirmer', 'GSE180759', 'Schirmer', 'Schirmer', 'GSE180759', 'GSE180759', 'GSE180759', 'Schirmer', 'GSE180759', 'GSE180759', 'Schirmer', 'GSE180759', 'Schirmer', 'GSE180759', 'GSE180759', 'Schirmer', 'Schirmer', 'GSE180759', 'Schirmer', 'GSE180759', 'Schirmer', 'GSE180759', 'GSE180759', 'Schirmer', 'Schirmer', 'GSE180759', 'Schirmer', 'Schirmer', 'GSE180759', 'Schirmer', 'Schirmer', 'Schirmer', 'Schirmer', 'GSE180759', 'GSE180759', 'Schirmer', 'Schirmer', 'GSE180759', 'Schirmer', 'Schirmer', 'Schirmer', 'Schirmer', 'Schirmer', 'Schirmer', 'Schirmer', 'Schirmer', 'Schirmer', 'GSE180759', 'GSE180759', 'Schirmer', 'GSE180759', 'Schirmer', 'GSE180759', 'GSE180759', 'GSE180759', 'GSE180759', 'Schirmer', 'Schirmer', 'Schirmer', 'Schirmer', 'Schirmer', 'Schirmer', 'GSE180759', 'Schirmer', 'GSE180759', 'Schirmer', 'Schirmer', 'Schirmer', 'Schirmer', 'GSE180759', 'Schirmer', 'Schirmer', 'Schirmer', 'Schirmer', 'GSE180759', 'Schirmer', 'Schirmer', 'GSE180759', 'GSE180759', 'GSE180759', 'Schirmer', 'GSE180759', 'Schirmer', 'Schirmer', 'Schirmer', 'Schirmer', 'GSE180759', 'Schirmer', 'Schirmer', 'Schirmer', 'GSE180759', 'Schirmer', 'Schirmer', 'Schirmer', 'Schirmer', 'Schirmer', 'GSE180759', 'GSE180759', 'GSE180759', 'Schirmer', 'Schirmer', 'Schirmer', 'GSE180759', 'Schirmer', 'GSE180759', 'Schirmer', 'Schirmer', 'Schirmer', 'Schirmer', 'GSE180759', 'GSE180759', 'GSE180759', 'Schirmer', 'GSE180759', 'GSE180759', 'GSE180759', 'Schirmer', 'GSE180759', 'GSE180759', 'Schirmer', 'Schirmer', 'GSE180759', 'GSE180759', 'GSE180759', 'GSE180759', 'GSE180759', 'Schirmer', 'Schirmer', 'GSE180759', 'GSE180759', 'Schirmer', 'Schirmer', 'GSE180759', 'Schirmer', 'GSE180759', 'GSE180759', 'Schirmer', 'Schirmer', 'Schirmer', 'Schirmer', 'Schirmer', 'Schirmer', 'GSE180759', 'GSE180759', 'Schirmer', 'Schirmer', 'Schirmer', 'Schirmer', 'Schirmer', 'Schirmer', 'Schirmer', 'Schirmer', 'Schirmer', 'Schirmer', 'Schirmer', 'GSE180759', 'Schirmer', 'GSE180759', 'GSE180759', 'GSE180759', 'GSE180759', 'Schirmer', 'GSE180759', 'Schirmer', 'Schirmer', 'GSE180759', 'GSE180759', 'GSE180759', 'GSE180759', 'GSE180759', 'Schirmer', 'Schirmer', 'Schirmer', 'Schirmer', 'GSE180759', 'GSE180759', 'GSE180759', 'Schirmer', 'Schirmer', 'Schirmer', 'GSE180759', 'Schirmer', 'Schirmer', 'GSE180759', 'Schirmer', 'GSE180759', 'GSE180759', 'Schirmer', 'Schirmer', 'GSE180759', 'GSE180759', 'GSE180759', 'Schirmer', 'GSE180759', 'GSE180759', 'GSE180759', 'GSE180759', 'Schirmer', 'GSE180759', 'Schirmer', 'Schirmer', 'Schirmer', 'GSE180759', 'GSE180759', 'GSE180759', 'GSE180759', 'Schirmer', 'Schirmer', 'Schirmer', 'Schirmer', 'GSE180759', 'GSE180759', 'GSE180759', 'GSE180759', 'GSE180759', 'GSE180759', 'Schirmer', 'GSE180759', 'GSE180759', 'Schirmer', 'Schirmer', 'GSE180759', 'Schirmer', 'GSE180759', 'Schirmer', 'Schirmer', 'Schirmer', 'GSE180759', 'GSE180759', 'Schirmer', 'GSE180759', 'GSE180759', 'Schirmer', 'Schirmer', 'Schirmer', 'GSE180759', 'Schirmer', 'Schirmer', 'GSE180759', 'Schirmer', 'Schirmer', 'GSE180759', 'GSE180759', 'Schirmer', 'GSE180759', 'GSE180759', 'GSE180759', 'GSE180759', 'Schirmer', 'Schirmer', 'GSE180759', 'Schirmer', 'Schirmer', 'GSE180759', 'GSE180759', 'GSE180759', 'Schirmer', 'GSE180759', 'Schirmer', 'Schirmer', 'Schirmer', 'GSE180759', 'Schirmer', 'GSE180759', 'GSE180759', 'GSE180759', 'GSE180759', 'GSE180759', 'Schirmer', 'Schirmer', 'GSE180759', 'Schirmer', 'Schirmer', 'GSE180759', 'Schirmer', 'Schirmer', 'Schirmer', 'GSE180759', 'GSE180759', 'Schirmer', 'Schirmer', 'Schirmer', 'Schirmer', 'Schirmer', 'Schirmer', 'GSE180759', 'Schirmer', 'GSE180759', 'Schirmer', 'Schirmer', 'GSE180759', 'Schirmer', 'GSE180759', 'Schirmer', 'GSE180759', 'Schirmer', 'Schirmer', 'GSE180759', 'GSE180759', 'GSE180759', 'Schirmer', 'Schirmer', 'Schirmer', 'Schirmer', 'Schirmer', 'GSE180759', 'Schirmer', 'GSE180759', 'GSE180759', 'Schirmer', 'Schirmer', 'Schirmer', 'Schirmer', 'Schirmer', 'GSE180759', 'GSE180759', 'Schirmer', 'GSE180759', 'Schirmer', 'GSE180759', 'Schirmer', 'Schirmer', 'Schirmer', 'GSE180759', 'GSE180759', 'Schirmer', 'Schirmer', 'Schirmer', 'GSE180759', 'Schirmer', 'GSE180759', 'Schirmer', 'Schirmer', 'GSE180759', 'Schirmer', 'Schirmer', 'GSE180759', 'Schirmer', 'GSE180759', 'Schirmer', 'GSE180759', 'GSE180759', 'Schirmer', 'Schirmer', 'Schirmer', 'GSE180759', 'GSE180759', 'GSE180759', 'Schirmer', 'Schirmer', 'GSE180759', 'Schirmer', 'Schirmer', 'GSE180759', 'GSE180759', 'Schirmer', 'Schirmer', 'Schirmer', 'GSE180759', 'Schirmer', 'Schirmer', 'GSE180759', 'GSE180759', 'Schirmer', 'GSE180759', 'GSE180759', 'GSE180759', 'Schirmer', 'GSE180759', 'GSE180759', 'GSE180759', 'Schirmer', 'Schirmer', 'Schirmer', 'Schirmer', 'Schirmer', 'GSE180759', 'GSE180759', 'GSE180759', 'GSE180759', 'GSE180759', 'GSE180759', 'GSE180759', 'Schirmer', 'Schirmer', 'GSE180759', 'Schirmer', 'Schirmer', 'GSE180759', 'Schirmer', 'Schirmer', 'Schirmer', 'Schirmer', 'Schirmer', 'Schirmer', 'GSE180759', 'GSE180759', 'Schirmer', 'GSE180759', 'Schirmer', 'Schirmer', 'GSE180759', 'GSE180759', 'GSE180759', 'Schirmer', 'Schirmer', 'GSE180759', 'GSE180759', 'Schirmer', 'GSE180759', 'Schirmer', 'GSE180759', 'Schirmer', 'GSE180759', 'Schirmer', 'GSE180759', 'GSE180759', 'GSE180759', 'Schirmer', 'Schirmer', 'Schirmer', 'Schirmer', 'GSE180759', 'Schirmer', 'Schirmer', 'GSE180759', 'Schirmer', 'Schirmer', 'GSE180759', 'Schirmer', 'GSE180759', 'Schirmer', 'Schirmer', 'GSE180759', 'GSE180759', 'GSE180759', 'Schirmer', 'Schirmer', 'GSE180759', 'Schirmer', 'Schirmer', 'GSE180759', 'GSE180759', 'GSE180759', 'GSE180759', 'Schirmer', 'Schirmer', 'Schirmer', 'GSE180759', 'Schirmer', 'Schirmer', 'GSE180759', 'GSE180759', 'Schirmer', 'Schirmer', 'Schirmer', 'GSE180759', 'Schirmer', 'GSE180759', 'Schirmer', 'GSE180759', 'Schirmer', 'Schirmer', 'Schirmer', 'Schirmer', 'Schirmer', 'Schirmer', 'Schirmer', 'Schirmer', 'Schirmer', 'Schirmer', 'Schirmer', 'GSE180759', 'GSE180759', 'GSE180759', 'Schirmer', 'Schirmer', 'Schirmer', 'GSE180759', 'Schirmer', 'Schirmer', 'GSE180759', 'Schirmer', 'Schirmer', 'GSE180759', 'Schirmer', 'Schirmer', 'Schirmer', 'Schirmer', 'Schirmer', 'GSE180759', 'GSE180759', 'GSE180759', 'Schirmer', 'Schirmer', 'Schirmer', 'Schirmer', 'GSE180759', 'Schirmer', 'GSE180759', 'GSE180759', 'GSE180759', 'Schirmer', 'Schirmer', 'GSE180759', 'GSE180759', 'GSE180759', 'Schirmer', 'Schirmer', 'Schirmer', 'GSE180759', 'Schirmer', 'Schirmer', 'Schirmer', 'Schirmer', 'Schirmer', 'Schirmer', 'Schirmer', 'Schirmer', 'Schirmer', 'Schirmer', 'Schirmer', 'Schirmer', 'Schirmer', 'Schirmer', 'Schirmer', 'Schirmer', 'GSE180759', 'Schirmer', 'GSE180759', 'GSE180759', 'GSE180759', 'GSE180759', 'Schirmer', 'Schirmer', 'Schirmer', 'Schirmer', 'Schirmer', 'GSE180759', 'GSE180759', 'GSE180759', 'GSE180759', 'Schirmer', 'Schirmer', 'GSE180759', 'Schirmer', 'Schirmer', 'GSE180759', 'Schirmer', 'Schirmer', 'Schirmer', 'Schirmer', 'GSE180759', 'GSE180759', 'Schirmer', 'Schirmer', 'GSE180759', 'GSE180759', 'Schirmer', 'Schirmer', 'Schirmer', 'GSE180759', 'GSE180759', 'GSE180759', 'GSE180759', 'GSE180759', 'Schirmer', 'Schirmer', 'Schirmer', 'GSE180759', 'Schirmer', 'Schirmer', 'Schirmer', 'Schirmer', 'GSE180759', 'Schirmer', 'Schirmer', 'GSE180759', 'Schirmer', 'GSE180759', 'GSE180759', 'Schirmer', 'Schirmer', 'Schirmer', 'Schirmer', 'Schirmer', 'GSE180759', 'Schirmer', 'GSE180759', 'Schirmer', 'Schirmer', 'Schirmer', 'Schirmer', 'GSE180759', 'Schirmer', 'Schirmer', 'Schirmer', 'GSE180759', 'GSE180759', 'GSE180759', 'Schirmer', 'Schirmer', 'Schirmer', 'Schirmer', 'Schirmer', 'Schirmer', 'GSE180759', 'Schirmer', 'GSE180759', 'Schirmer', 'GSE180759', 'GSE180759', 'Schirmer', 'GSE180759', 'Schirmer', 'GSE180759', 'GSE180759', 'GSE180759', 'Schirmer', 'Schirmer', 'Schirmer', 'Schirmer', 'Schirmer', 'Schirmer', 'Schirmer', 'Schirmer', 'GSE180759', 'Schirmer', 'GSE180759', 'GSE180759', 'Schirmer', 'GSE180759', 'GSE180759', 'GSE180759', 'GSE180759', 'Schirmer', 'Schirmer', 'Schirmer', 'Schirmer', 'GSE180759', 'GSE180759', 'Schirmer', 'Schirmer', 'Schirmer', 'GSE180759', 'Schirmer', 'Schirmer', 'GSE180759', 'GSE180759', 'Schirmer', 'Schirmer', 'GSE180759', 'GSE180759', 'Schirmer', 'GSE180759', 'Schirmer', 'Schirmer', 'Schirmer', 'Schirmer', 'Schirmer', 'GSE180759', 'Schirmer', 'Schirmer', 'Schirmer', 'GSE180759', 'GSE180759', 'GSE180759', 'Schirmer', 'Schirmer', 'GSE180759', 'GSE180759', 'GSE180759', 'Schirmer', 'Schirmer', 'Schirmer', 'Schirmer', 'Schirmer', 'Schirmer', 'Schirmer', 'GSE180759', 'Schirmer', 'GSE180759', 'Schirmer', 'GSE180759', 'Schirmer', 'GSE180759', 'Schirmer', 'GSE180759', 'Schirmer', 'Schirmer', 'Schirmer', 'Schirmer', 'Schirmer', 'Schirmer', 'Schirmer', 'Schirmer', 'Schirmer', 'Schirmer', 'Schirmer', 'GSE180759', 'GSE180759', 'GSE180759', 'GSE180759', 'GSE180759', 'GSE180759', 'GSE180759', 'GSE180759', 'GSE180759', 'GSE180759', 'GSE180759', 'Schirmer', 'Schirmer', 'Schirmer', 'Schirmer', 'Schirmer', 'Schirmer', 'GSE180759', 'Schirmer', 'GSE180759', 'Schirmer', 'Schirmer', 'Schirmer', 'Schirmer', 'Schirmer', 'Schirmer', 'Schirmer', 'Schirmer', 'GSE180759', 'Schirmer', 'Schirmer', 'GSE180759', 'GSE180759', 'Schirmer', 'Schirmer', 'Schirmer', 'Schirmer', 'Schirmer', 'GSE180759', 'GSE180759', 'GSE180759', 'GSE180759', 'GSE180759', 'Schirmer', 'GSE180759', 'GSE180759', 'Schirmer', 'Schirmer', 'GSE180759', 'GSE180759', 'Schirmer', 'GSE180759', 'GSE180759']]\n"
     ]
    }
   ],
   "source": [
    "for batch in train_dataloader:\n",
    "    print(batch)  # 打印批次内容\n",
    "    break "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "49da665c-ab32-499b-acee-2687e4787b2b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "torch.Size([1000, 34667])"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "batch[0].shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "57abc525-9a13-4ca3-9ed1-9abf582d4d1c",
   "metadata": {},
   "outputs": [],
   "source": [
    "class TripletSelector:\n",
    "    \"\"\"For each anchor-positive pair, mine negative samples to create a triplet.\"\"\"\n",
    "\n",
    "    def __init__(\n",
    "        self,\n",
    "        margin: float,\n",
    "        negative_selection: str = \"semihard\",\n",
    "        perturb_labels: bool = True,\n",
    "        perturb_labels_fraction: float = 0.5,\n",
    "    ):\n",
    "        \"\"\"Constructor.\n",
    "\n",
    "        Parameters\n",
    "        ----------\n",
    "        margin: float\n",
    "            Triplet loss margin.\n",
    "        negative_selection: str, default: \"semihard\"\n",
    "            Method for negative selection: {\"semihard\", \"hardest\", \"random\"}.\n",
    "        perturb_labels: bool, default: True\n",
    "            Whether to perturb the ontology labels by coarse graining one level up.\n",
    "        perturb_labels_fraction: float, default: 0.5\n",
    "            The fraction of labels to perturb.\n",
    "\n",
    "        Examples\n",
    "        --------\n",
    "        >>> triplet_selector = TripletSelector(margin=0.05,\n",
    "                negative_selection=\"semihard\",\n",
    "                perturb_labels=True,\n",
    "                perturb_labels_fraction=0.5,\n",
    "            )\n",
    "        \"\"\"\n",
    "\n",
    "        self.margin = margin\n",
    "        self.negative_selection = negative_selection\n",
    "\n",
    "        self.onto = import_cell_ontology()\n",
    "        self.id2name = get_id_mapper(self.onto)\n",
    "        self.name2id = {value: key for key, value in self.id2name.items()}\n",
    "\n",
    "        self.perturb_labels = perturb_labels\n",
    "        self.perturb_labels_fraction = perturb_labels_fraction\n",
    "\n",
    "    def get_triplets_idx(\n",
    "        self,\n",
    "        embeddings: Union[np.ndarray, torch.Tensor],\n",
    "        labels: Union[np.ndarray, torch.Tensor],\n",
    "        int2label: dict,\n",
    "        studies: Optional[Union[np.ndarray, torch.Tensor, list]] = None,\n",
    "    ):\n",
    "        \"\"\"Get triplets as anchor, positive, and negative cell indices.\n",
    "\n",
    "        Parameters\n",
    "        ----------\n",
    "        embeddings: numpy.ndarray, torch.Tensor\n",
    "            Cell embeddings.\n",
    "        labels: numpy.ndarray, torch.Tensor\n",
    "            Cell labels in integer form.\n",
    "        int2label: dict\n",
    "            Dictionary to map labels in integer form to string\n",
    "        studies: numpy.ndarray, torch.Tensor, optional, default: None\n",
    "            Studies metadata for each cell.\n",
    "\n",
    "        Returns\n",
    "        -------\n",
    "        triplets: Tuple[List, List, List]\n",
    "            A tuple of lists containing anchor, positive, and negative cell indices.\n",
    "        num_hard_triplets: int\n",
    "            Number of hard triplets.\n",
    "        num_viable_triplets: int\n",
    "            Number of viable triplets.\n",
    "        )\n",
    "        \"\"\"\n",
    "\n",
    "        if isinstance(embeddings, torch.Tensor):\n",
    "            distance_matrix = self.pdist(embeddings.detach().cpu().numpy())\n",
    "        else:\n",
    "            distance_matrix = self.pdist(embeddings)\n",
    "\n",
    "        if isinstance(labels, torch.Tensor):\n",
    "            labels = labels.detach().cpu().numpy()\n",
    "\n",
    "        if studies is not None and isinstance(studies, torch.Tensor):\n",
    "            studies = studies.detach().cpu().numpy()\n",
    "\n",
    "        labels_ids = np.array([self.name2id[int2label[label]] for label in labels])\n",
    "        labels_set = set(labels)\n",
    "\n",
    "        if self.perturb_labels:\n",
    "            labels_ids_set = set(labels_ids.tolist())\n",
    "            label2int = {value: key for key, value in int2label.items()}\n",
    "            perturb_list = random.choices(\n",
    "                np.arange(len(labels)),\n",
    "                k=int(len(labels) * self.perturb_labels_fraction),\n",
    "            )\n",
    "            for i in perturb_list:  # cells chosen for perturbation of labels\n",
    "                term_id = self.name2id[int2label[labels[i]]]\n",
    "                parents = set()\n",
    "                max_ancestors = (\n",
    "                    1  # range of ancestor levels. 1: parents, 2: grandparents, ...\n",
    "                )\n",
    "                ancestor_level = 0\n",
    "                while ancestor_level < max_ancestors:\n",
    "                    ancestor_level += 1\n",
    "                    if not parents:\n",
    "                        parents = get_parents(self.onto, term_id)\n",
    "                    else:\n",
    "                        current = set()\n",
    "                        for p in parents:\n",
    "                            current = current | get_parents(self.onto, p)\n",
    "                        parents = current\n",
    "                    found = any((parent in labels_ids_set for parent in parents))\n",
    "                    if found is True:\n",
    "                        parents = list(parents)\n",
    "                        np.random.shuffle(parents)\n",
    "                        p = next(\n",
    "                            parent for parent in parents if parent in labels_ids_set\n",
    "                        )\n",
    "                        labels[i] = label2int[self.id2name[p]]\n",
    "                        break  # label perturbed, skip the rest of the ancestors\n",
    "\n",
    "        triplets = []\n",
    "        num_hard_triplets = 0\n",
    "        num_viable_triplets = 0\n",
    "        for label in labels_set:\n",
    "            term_id = self.name2id[int2label[label]]\n",
    "            ancestors = get_all_ancestors(self.onto, term_id)\n",
    "            descendants = get_all_descendants(self.onto, term_id)\n",
    "            violating_terms = ancestors.union(descendants)\n",
    "\n",
    "            label_mask = labels == label\n",
    "            label_indices = np.where(label_mask)[0]\n",
    "            if len(label_indices) < 2:\n",
    "                continue\n",
    "\n",
    "            # negatives are labels that are not the same as current and not in violating terms\n",
    "            negative_indices = np.where(\n",
    "                np.logical_not(label_mask | np.isin(labels_ids, list(violating_terms)))\n",
    "            )[0]\n",
    "\n",
    "            # compute all pairs of anchor-positives\n",
    "            anchor_positives = list(combinations(label_indices, 2))\n",
    "\n",
    "            # enforce anchor and positive coming from different studies\n",
    "            if studies is not None:\n",
    "                anchor_positives = [\n",
    "                    (anchor, positive)\n",
    "                    for anchor, positive in anchor_positives\n",
    "                    if studies[anchor] != studies[positive]\n",
    "                ]\n",
    "\n",
    "            for anchor_positive in anchor_positives:\n",
    "                loss_values = (\n",
    "                    distance_matrix[anchor_positive[0], anchor_positive[1]]\n",
    "                    - distance_matrix[[anchor_positive[0]], negative_indices]\n",
    "                    + self.margin\n",
    "                )\n",
    "                num_hard_triplets += (loss_values > 0).sum()\n",
    "                num_viable_triplets += loss_values.size\n",
    "\n",
    "                # select one negative for anchor positive pair based on selection function\n",
    "                if self.negative_selection == \"semihard\":\n",
    "                    hard_negative = self.semihard_negative(loss_values)\n",
    "                elif self.negative_selection == \"hardest\":\n",
    "                    hard_negative = self.hardest_negative(loss_values)\n",
    "                elif self.negative_selection == \"random\":\n",
    "                    hard_negative = self.random_negative(loss_values)\n",
    "                else:\n",
    "                    hard_negative = None\n",
    "\n",
    "                if hard_negative is not None:\n",
    "                    hard_negative = negative_indices[hard_negative]\n",
    "                    triplets.append(\n",
    "                        [anchor_positive[0], anchor_positive[1], hard_negative]\n",
    "                    )\n",
    "\n",
    "        if len(triplets) == 0:\n",
    "            triplets.append([0, 0, 0])\n",
    "\n",
    "        anchor_idx, positive_idx, negative_idx = tuple(\n",
    "            map(list, zip(*triplets))\n",
    "        )  # tuple([list(t) for t in zip(*triplets)])\n",
    "\n",
    "        return (\n",
    "            (\n",
    "                anchor_idx,\n",
    "                positive_idx,\n",
    "                negative_idx,\n",
    "            ),\n",
    "            num_hard_triplets,\n",
    "            num_viable_triplets,\n",
    "        )\n",
    "\n",
    "\n",
    "    def get_triplets(\n",
    "        self,\n",
    "        embeddings: Union[np.ndarray, torch.Tensor],\n",
    "        labels: Union[np.ndarray, torch.Tensor],\n",
    "        int2label: dict,\n",
    "        studies: Optional[Union[np.ndarray, torch.Tensor, list]] = None,\n",
    "    ):\n",
    "        \"\"\"Get triplets as anchor, positive, and negative cell embeddings.\n",
    "\n",
    "        Parameters\n",
    "        ----------\n",
    "        embeddings: numpy.ndarray, torch.Tensor\n",
    "            Cell embeddings.\n",
    "        labels: numpy.ndarray, torch.Tensor\n",
    "            Cell labels in integer form.\n",
    "        int2label: dict\n",
    "            Dictionary to map labels in integer form to string\n",
    "        studies: numpy.ndarray, torch.Tensor, optional, default: None\n",
    "            Studies metadata for each cell.\n",
    "\n",
    "        Returns\n",
    "        -------\n",
    "        triplets: Tuple[numpy.ndarray, numpy.ndarray, numpy.ndarray]\n",
    "            A tuple of numpy arrays containing anchor, positive, and negative cell embeddings.\n",
    "        num_hard_triplets: int\n",
    "            Number of hard triplets.\n",
    "        num_viable_triplets: int\n",
    "            Number of viable triplets.\n",
    "        \"\"\"\n",
    "\n",
    "        (\n",
    "            triplets_idx,\n",
    "            num_hard_triplets,\n",
    "            num_viable_triplets,\n",
    "        ) = self.get_triplets_idx(embeddings, labels, int2label, studies)\n",
    "        anchor_idx, positive_idx, negative_idx = triplets_idx\n",
    "        return (\n",
    "            (\n",
    "                embeddings[anchor_idx],\n",
    "                embeddings[positive_idx],\n",
    "                embeddings[negative_idx],\n",
    "            ),\n",
    "            num_hard_triplets,\n",
    "            num_viable_triplets,\n",
    "        )\n",
    "\n",
    "\n",
    "    def pdist(self, vectors: np.ndarray):\n",
    "        \"\"\"Get pair-wise distance between all cell embeddings.\n",
    "\n",
    "        Parameters\n",
    "        ----------\n",
    "        vectors: numpy.ndarray\n",
    "            Cell embeddings.\n",
    "\n",
    "        Returns\n",
    "        -------\n",
    "        numpy.ndarray\n",
    "            Distance matrix of cell embeddings.\n",
    "        \"\"\"\n",
    "\n",
    "        vectors_squared_sum = (vectors**2).sum(axis=1)\n",
    "        distance_matrix = (\n",
    "            -2 * np.matmul(vectors, np.matrix.transpose(vectors))\n",
    "            + vectors_squared_sum.reshape(1, -1)\n",
    "            + vectors_squared_sum.reshape(-1, 1)\n",
    "        )\n",
    "        return distance_matrix\n",
    "\n",
    "\n",
    "    def hardest_negative(self, loss_values):\n",
    "        \"\"\"Get hardest negative.\n",
    "\n",
    "        Parameters\n",
    "        ----------\n",
    "        loss_values: numpy.ndarray\n",
    "            Triplet loss of all negatives for given anchor positive pair.\n",
    "\n",
    "        Returns\n",
    "        -------\n",
    "        int\n",
    "            Index of selection.\n",
    "        \"\"\"\n",
    "\n",
    "        hard_negative = np.argmax(loss_values)\n",
    "        return hard_negative if loss_values[hard_negative] > 0 else None\n",
    "\n",
    "\n",
    "    def random_negative(self, loss_values):\n",
    "        \"\"\"Get random negative.\n",
    "\n",
    "        Parameters\n",
    "        ----------\n",
    "        loss_values: numpy.ndarray\n",
    "            Triplet loss of all negatives for given anchor positive pair.\n",
    "\n",
    "        Returns\n",
    "        -------\n",
    "        int\n",
    "            Index of selection.\n",
    "        \"\"\"\n",
    "\n",
    "        hard_negatives = np.where(loss_values > 0)[0]\n",
    "        return np.random.choice(hard_negatives) if len(hard_negatives) > 0 else None\n",
    "\n",
    "\n",
    "    def semihard_negative(self, loss_values):\n",
    "        \"\"\"Get a random semihard negative.\n",
    "\n",
    "        Parameters\n",
    "        ----------\n",
    "        loss_values: numpy.ndarray\n",
    "            Triplet loss of all negatives for given anchor positive pair.\n",
    "\n",
    "        Returns\n",
    "        -------\n",
    "        int\n",
    "            Index of selection.\n",
    "        \"\"\"\n",
    "\n",
    "        semihard_negatives = np.where(\n",
    "            np.logical_and(loss_values < self.margin, loss_values > 0)\n",
    "        )[0]\n",
    "        return (\n",
    "            np.random.choice(semihard_negatives)\n",
    "            if len(semihard_negatives) > 0\n",
    "            else None\n",
    "        )\n",
    "\n",
    "\n",
    "    def get_asw(\n",
    "        self,\n",
    "        embeddings: Union[np.ndarray, torch.Tensor],\n",
    "        labels: List[str],\n",
    "        int2label: dict,\n",
    "        metric: str = \"cosine\",\n",
    "    ) -> float:\n",
    "        \"\"\"Get the average silhouette width of celltypes, being aware of cell ontology such that\n",
    "           ancestors are not considered inter-cluster and descendants are considered intra-cluster.\n",
    "\n",
    "        Parameters\n",
    "        ----------\n",
    "        embeddings: numpy.ndarray, torch.Tensor\n",
    "            Cell embeddings.\n",
    "        labels: List[str]\n",
    "            Celltype names.\n",
    "        int2label: dict\n",
    "            Dictionary to map labels in integer form to string\n",
    "        metric: str, default: \"cosine\"\n",
    "            The distance metric to use for scipy.spatial.distance.cdist().\n",
    "\n",
    "        Returns\n",
    "        -------\n",
    "        asw: float\n",
    "            The average silhouette width.\n",
    "\n",
    "        Examples\n",
    "        --------\n",
    "        >>> asw = ontology_silhouette_width(embeddings, labels, metric=\"cosine\")\n",
    "        \"\"\"\n",
    "\n",
    "        if isinstance(embeddings, torch.Tensor):\n",
    "            distance_matrix = self.pdist(embeddings.detach().cpu().numpy())\n",
    "        else:\n",
    "            distance_matrix = self.pdist(embeddings)\n",
    "\n",
    "        if isinstance(labels, torch.Tensor):\n",
    "            labels = labels.detach().cpu().numpy()\n",
    "\n",
    "        sw = []\n",
    "        for i, label1 in enumerate(labels):\n",
    "            term_id1 = self.name2id[int2label[label1]]\n",
    "            ancestors = get_all_ancestors(self.onto, term_id1)\n",
    "            descendants = get_all_descendants(self.onto, term_id1)\n",
    "\n",
    "            a_i = []\n",
    "            b_i = {}\n",
    "            for j, label2 in enumerate(labels):\n",
    "                if i == j:\n",
    "                    continue\n",
    "\n",
    "                term_id2 = self.name2id[int2label[label2]]\n",
    "                if term_id2 == term_id1 or term_id2 in descendants:  # intra-cluster\n",
    "                    a_i.append(distance_matrix[i, j])\n",
    "                elif (\n",
    "                    term_id2 != term_id1 and term_id2 not in ancestors\n",
    "                ):  # inter-cluster\n",
    "                    if term_id2 not in b_i:\n",
    "                        b_i[term_id2] = []\n",
    "                    b_i[term_id2].append(distance_matrix[i, j])\n",
    "\n",
    "            if len(a_i) <= 1 or not b_i:\n",
    "                continue\n",
    "            a_i = np.sum(a_i) / (len(a_i) - 1)\n",
    "            b_i = np.min(\n",
    "                [\n",
    "                    np.sum(values) / len(values)\n",
    "                    for values in b_i.values()\n",
    "                    if len(values) > 1\n",
    "                ]\n",
    "            )\n",
    "\n",
    "            s_i = (b_i - a_i) / np.max([a_i, b_i])\n",
    "            sw.append(s_i)\n",
    "        return np.mean(sw)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "5f563f30-9214-4cbb-a52d-1a9f2c3d6d26",
   "metadata": {},
   "outputs": [],
   "source": [
    "triplet_selector = TripletSelector(margin=0.05,\n",
    "                 negative_selection=\"semihard\",\n",
    "                 perturb_labels=True,\n",
    "                 perturb_labels_fraction=0.5 )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "b7fbf3fe-e832-46ef-89dc-9148d94925e7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{0: 'oligodendrocyte',\n",
       " 1: 'lymphocyte',\n",
       " 2: 'oligodendrocyte precursor cell',\n",
       " 3: 'phagocyte',\n",
       " 4: 'endothelial cell',\n",
       " 5: 'microglial cell',\n",
       " 6: 'neuron',\n",
       " 7: 'astrocyte'}"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "int2label"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "06053a50-8515-48c6-a909-af61547460a5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'oligodendrocyte': 0,\n",
       " 'lymphocyte': 1,\n",
       " 'oligodendrocyte precursor cell': 2,\n",
       " 'phagocyte': 3,\n",
       " 'endothelial cell': 4,\n",
       " 'microglial cell': 5,\n",
       " 'neuron': 6,\n",
       " 'astrocyte': 7}"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "label2int"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "a4bb7c00-f691-42fe-bd4c-b4444fc9463a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{0: 'oligodendrocyte',\n",
       " 1: 'lymphocyte',\n",
       " 2: 'oligodendrocyte precursor cell',\n",
       " 3: 'phagocyte',\n",
       " 4: 'endothelial cell',\n",
       " 5: 'microglial cell',\n",
       " 6: 'neuron',\n",
       " 7: 'astrocyte'}"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "int2label  # 这是一个形状为 (N,) 的数组或张量，包含每个样本的标签"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "fe361b6c-2d09-4789-b5ea-5ee2a39b79bc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2810"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(triplet_selector.id2name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "60ddf264-d84c-4d2a-95e6-9c44386475ae",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'CL:0000127'"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "triplet_selector.name2id.get('astrocyte')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "8debb5c8-ca1b-4ce4-a20d-7326c00f8767",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'cell': 'CL:0000000',\n",
       " 'primary cultured cell': 'CL:0000001',\n",
       " 'neural crest derived fibroblast': 'CL:0000005',\n",
       " 'neuronal receptor cell': 'CL:0000006',\n",
       " 'early embryonic cell (metazoa)': 'CL:0000007',\n",
       " 'migratory cranial neural crest cell': 'CL:0000008',\n",
       " 'cultured cell': 'CL:0000010',\n",
       " 'migratory trunk neural crest cell': 'CL:0000011',\n",
       " 'germ line stem cell': 'CL:0000014',\n",
       " 'male germ cell': 'CL:0000015',\n",
       " 'male germ line stem cell': 'CL:0000016',\n",
       " 'spermatocyte': 'CL:0000017',\n",
       " 'spermatid': 'CL:0000018',\n",
       " 'sperm': 'CL:0000019',\n",
       " 'spermatogonium': 'CL:0000020',\n",
       " 'female germ cell': 'CL:0000021',\n",
       " 'female germ line stem cell': 'CL:0000022',\n",
       " 'oocyte': 'CL:0000023',\n",
       " 'oogonial cell': 'CL:0000024',\n",
       " 'egg cell': 'CL:0000025',\n",
       " 'invertebrate nurse cell': 'CL:0000026',\n",
       " 'smooth muscle cell neural crest derived': 'CL:0000027',\n",
       " 'CNS neuron (sensu Nematoda and Protostomia)': 'CL:0000028',\n",
       " 'neural crest derived neuron': 'CL:0000029',\n",
       " 'glioblast': 'CL:0000030',\n",
       " 'neuroblast (sensu Vertebrata)': 'CL:0000031',\n",
       " 'neuroplacodal cell': 'CL:0000032',\n",
       " 'apocrine cell': 'CL:0000033',\n",
       " 'stem cell': 'CL:0000034',\n",
       " 'single fate stem cell': 'CL:0000035',\n",
       " 'epithelial fate stem cell': 'CL:0000036',\n",
       " 'hematopoietic stem cell': 'CL:0000037',\n",
       " 'erythroid progenitor cell': 'CL:0000038',\n",
       " 'germ line cell': 'CL:0000039',\n",
       " 'monoblast': 'CL:0000040',\n",
       " 'mature eosinophil': 'CL:0000041',\n",
       " 'neutrophilic myeloblast': 'CL:0000042',\n",
       " 'mature basophil': 'CL:0000043',\n",
       " 'neural stem cell': 'CL:0000047',\n",
       " 'multi fate stem cell': 'CL:0000048',\n",
       " 'common myeloid progenitor': 'CL:0000049',\n",
       " 'megakaryocyte-erythroid progenitor cell': 'CL:0000050',\n",
       " 'common lymphoid progenitor': 'CL:0000051',\n",
       " 'totipotent stem cell': 'CL:0000052',\n",
       " 'bone matrix secreting cell': 'CL:0000054',\n",
       " 'non-terminally differentiated cell': 'CL:0000055',\n",
       " 'myoblast': 'CL:0000056',\n",
       " 'fibroblast': 'CL:0000057',\n",
       " 'chondroblast': 'CL:0000058',\n",
       " 'ameloblast': 'CL:0000059',\n",
       " 'odontoblast': 'CL:0000060',\n",
       " 'cementoblast': 'CL:0000061',\n",
       " 'osteoblast': 'CL:0000062',\n",
       " 'ciliated cell': 'CL:0000064',\n",
       " 'ependymal cell': 'CL:0000065',\n",
       " 'epithelial cell': 'CL:0000066',\n",
       " 'ciliated epithelial cell': 'CL:0000067',\n",
       " 'duct epithelial cell': 'CL:0000068',\n",
       " 'branched duct epithelial cell': 'CL:0000069',\n",
       " 'blood vessel endothelial cell': 'CL:0000071',\n",
       " 'non-branched duct epithelial cell': 'CL:0000072',\n",
       " 'columnar/cuboidal epithelial cell': 'CL:0000075',\n",
       " 'squamous epithelial cell': 'CL:0000076',\n",
       " 'mesothelial cell': 'CL:0000077',\n",
       " 'peridermal cell': 'CL:0000078',\n",
       " 'stratified epithelial cell': 'CL:0000079',\n",
       " 'circulating cell': 'CL:0000080',\n",
       " 'blood cell': 'CL:0000081',\n",
       " 'epithelial cell of lung': 'CL:0000082',\n",
       " 'epithelial cell of pancreas': 'CL:0000083',\n",
       " 'T cell': 'CL:0000084',\n",
       " 'Kupffer cell': 'CL:0000091',\n",
       " 'osteoclast': 'CL:0000092',\n",
       " 'granulocyte': 'CL:0000094',\n",
       " 'neuron associated cell': 'CL:0000095',\n",
       " 'mature neutrophil': 'CL:0000096',\n",
       " 'mast cell': 'CL:0000097',\n",
       " 'sensory epithelial cell': 'CL:0000098',\n",
       " 'interneuron': 'CL:0000099',\n",
       " 'motor neuron': 'CL:0000100',\n",
       " 'sensory neuron': 'CL:0000101',\n",
       " 'polymodal neuron': 'CL:0000102',\n",
       " 'bipolar neuron': 'CL:0000103',\n",
       " 'multipolar neuron': 'CL:0000104',\n",
       " 'pseudounipolar neuron': 'CL:0000105',\n",
       " 'unipolar neuron': 'CL:0000106',\n",
       " 'autonomic neuron': 'CL:0000107',\n",
       " 'cholinergic neuron': 'CL:0000108',\n",
       " 'adrenergic neuron': 'CL:0000109',\n",
       " 'peptidergic neuron': 'CL:0000110',\n",
       " 'columnar neuron': 'CL:0000112',\n",
       " 'mononuclear phagocyte': 'CL:0000113',\n",
       " 'surface ectodermal cell': 'CL:0000114',\n",
       " 'endothelial cell': 'CL:0000115',\n",
       " 'pioneer neuron': 'CL:0000116',\n",
       " 'CNS neuron (sensu Vertebrata)': 'CL:0000117',\n",
       " 'basket cell': 'CL:0000118',\n",
       " 'cerebellar Golgi cell': 'CL:0000119',\n",
       " 'granule cell': 'CL:0000120',\n",
       " 'Purkinje cell': 'CL:0000121',\n",
       " 'stellate neuron': 'CL:0000122',\n",
       " 'neuron associated cell (sensu Vertebrata)': 'CL:0000123',\n",
       " 'glial cell': 'CL:0000125',\n",
       " 'macroglial cell': 'CL:0000126',\n",
       " 'astrocyte': 'CL:0000127',\n",
       " 'oligodendrocyte': 'CL:0000128',\n",
       " 'microglial cell': 'CL:0000129',\n",
       " 'neuron associated cell (sensu Nematoda and Protostomia)': 'CL:0000130',\n",
       " 'gut endothelial cell': 'CL:0000131',\n",
       " 'corneal endothelial cell': 'CL:0000132',\n",
       " 'neurectodermal cell': 'CL:0000133',\n",
       " 'mesenchymal stem cell': 'CL:0000134',\n",
       " 'fibrocyte': 'CL:0000135',\n",
       " 'adipocyte': 'CL:0000136',\n",
       " 'osteocyte': 'CL:0000137',\n",
       " 'chondrocyte': 'CL:0000138',\n",
       " 'odontocyte': 'CL:0000140',\n",
       " 'cementocyte': 'CL:0000141',\n",
       " 'hyalocyte': 'CL:0000142',\n",
       " 'guidepost cell': 'CL:0000143',\n",
       " 'professional antigen presenting cell': 'CL:0000145',\n",
       " 'simple columnar epithelial cell': 'CL:0000146',\n",
       " 'pigment cell': 'CL:0000147',\n",
       " 'melanocyte': 'CL:0000148',\n",
       " 'visual pigment cell': 'CL:0000149',\n",
       " 'glandular epithelial cell': 'CL:0000150',\n",
       " 'secretory cell': 'CL:0000151',\n",
       " 'exocrine cell': 'CL:0000152',\n",
       " 'glycosaminoglycan secreting cell': 'CL:0000153',\n",
       " 'protein secreting cell': 'CL:0000154',\n",
       " 'peptic cell': 'CL:0000155',\n",
       " 'surfactant secreting cell': 'CL:0000157',\n",
       " 'club cell': 'CL:0000158',\n",
       " 'seromucus secreting cell': 'CL:0000159',\n",
       " 'goblet cell': 'CL:0000160',\n",
       " 'parietal cell': 'CL:0000162',\n",
       " 'endocrine cell': 'CL:0000163',\n",
       " 'enteroendocrine cell': 'CL:0000164',\n",
       " 'neuroendocrine cell': 'CL:0000165',\n",
       " 'chromaffin cell': 'CL:0000166',\n",
       " 'peptide hormone secreting cell': 'CL:0000167',\n",
       " 'insulin secreting cell': 'CL:0000168',\n",
       " 'type B pancreatic cell': 'CL:0000169',\n",
       " 'glucagon secreting cell': 'CL:0000170',\n",
       " 'pancreatic A cell': 'CL:0000171',\n",
       " 'somatostatin secreting cell': 'CL:0000172',\n",
       " 'pancreatic D cell': 'CL:0000173',\n",
       " 'steroid hormone secreting cell': 'CL:0000174',\n",
       " 'luteal cell': 'CL:0000175',\n",
       " 'ecdysteroid secreting cell': 'CL:0000176',\n",
       " 'testosterone secreting cell': 'CL:0000177',\n",
       " 'Leydig cell': 'CL:0000178',\n",
       " 'progesterone secreting cell': 'CL:0000179',\n",
       " 'estradiol secreting cell': 'CL:0000180',\n",
       " 'hepatocyte': 'CL:0000182',\n",
       " 'contractile cell': 'CL:0000183',\n",
       " 'myoepithelial cell': 'CL:0000185',\n",
       " 'myofibroblast cell': 'CL:0000186',\n",
       " 'muscle cell': 'CL:0000187',\n",
       " 'cell of skeletal muscle': 'CL:0000188',\n",
       " 'slow muscle cell': 'CL:0000189',\n",
       " 'fast muscle cell': 'CL:0000190',\n",
       " 'smooth muscle cell': 'CL:0000192',\n",
       " 'cardiac muscle cell (sensu Arthopoda)': 'CL:0000193',\n",
       " 'insect flight muscle cell': 'CL:0000196',\n",
       " 'sensory receptor cell': 'CL:0000197',\n",
       " 'pain receptor cell': 'CL:0000198',\n",
       " 'mechanoreceptor cell': 'CL:0000199',\n",
       " 'touch receptor cell': 'CL:0000200',\n",
       " 'auditory hair cell': 'CL:0000202',\n",
       " 'gravity sensitive cell': 'CL:0000203',\n",
       " 'acceleration receptive cell': 'CL:0000204',\n",
       " 'thermoreceptor cell': 'CL:0000205',\n",
       " 'chemoreceptor cell': 'CL:0000206',\n",
       " 'olfactory receptor cell': 'CL:0000207',\n",
       " 'pH receptor cell': 'CL:0000208',\n",
       " 'taste receptor cell': 'CL:0000209',\n",
       " 'photoreceptor cell': 'CL:0000210',\n",
       " 'electrically active cell': 'CL:0000211',\n",
       " 'absorptive cell': 'CL:0000212',\n",
       " 'lining cell': 'CL:0000213',\n",
       " 'synovial cell': 'CL:0000214',\n",
       " 'barrier cell': 'CL:0000215',\n",
       " 'Sertoli cell': 'CL:0000216',\n",
       " 'insulating cell': 'CL:0000217',\n",
       " 'myelinating Schwann cell': 'CL:0000218',\n",
       " 'motile cell': 'CL:0000219',\n",
       " 'ectodermal cell': 'CL:0000221',\n",
       " 'mesodermal cell': 'CL:0000222',\n",
       " 'endodermal cell': 'CL:0000223',\n",
       " 'anucleate cell': 'CL:0000225',\n",
       " 'single nucleate cell': 'CL:0000226',\n",
       " 'binucleate cell': 'CL:0000227',\n",
       " 'multinucleate cell': 'CL:0000228',\n",
       " 'erythrocyte': 'CL:0000232',\n",
       " 'platelet': 'CL:0000233',\n",
       " 'phagocyte': 'CL:0000234',\n",
       " 'macrophage': 'CL:0000235',\n",
       " 'B cell': 'CL:0000236',\n",
       " 'keratinizing barrier epithelial cell': 'CL:0000237',\n",
       " 'non keratinizing barrier epithelial cell': 'CL:0000238',\n",
       " 'brush border epithelial cell': 'CL:0000239',\n",
       " 'stratified squamous epithelial cell': 'CL:0000240',\n",
       " 'stratified cuboidal epithelial cell': 'CL:0000241',\n",
       " 'Merkel cell': 'CL:0000242',\n",
       " 'transitional epithelial cell': 'CL:0000244',\n",
       " 'Mauthner neuron': 'CL:0000246',\n",
       " 'Rohon-Beard neuron': 'CL:0000247',\n",
       " 'hatching gland cell': 'CL:0000249',\n",
       " 'extramedullary cell': 'CL:0000251',\n",
       " 'eurydendroid cell': 'CL:0000253',\n",
       " 'eukaryotic cell': 'CL:0000255',\n",
       " 'uric acid accumulating cell': 'CL:0000256',\n",
       " 'Eumycetozoan cell': 'CL:0000257',\n",
       " 'hyphal cell': 'CL:0000286',\n",
       " 'eye photoreceptor cell': 'CL:0000287',\n",
       " 'structural cell': 'CL:0000293',\n",
       " 'somatotropin secreting cell': 'CL:0000295',\n",
       " 'gamete': 'CL:0000300',\n",
       " 'pole cell': 'CL:0000301',\n",
       " 'crystallin accumulating cell': 'CL:0000306',\n",
       " 'tracheal epithelial cell': 'CL:0000307',\n",
       " 'metal ion accumulating cell': 'CL:0000308',\n",
       " 'copper accumulating cell': 'CL:0000309',\n",
       " 'iron accumulating cell': 'CL:0000310',\n",
       " 'keratin accumulating cell': 'CL:0000311',\n",
       " 'keratinocyte': 'CL:0000312',\n",
       " 'serous secreting cell': 'CL:0000313',\n",
       " 'milk secreting cell': 'CL:0000314',\n",
       " 'tear secreting cell': 'CL:0000315',\n",
       " 'sebum secreting cell': 'CL:0000317',\n",
       " 'sweat secreting cell': 'CL:0000318',\n",
       " 'mucus secreting cell': 'CL:0000319',\n",
       " 'pulmonary alveolar epithelial cell': 'CL:0000322',\n",
       " 'lysozyme secreting cell': 'CL:0000323',\n",
       " 'metanephric mesenchyme stem cell': 'CL:0000324',\n",
       " 'stuff accumulating cell': 'CL:0000325',\n",
       " 'glycogen accumulating cell': 'CL:0000326',\n",
       " 'extracellular matrix secreting cell': 'CL:0000327',\n",
       " 'oxygen accumulating cell': 'CL:0000329',\n",
       " 'migratory neural crest cell': 'CL:0000333',\n",
       " 'vegetative cell (sensu Fungi)': 'CL:0000334',\n",
       " 'mesenchyme condensation cell': 'CL:0000335',\n",
       " 'adrenal medulla chromaffin cell': 'CL:0000336',\n",
       " 'neuroblast (sensu Nematoda and Protostomia)': 'CL:0000338',\n",
       " 'glioblast (sensu Vertebrata)': 'CL:0000339',\n",
       " 'glioblast (sensu Nematoda and Protostomia)': 'CL:0000340',\n",
       " 'pigment cell (sensu Nematoda and Protostomia)': 'CL:0000341',\n",
       " 'pigment cell (sensu Vertebrata)': 'CL:0000342',\n",
       " 'visual pigment cell (sensu Vertebrata)': 'CL:0000343',\n",
       " 'dental papilla cell': 'CL:0000345',\n",
       " 'hair follicle dermal papilla cell': 'CL:0000346',\n",
       " 'scleral cell': 'CL:0000347',\n",
       " 'choroidal cell of the eye': 'CL:0000348',\n",
       " 'extraembryonic cell': 'CL:0000349',\n",
       " 'amnioserosal cell': 'CL:0000350',\n",
       " 'trophoblast cell': 'CL:0000351',\n",
       " 'epiblast cell': 'CL:0000352',\n",
       " 'blastoderm cell': 'CL:0000353',\n",
       " 'blastemal cell': 'CL:0000354',\n",
       " 'multi-potent skeletal muscle stem cell': 'CL:0000355',\n",
       " 'hair matrix stem cell': 'CL:0000356',\n",
       " 'stratified epithelial stem cell': 'CL:0000357',\n",
       " 'sphincter associated smooth muscle cell': 'CL:0000358',\n",
       " 'vascular associated smooth muscle cell': 'CL:0000359',\n",
       " 'morula cell': 'CL:0000360',\n",
       " 'gastrula cell': 'CL:0000361',\n",
       " 'epidermal cell': 'CL:0000362',\n",
       " 'animal zygote': 'CL:0000365',\n",
       " 'sheath cell (sensu Nematoda)': 'CL:0000367',\n",
       " 'protoplast': 'CL:0000371',\n",
       " 'tormogen cell': 'CL:0000372',\n",
       " 'histoblast': 'CL:0000373',\n",
       " 'trichogen cell': 'CL:0000374',\n",
       " 'humidity receptor cell': 'CL:0000376',\n",
       " 'tracheoblast': 'CL:0000377',\n",
       " 'supporting cell (sensu Nematoda and Protostomia)': 'CL:0000378',\n",
       " 'sensory processing neuron': 'CL:0000379',\n",
       " 'thecogen cell': 'CL:0000380',\n",
       " 'scolopale cell': 'CL:0000382',\n",
       " 'nephrogenic mesenchyme stem cell': 'CL:0000383',\n",
       " 'ligament cell': 'CL:0000384',\n",
       " 'prohemocyte (sensu Nematoda and Protostomia)': 'CL:0000385',\n",
       " 'attachment cell': 'CL:0000386',\n",
       " 'hemocyte (sensu Arthropoda)': 'CL:0000387',\n",
       " 'tendon cell': 'CL:0000388',\n",
       " 'socket cell (sensu Nematoda)': 'CL:0000389',\n",
       " 'blood cell (sensu Nematoda and Protostomia)': 'CL:0000390',\n",
       " 'crystal cell': 'CL:0000392',\n",
       " 'electrically responsive cell': 'CL:0000393',\n",
       " 'plasmatocyte': 'CL:0000394',\n",
       " 'procrystal cell': 'CL:0000395',\n",
       " 'lamellocyte': 'CL:0000396',\n",
       " 'ganglion interneuron': 'CL:0000397',\n",
       " 'CNS interneuron': 'CL:0000402',\n",
       " 'electrically signaling cell': 'CL:0000404',\n",
       " 'CNS short range interneuron': 'CL:0000406',\n",
       " 'scolopidial ligament cell': 'CL:0000407',\n",
       " 'male gamete': 'CL:0000408',\n",
       " 'CNS long range interneuron': 'CL:0000410',\n",
       " 'Caenorhabditis hypodermal cell': 'CL:0000411',\n",
       " 'polyploid cell': 'CL:0000412',\n",
       " 'haploid cell': 'CL:0000413',\n",
       " 'diploid cell': 'CL:0000415',\n",
       " 'polytene cell': 'CL:0000416',\n",
       " 'endopolyploid cell': 'CL:0000417',\n",
       " 'arcade cell': 'CL:0000418',\n",
       " 'seam cell': 'CL:0000419',\n",
       " 'syncytial epithelial cell': 'CL:0000420',\n",
       " 'coelomocyte': 'CL:0000421',\n",
       " 'mitogenic signaling cell': 'CL:0000422',\n",
       " 'tip cell': 'CL:0000423',\n",
       " 'excretory cell': 'CL:0000424',\n",
       " 'pore cell': 'CL:0000425',\n",
       " 'GLR cell': 'CL:0000427',\n",
       " 'yolk cell': 'CL:0000428',\n",
       " 'imaginal disc cell': 'CL:0000429',\n",
       " 'xanthophore cell': 'CL:0000430',\n",
       " 'iridophore': 'CL:0000431',\n",
       " 'reticular cell': 'CL:0000432',\n",
       " 'eccrine cell': 'CL:0000434',\n",
       " 'alkali secreting cell': 'CL:0000435',\n",
       " 'vaginal lubricant secreting cell': 'CL:0000436',\n",
       " 'gonadtroph': 'CL:0000437',\n",
       " 'luteinizing hormone secreting cell': 'CL:0000438',\n",
       " 'prolactin secreting cell': 'CL:0000439',\n",
       " 'melanocyte stimulating hormone secreting cell': 'CL:0000440',\n",
       " 'follicle stem cell (sensu Arthropoda)': 'CL:0000441',\n",
       " 'follicular dendritic cell': 'CL:0000442',\n",
       " 'calcitonin secreting cell': 'CL:0000443',\n",
       " 'obliquely striated muscle cell': 'CL:0000444',\n",
       " 'apoptosis fated cell': 'CL:0000445',\n",
       " 'chief cell of parathyroid gland': 'CL:0000446',\n",
       " 'carbohydrate secreting cell': 'CL:0000447',\n",
       " 'white adipocyte': 'CL:0000448',\n",
       " 'brown adipocyte': 'CL:0000449',\n",
       " 'dendritic cell': 'CL:0000451',\n",
       " 'thyroid hormone secreting cell': 'CL:0000452',\n",
       " 'Langerhans cell': 'CL:0000453',\n",
       " 'epinephrine secreting cell': 'CL:0000454',\n",
       " 'mineralocorticoid secreting cell': 'CL:0000456',\n",
       " 'biogenic amine secreting cell': 'CL:0000457',\n",
       " 'serotonin secreting cell': 'CL:0000458',\n",
       " 'noradrenergic cell': 'CL:0000459',\n",
       " 'glucocorticoid secreting cell': 'CL:0000460',\n",
       " 'adepithelial cell': 'CL:0000462',\n",
       " 'epidermal cell (sensu Arthropoda)': 'CL:0000463',\n",
       " 'epidermoblast': 'CL:0000464',\n",
       " 'cardioblast (sensu Arthropoda)': 'CL:0000465',\n",
       " 'adrenocorticotropic hormone secreting cell': 'CL:0000467',\n",
       " 'neuroglioblast (sensu Nematoda and Protostomia)': 'CL:0000468',\n",
       " 'ganglion mother cell': 'CL:0000469',\n",
       " 'digestive enzyme secreting cell': 'CL:0000470',\n",
       " 'defensive cell': 'CL:0000473',\n",
       " 'pericardial nephrocyte': 'CL:0000474',\n",
       " 'thyrotroph': 'CL:0000476',\n",
       " 'follicle cell of egg chamber': 'CL:0000477',\n",
       " 'oxytocin stimulating hormone secreting cell': 'CL:0000478',\n",
       " 'vasopressin stimulating hormone secreting cell': 'CL:0000479',\n",
       " 'secretin stimulating hormone secreting cell': 'CL:0000480',\n",
       " 'cholecystokin stimulating hormone secreting cell': 'CL:0000481',\n",
       " 'juvenile hormone secreting cell': 'CL:0000482',\n",
       " 'bombesin stimulating hormone secreting cell': 'CL:0000483',\n",
       " 'connective tissue type mast cell': 'CL:0000484',\n",
       " 'mucosal type mast cell': 'CL:0000485',\n",
       " 'garland cell': 'CL:0000486',\n",
       " 'oenocyte': 'CL:0000487',\n",
       " 'visible light photoreceptor cell': 'CL:0000488',\n",
       " 'scotopic photoreceptor cell': 'CL:0000489',\n",
       " 'photopic photoreceptor cell': 'CL:0000490',\n",
       " 'CD4-positive helper T cell': 'CL:0000492',\n",
       " 'UV sensitive photoreceptor cell': 'CL:0000494',\n",
       " 'blue sensitive photoreceptor cell': 'CL:0000495',\n",
       " 'green sensitive photoreceptor cell': 'CL:0000496',\n",
       " 'red sensitive photoreceptor cell': 'CL:0000497',\n",
       " 'inhibitory interneuron': 'CL:0000498',\n",
       " 'stromal cell': 'CL:0000499',\n",
       " 'follicular epithelial cell': 'CL:0000500',\n",
       " 'granulosa cell': 'CL:0000501',\n",
       " 'type D enteroendocrine cell': 'CL:0000502',\n",
       " 'theca cell': 'CL:0000503',\n",
       " 'enterochromaffin-like cell': 'CL:0000504',\n",
       " 'substance P secreting cell': 'CL:0000505',\n",
       " 'enkephalin secreting cell': 'CL:0000506',\n",
       " 'endorphin secreting cell': 'CL:0000507',\n",
       " 'type G enteroendocrine cell': 'CL:0000508',\n",
       " 'gastrin secreting cell': 'CL:0000509',\n",
       " 'paneth cell': 'CL:0000510',\n",
       " 'androgen binding protein secreting cell': 'CL:0000511',\n",
       " 'cardiac muscle myoblast': 'CL:0000513',\n",
       " 'smooth muscle myoblast': 'CL:0000514',\n",
       " 'skeletal muscle myoblast': 'CL:0000515',\n",
       " 'perineuronal satellite cell': 'CL:0000516',\n",
       " 'macrophage derived foam cell': 'CL:0000517',\n",
       " 'phagocyte (sensu Vertebrata)': 'CL:0000518',\n",
       " 'phagocyte (sensu Nematoda and Protostomia)': 'CL:0000519',\n",
       " 'prokaryotic cell': 'CL:0000520',\n",
       " 'fungal cell': 'CL:0000521',\n",
       " 'mononuclear cytotrophoblast cell': 'CL:0000523',\n",
       " 'spheroplast': 'CL:0000524',\n",
       " 'syncytiotrophoblast cell': 'CL:0000525',\n",
       " 'afferent neuron': 'CL:0000526',\n",
       " 'efferent neuron': 'CL:0000527',\n",
       " 'nitrergic neuron': 'CL:0000528',\n",
       " 'pigmented epithelial cell': 'CL:0000529',\n",
       " 'primary neuron (sensu Teleostei)': 'CL:0000530',\n",
       " 'primary sensory neuron (sensu Teleostei)': 'CL:0000531',\n",
       " 'CAP motoneuron': 'CL:0000532',\n",
       " 'primary motor neuron (sensu Teleostei)': 'CL:0000533',\n",
       " 'primary interneuron (sensu Teleostei)': 'CL:0000534',\n",
       " 'secondary neuron (sensu Teleostei)': 'CL:0000535',\n",
       " 'secondary motor neuron (sensu Teleostei)': 'CL:0000536',\n",
       " 'neuron': 'CL:0000540',\n",
       " 'melanoblast': 'CL:0000541',\n",
       " 'lymphocyte': 'CL:0000542',\n",
       " 'slowly adapting mechanoreceptor cell': 'CL:0000544',\n",
       " 'T-helper 1 cell': 'CL:0000545',\n",
       " 'T-helper 2 cell': 'CL:0000546',\n",
       " 'proerythroblast': 'CL:0000547',\n",
       " 'basophilic erythroblast': 'CL:0000549',\n",
       " 'polychromatophilic erythroblast': 'CL:0000550',\n",
       " 'unimodal nocireceptor': 'CL:0000551',\n",
       " 'orthochromatic erythroblast': 'CL:0000552',\n",
       " 'megakaryocyte progenitor cell': 'CL:0000553',\n",
       " 'gastrin stimulating hormone secreting cell': 'CL:0000554',\n",
       " 'neuronal brush cell': 'CL:0000555',\n",
       " 'megakaryocyte': 'CL:0000556',\n",
       " 'granulocyte monocyte progenitor cell': 'CL:0000557',\n",
       " 'reticulocyte': 'CL:0000558',\n",
       " 'promonocyte': 'CL:0000559',\n",
       " 'band form neutrophil': 'CL:0000560',\n",
       " 'amacrine cell': 'CL:0000561',\n",
       " 'nucleate erythrocyte': 'CL:0000562',\n",
       " 'endospore': 'CL:0000563',\n",
       " 'neutrophilic promyelocyte': 'CL:0000564',\n",
       " 'fat body cell': 'CL:0000565',\n",
       " 'angioblastic mesenchymal cell': 'CL:0000566',\n",
       " 'polymodal nocireceptor': 'CL:0000567',\n",
       " 'amine precursor uptake and decarboxylation cell': 'CL:0000568',\n",
       " 'cardiac mesenchymal cell': 'CL:0000569',\n",
       " 'parafollicular cell': 'CL:0000570',\n",
       " 'leucophore': 'CL:0000571',\n",
       " 'retinal cone cell': 'CL:0000573',\n",
       " 'erythrophore': 'CL:0000574',\n",
       " 'corneal epithelial cell': 'CL:0000575',\n",
       " 'monocyte': 'CL:0000576',\n",
       " 'type EC enteroendocrine cell': 'CL:0000577',\n",
       " 'experimentally modified cell in vitro': 'CL:0000578',\n",
       " 'border follicle cell': 'CL:0000579',\n",
       " 'neutrophilic myelocyte': 'CL:0000580',\n",
       " 'peritoneal macrophage': 'CL:0000581',\n",
       " 'neutrophilic metamyelocyte': 'CL:0000582',\n",
       " 'alveolar macrophage': 'CL:0000583',\n",
       " 'enterocyte': 'CL:0000584',\n",
       " 'germ cell': 'CL:0000586',\n",
       " 'cold sensing thermoreceptor cell': 'CL:0000587',\n",
       " 'odontoclast': 'CL:0000588',\n",
       " 'cochlear inner hair cell': 'CL:0000589',\n",
       " 'small luteal cell': 'CL:0000590',\n",
       " 'warmth sensing thermoreceptor cell': 'CL:0000591',\n",
       " 'large luteal cell': 'CL:0000592',\n",
       " 'androgen secreting cell': 'CL:0000593',\n",
       " 'skeletal muscle satellite cell': 'CL:0000594',\n",
       " 'enucleate erythrocyte': 'CL:0000595',\n",
       " 'sexual spore': 'CL:0000596',\n",
       " 'microconidium': 'CL:0000597',\n",
       " 'pyramidal neuron': 'CL:0000598',\n",
       " 'conidium': 'CL:0000599',\n",
       " 'heterokaryon': 'CL:0000600',\n",
       " 'cochlear outer hair cell': 'CL:0000601',\n",
       " 'pressoreceptor cell': 'CL:0000602',\n",
       " 'dikaryon': 'CL:0000603',\n",
       " 'retinal rod cell': 'CL:0000604',\n",
       " 'fungal asexual spore': 'CL:0000605',\n",
       " 'macroconidium': 'CL:0000606',\n",
       " 'ascospore': 'CL:0000607',\n",
       " 'zygospore': 'CL:0000608',\n",
       " 'vestibular hair cell': 'CL:0000609',\n",
       " 'eosinophil progenitor cell': 'CL:0000611',\n",
       " 'eosinophilic myelocyte': 'CL:0000612',\n",
       " 'basophil progenitor cell': 'CL:0000613',\n",
       " 'basophilic myelocyte': 'CL:0000614',\n",
       " 'basidiospore': 'CL:0000615',\n",
       " 'GABAergic neuron': 'CL:0000617',\n",
       " 'sheath cell': 'CL:0000618',\n",
       " 'supporting cell (sensu Nematoda)': 'CL:0000619',\n",
       " 'fusion competent myoblast': 'CL:0000621',\n",
       " 'acinar cell': 'CL:0000622',\n",
       " 'natural killer cell': 'CL:0000623',\n",
       " 'CD4-positive, alpha-beta T cell': 'CL:0000624',\n",
       " 'CD8-positive, alpha-beta T cell': 'CL:0000625',\n",
       " 'olfactory granule cell': 'CL:0000626',\n",
       " 'transporting cell': 'CL:0000627',\n",
       " 'photosynthetic cell': 'CL:0000628',\n",
       " 'storage cell': 'CL:0000629',\n",
       " 'supporting cell': 'CL:0000630',\n",
       " 'labyrinth supporting cell': 'CL:0000631',\n",
       " 'hepatic stellate cell': 'CL:0000632',\n",
       " 'Hensen cell': 'CL:0000633',\n",
       " 'Claudius cell': 'CL:0000634',\n",
       " \"Deiter's cell\": 'CL:0000635',\n",
       " 'Mueller cell': 'CL:0000636',\n",
       " 'chromophil cell of anterior pituitary gland': 'CL:0000637',\n",
       " 'acidophil cell of pars distalis of adenohypophysis': 'CL:0000638',\n",
       " 'basophil cell of pars distalis of adenohypophysis': 'CL:0000639',\n",
       " 'chromophobe cell': 'CL:0000641',\n",
       " 'folliculostellate cell': 'CL:0000642',\n",
       " 'Bergmann glial cell': 'CL:0000644',\n",
       " 'pituicyte': 'CL:0000645',\n",
       " 'basal cell': 'CL:0000646',\n",
       " 'multinucleated giant cell': 'CL:0000647',\n",
       " 'kidney granular cell': 'CL:0000648',\n",
       " 'prickle cell': 'CL:0000649',\n",
       " 'mesangial cell': 'CL:0000650',\n",
       " 'mucous neck cell': 'CL:0000651',\n",
       " 'pinealocyte': 'CL:0000652',\n",
       " 'podocyte': 'CL:0000653',\n",
       " 'primary oocyte': 'CL:0000654',\n",
       " 'secondary oocyte': 'CL:0000655',\n",
       " 'primary spermatocyte': 'CL:0000656',\n",
       " 'secondary spermatocyte': 'CL:0000657',\n",
       " 'cuticle secreting cell': 'CL:0000658',\n",
       " 'eggshell secreting cell': 'CL:0000659',\n",
       " 'glycocalyx secreting cell': 'CL:0000660',\n",
       " 'distal tip cell (sensu Nematoda)': 'CL:0000661',\n",
       " 'neuroglioblast (sensu Nematoda)': 'CL:0000662',\n",
       " 'fenestrated endothelial cell': 'CL:0000666',\n",
       " 'collagen secreting cell': 'CL:0000667',\n",
       " 'pericyte': 'CL:0000669',\n",
       " 'primordial germ cell': 'CL:0000670',\n",
       " 'centripetally migrating follicle cell': 'CL:0000671',\n",
       " 'Kenyon cell': 'CL:0000673',\n",
       " 'interfollicle cell': 'CL:0000674',\n",
       " 'female gamete': 'CL:0000675',\n",
       " 'cap cell': 'CL:0000676',\n",
       " 'gut absorptive cell': 'CL:0000677',\n",
       " 'commissural neuron': 'CL:0000678',\n",
       " 'glutamatergic neuron': 'CL:0000679',\n",
       " 'muscle precursor cell': 'CL:0000680',\n",
       " 'radial glial cell': 'CL:0000681',\n",
       " 'M cell of gut': 'CL:0000682',\n",
       " 'ependymoglial cell': 'CL:0000683',\n",
       " 'littoral cell of liver': 'CL:0000684',\n",
       " 'cerebrospinal fluid secreting cell': 'CL:0000686',\n",
       " 'R1 photoreceptor cell': 'CL:0000687',\n",
       " 'perijunctional fibroblast': 'CL:0000688',\n",
       " 'myoendocrine cell': 'CL:0000689',\n",
       " 'R2 photoreceptor cell': 'CL:0000690',\n",
       " 'stellate interneuron': 'CL:0000691',\n",
       " 'terminal Schwann cell': 'CL:0000692',\n",
       " 'neurogliaform cell': 'CL:0000693',\n",
       " 'R3 photoreceptor cell': 'CL:0000694',\n",
       " 'Cajal-Retzius cell': 'CL:0000695',\n",
       " 'PP cell': 'CL:0000696',\n",
       " 'R4 photoreceptor cell': 'CL:0000697',\n",
       " 'paraganglial type 1 cell': 'CL:0000699',\n",
       " 'dopaminergic neuron': 'CL:0000700',\n",
       " 'paraganglia type 2 cell': 'CL:0000701',\n",
       " 'R5 photoreceptor cell': 'CL:0000702',\n",
       " 'sustentacular cell': 'CL:0000703',\n",
       " 'endothelial tip cell': 'CL:0000704',\n",
       " 'R6 photoreceptor cell': 'CL:0000705',\n",
       " 'choroid plexus epithelial cell': 'CL:0000706',\n",
       " 'R7 photoreceptor cell': 'CL:0000707',\n",
       " 'leptomeningeal cell': 'CL:0000708',\n",
       " 'R8 photoreceptor cell': 'CL:0000709',\n",
       " 'neurecto-epithelial cell': 'CL:0000710',\n",
       " 'cumulus cell': 'CL:0000711',\n",
       " 'stratum granulosum cell': 'CL:0000712',\n",
       " 'corona radiata cell': 'CL:0000713',\n",
       " 'embryonic crystal cell': 'CL:0000715',\n",
       " 'lymph gland crystal cell': 'CL:0000716',\n",
       " 'compound eye cone cell': 'CL:0000718',\n",
       " 'posterior cone cell (sensu Endopterygota)': 'CL:0000719',\n",
       " 'anterior cone cell (sensu Endopterygota)': 'CL:0000720',\n",
       " 'equatorial cone cell (sensu Endopterygota)': 'CL:0000721',\n",
       " 'cystoblast': 'CL:0000722',\n",
       " 'somatic stem cell': 'CL:0000723',\n",
       " 'heterocyst': 'CL:0000724',\n",
       " 'nitrogen fixing cell': 'CL:0000725',\n",
       " 'chlamydospore': 'CL:0000726',\n",
       " 'leading edge cell': 'CL:0000730',\n",
       " 'urothelial cell': 'CL:0000731',\n",
       " 'amoeboid cell': 'CL:0000732',\n",
       " 'lymph gland plasmatocyte': 'CL:0000733',\n",
       " 'lymph gland hemocyte': 'CL:0000735',\n",
       " 'striated muscle cell': 'CL:0000737',\n",
       " 'leukocyte': 'CL:0000738',\n",
       " 'retinal ganglion cell': 'CL:0000740',\n",
       " 'spinal accessory motor neuron': 'CL:0000741',\n",
       " 'periarticular chondrocyte': 'CL:0000742',\n",
       " 'hypertrophic chondrocyte': 'CL:0000743',\n",
       " 'columnar chondrocyte': 'CL:0000744',\n",
       " 'retina horizontal cell': 'CL:0000745',\n",
       " 'cardiac muscle cell': 'CL:0000746',\n",
       " 'cyanophore': 'CL:0000747',\n",
       " 'retinal bipolar neuron': 'CL:0000748',\n",
       " 'ON-bipolar cell': 'CL:0000749',\n",
       " 'OFF-bipolar cell': 'CL:0000750',\n",
       " 'rod bipolar cell': 'CL:0000751',\n",
       " 'cone retinal bipolar cell': 'CL:0000752',\n",
       " 'type 1 cone bipolar cell (sensu Mus)': 'CL:0000753',\n",
       " 'type 2 cone bipolar cell (sensu Mus)': 'CL:0000754',\n",
       " 'type 3 cone bipolar cell (sensu Mus)': 'CL:0000755',\n",
       " 'type 4 cone bipolar cell (sensu Mus)': 'CL:0000756',\n",
       " 'type 5 cone bipolar cell (sensu Mus)': 'CL:0000757',\n",
       " 'type 6 cone bipolar cell (sensu Mus)': 'CL:0000758',\n",
       " 'type 7 cone bipolar cell (sensu Mus)': 'CL:0000759',\n",
       " 'type 8 cone bipolar cell (sensu Mus)': 'CL:0000760',\n",
       " 'type 9 cone bipolar cell (sensu Mus)': 'CL:0000761',\n",
       " 'nucleated thrombocyte': 'CL:0000762',\n",
       " 'myeloid cell': 'CL:0000763',\n",
       " 'erythroid lineage cell': 'CL:0000764',\n",
       " 'erythroblast': 'CL:0000765',\n",
       " 'myeloid leukocyte': 'CL:0000766',\n",
       " 'basophil': 'CL:0000767',\n",
       " 'immature basophil': 'CL:0000768',\n",
       " 'basophilic metamyelocyte': 'CL:0000769',\n",
       " 'band form basophil': 'CL:0000770',\n",
       " 'eosinophil': 'CL:0000771',\n",
       " 'immature eosinophil': 'CL:0000772',\n",
       " 'eosinophilic metamyelocyte': 'CL:0000773',\n",
       " 'band form eosinophil': 'CL:0000774',\n",
       " 'neutrophil': 'CL:0000775',\n",
       " 'immature neutrophil': 'CL:0000776',\n",
       " 'mesangial phagocyte': 'CL:0000777',\n",
       " 'mononuclear osteoclast': 'CL:0000778',\n",
       " 'multinuclear osteoclast': 'CL:0000779',\n",
       " 'multinuclear odontoclast': 'CL:0000780',\n",
       " 'mononuclear odontoclast': 'CL:0000781',\n",
       " 'myeloid dendritic cell': 'CL:0000782',\n",
       " 'multinucleated phagocyte': 'CL:0000783',\n",
       " 'plasmacytoid dendritic cell': 'CL:0000784',\n",
       " 'mature B cell': 'CL:0000785',\n",
       " 'plasma cell': 'CL:0000786',\n",
       " 'memory B cell': 'CL:0000787',\n",
       " 'naive B cell': 'CL:0000788',\n",
       " 'alpha-beta T cell': 'CL:0000789',\n",
       " 'immature alpha-beta T cell': 'CL:0000790',\n",
       " 'mature alpha-beta T cell': 'CL:0000791',\n",
       " 'CD4-positive, CD25-positive, alpha-beta regulatory T cell': 'CL:0000792',\n",
       " 'CD4-positive, alpha-beta intraepithelial T cell': 'CL:0000793',\n",
       " 'CD8-positive, alpha-beta cytotoxic T cell': 'CL:0000794',\n",
       " 'CD8-positive, alpha-beta regulatory T cell': 'CL:0000795',\n",
       " 'CD8-alpha-beta-positive, alpha-beta intraepithelial T cell': 'CL:0000796',\n",
       " 'alpha-beta intraepithelial T cell': 'CL:0000797',\n",
       " 'gamma-delta T cell': 'CL:0000798',\n",
       " 'immature gamma-delta T cell': 'CL:0000799',\n",
       " 'mature gamma-delta T cell': 'CL:0000800',\n",
       " 'gamma-delta intraepithelial T cell': 'CL:0000801',\n",
       " 'CD8-alpha alpha positive, gamma-delta intraepithelial T cell': 'CL:0000802',\n",
       " 'CD4-negative CD8-negative gamma-delta intraepithelial T cell': 'CL:0000803',\n",
       " 'immature single positive thymocyte': 'CL:0000805',\n",
       " 'DN2 thymocyte': 'CL:0000806',\n",
       " 'DN3 thymocyte': 'CL:0000807',\n",
       " 'DN4 thymocyte': 'CL:0000808',\n",
       " 'double-positive, alpha-beta thymocyte': 'CL:0000809',\n",
       " 'CD4-positive, alpha-beta thymocyte': 'CL:0000810',\n",
       " 'CD8-positive, alpha-beta thymocyte': 'CL:0000811',\n",
       " 'memory T cell': 'CL:0000813',\n",
       " 'mature NK T cell': 'CL:0000814',\n",
       " 'regulatory T cell': 'CL:0000815',\n",
       " 'immature B cell': 'CL:0000816',\n",
       " 'precursor B cell': 'CL:0000817',\n",
       " 'transitional stage B cell': 'CL:0000818',\n",
       " 'B-1 B cell': 'CL:0000819',\n",
       " 'B-1a B cell': 'CL:0000820',\n",
       " 'B-1b B cell': 'CL:0000821',\n",
       " 'B-2 B cell': 'CL:0000822',\n",
       " 'immature natural killer cell': 'CL:0000823',\n",
       " 'mature natural killer cell': 'CL:0000824',\n",
       " 'pro-NK cell': 'CL:0000825',\n",
       " 'pro-B cell': 'CL:0000826',\n",
       " 'pro-T cell': 'CL:0000827',\n",
       " 'thromboblast': 'CL:0000828',\n",
       " 'basophilic myeloblast': 'CL:0000829',\n",
       " 'basophilic promyelocyte': 'CL:0000830',\n",
       " 'mast cell progenitor': 'CL:0000831',\n",
       " 'eosinophilic myeloblast': 'CL:0000832',\n",
       " 'eosinophilic promyelocyte': 'CL:0000833',\n",
       " 'neutrophil progenitor cell': 'CL:0000834',\n",
       " 'myeloblast': 'CL:0000835',\n",
       " 'promyelocyte': 'CL:0000836',\n",
       " 'hematopoietic multipotent progenitor cell': 'CL:0000837',\n",
       " 'lymphoid lineage restricted progenitor cell': 'CL:0000838',\n",
       " 'myeloid lineage restricted progenitor cell': 'CL:0000839',\n",
       " 'immature conventional dendritic cell': 'CL:0000840',\n",
       " 'mature conventional dendritic cell': 'CL:0000841',\n",
       " 'mononuclear cell': 'CL:0000842',\n",
       " 'follicular B cell': 'CL:0000843',\n",
       " 'germinal center B cell': 'CL:0000844',\n",
       " 'marginal zone B cell of spleen': 'CL:0000845',\n",
       " 'vestibular dark cell': 'CL:0000846',\n",
       " 'ciliated olfactory receptor neuron': 'CL:0000847',\n",
       " 'microvillous olfactory receptor neuron': 'CL:0000848',\n",
       " 'crypt olfactory receptor neuron': 'CL:0000849',\n",
       " 'serotonergic neuron': 'CL:0000850',\n",
       " 'neuromast mantle cell': 'CL:0000851',\n",
       " 'neuromast supporting cell': 'CL:0000852',\n",
       " 'olfactory epithelial supporting cell': 'CL:0000853',\n",
       " 'interneuromast cell': 'CL:0000854',\n",
       " 'sensory hair cell': 'CL:0000855',\n",
       " 'neuromast hair cell': 'CL:0000856',\n",
       " 'slow muscle myoblast': 'CL:0000857',\n",
       " 'fast muscle myoblast': 'CL:0000858',\n",
       " 'classical monocyte': 'CL:0000860',\n",
       " 'elicited macrophage': 'CL:0000861',\n",
       " 'suppressor macrophage': 'CL:0000862',\n",
       " 'inflammatory macrophage': 'CL:0000863',\n",
       " 'tissue-resident macrophage': 'CL:0000864',\n",
       " 'gastrointestinal tract (lamina propria) macrophage': 'CL:0000865',\n",
       " 'thymic macrophage': 'CL:0000866',\n",
       " 'secondary lymphoid organ macrophage': 'CL:0000867',\n",
       " 'lymph node macrophage': 'CL:0000868',\n",
       " 'tonsillar macrophage': 'CL:0000869',\n",
       " \"Peyer's patch macrophage\": 'CL:0000870',\n",
       " 'splenic macrophage': 'CL:0000871',\n",
       " 'splenic marginal zone macrophage': 'CL:0000872',\n",
       " 'splenic metallophillic macrophage': 'CL:0000873',\n",
       " 'splenic red pulp macrophage': 'CL:0000874',\n",
       " 'non-classical monocyte': 'CL:0000875',\n",
       " 'splenic white pulp macrophage': 'CL:0000876',\n",
       " 'splenic tingible body macrophage': 'CL:0000877',\n",
       " 'central nervous system macrophage': 'CL:0000878',\n",
       " 'meningeal macrophage': 'CL:0000879',\n",
       " 'choroid plexus macrophage': 'CL:0000880',\n",
       " 'perivascular macrophage': 'CL:0000881',\n",
       " 'thymic medullary macrophage': 'CL:0000882',\n",
       " 'thymic cortical macrophage': 'CL:0000883',\n",
       " 'mucosa-associated lymphoid tissue macrophage': 'CL:0000884',\n",
       " 'gut-associated lymphoid tissue macrophage': 'CL:0000885',\n",
       " 'nasal and broncial associated lymphoid tissue macrophage': 'CL:0000886',\n",
       " 'lymph node subcapsular sinus macrophage': 'CL:0000887',\n",
       " 'lymph node tingible body macrophage': 'CL:0000888',\n",
       " 'myeloid suppressor cell': 'CL:0000889',\n",
       " 'alternatively activated macrophage': 'CL:0000890',\n",
       " 'foam cell': 'CL:0000891',\n",
       " 'smooth muscle cell derived foam cell': 'CL:0000892',\n",
       " 'thymocyte': 'CL:0000893',\n",
       " 'DN1 thymic pro-T cell': 'CL:0000894',\n",
       " 'naive thymus-derived CD4-positive, alpha-beta T cell': 'CL:0000895',\n",
       " 'activated CD4-positive, alpha-beta T cell': 'CL:0000896',\n",
       " 'CD4-positive, alpha-beta memory T cell': 'CL:0000897',\n",
       " 'naive T cell': 'CL:0000898',\n",
       " 'T-helper 17 cell': 'CL:0000899',\n",
       " 'naive thymus-derived CD8-positive, alpha-beta T cell': 'CL:0000900',\n",
       " 'Tr1 cell': 'CL:0000901',\n",
       " 'induced T-regulatory cell': 'CL:0000902',\n",
       " 'natural T-regulatory cell': 'CL:0000903',\n",
       " 'central memory CD4-positive, alpha-beta T cell': 'CL:0000904',\n",
       " 'effector memory CD4-positive, alpha-beta T cell': 'CL:0000905',\n",
       " 'activated CD8-positive, alpha-beta T cell': 'CL:0000906',\n",
       " 'central memory CD8-positive, alpha-beta T cell': 'CL:0000907',\n",
       " 'CD8-positive, alpha-beta cytokine secreting effector T cell': 'CL:0000908',\n",
       " 'CD8-positive, alpha-beta memory T cell': 'CL:0000909',\n",
       " 'cytotoxic T cell': 'CL:0000910',\n",
       " 'effector T cell': 'CL:0000911',\n",
       " 'helper T cell': 'CL:0000912',\n",
       " 'effector memory CD8-positive, alpha-beta T cell': 'CL:0000913',\n",
       " 'immature NK T cell': 'CL:0000914',\n",
       " 'CD8-alpha-alpha-positive, alpha-beta intraepithelial T cell': 'CL:0000915',\n",
       " 'dendritic epidermal T cell': 'CL:0000916',\n",
       " 'Tc1 cell': 'CL:0000917',\n",
       " 'Tc2 cell': 'CL:0000918',\n",
       " 'CD8-positive, CD25-positive, alpha-beta regulatory T cell': 'CL:0000919',\n",
       " 'CD8-positive, CD28-negative, alpha-beta regulatory T cell': 'CL:0000920',\n",
       " 'type I NK T cell': 'CL:0000921',\n",
       " 'type II NK T cell': 'CL:0000922',\n",
       " 'CD4-positive type I NK T cell': 'CL:0000923',\n",
       " 'CD4-negative, CD8-negative type I NK T cell': 'CL:0000924',\n",
       " 'activated CD4-positive type I NK T cell': 'CL:0000925',\n",
       " 'CD4-positive type I NK T cell secreting interferon-gamma': 'CL:0000926',\n",
       " 'CD4-positive type I NK T cell secreting interleukin-4': 'CL:0000927',\n",
       " 'activated CD4-negative, CD8-negative type I NK T cell': 'CL:0000928',\n",
       " 'CD4-negative, CD8-negative type I NK T cell secreting interferon-gamma': 'CL:0000929',\n",
       " 'CD4-negative, CD8-negative type I NK T cell secreting interleukin-4': 'CL:0000930',\n",
       " 'activated type II NK T cell': 'CL:0000931',\n",
       " 'type II NK T cell secreting interferon-gamma': 'CL:0000932',\n",
       " 'type II NK T cell secreting interleukin-4': 'CL:0000933',\n",
       " 'CD4-positive, alpha-beta cytotoxic T cell': 'CL:0000934',\n",
       " 'CD4-negative, CD8-negative, alpha-beta intraepithelial T cell': 'CL:0000935',\n",
       " 'early lymphoid progenitor': 'CL:0000936',\n",
       " 'pre-natural killer cell': 'CL:0000937',\n",
       " 'CD16-negative, CD56-bright natural killer cell, human': 'CL:0000938',\n",
       " 'CD16-positive, CD56-dim natural killer cell, human': 'CL:0000939',\n",
       " 'mucosal invariant T cell': 'CL:0000940',\n",
       " 'thymic conventional dendritic cell': 'CL:0000941',\n",
       " 'thymic plasmacytoid dendritic cell': 'CL:0000942',\n",
       " 'Be1 Cell': 'CL:0000943',\n",
       " 'Be2 cell': 'CL:0000944',\n",
       " 'lymphocyte of B lineage': 'CL:0000945',\n",
       " 'antibody secreting cell': 'CL:0000946',\n",
       " 'IgE plasma cell': 'CL:0000947',\n",
       " 'IgE memory B cell': 'CL:0000948',\n",
       " 'IgD plasmablast': 'CL:0000949',\n",
       " 'IgE plasmablast': 'CL:0000950',\n",
       " 'IgE short lived plasma cell': 'CL:0000951',\n",
       " 'preBCR-positive large pre-B-II cell': 'CL:0000952',\n",
       " 'preBCR-negative large pre-B-II cell': 'CL:0000953',\n",
       " 'small pre-B-II cell': 'CL:0000954',\n",
       " 'pre-B-II cell': 'CL:0000955',\n",
       " 'pre-B-I cell': 'CL:0000956',\n",
       " 'large pre-B-II cell': 'CL:0000957',\n",
       " 'T1 B cell': 'CL:0000958',\n",
       " 'T2 B cell': 'CL:0000959',\n",
       " 'T3 B cell': 'CL:0000960',\n",
       " 'Bm1 B cell': 'CL:0000961',\n",
       " 'Bm2 B cell': 'CL:0000962',\n",
       " 'Bm3-delta B cell': 'CL:0000963',\n",
       " \"Bm2' B cell\": 'CL:0000964',\n",
       " 'Bm3 B cell': 'CL:0000965',\n",
       " 'Bm4 B cell': 'CL:0000966',\n",
       " 'Bm5 B cell': 'CL:0000967',\n",
       " 'Be cell': 'CL:0000968',\n",
       " 'regulatory B cell': 'CL:0000969',\n",
       " 'unswitched memory B cell': 'CL:0000970',\n",
       " 'IgM memory B cell': 'CL:0000971',\n",
       " 'class switched memory B cell': 'CL:0000972',\n",
       " 'IgA memory B cell': 'CL:0000973',\n",
       " 'long lived plasma cell': 'CL:0000974',\n",
       " 'short lived plasma cell': 'CL:0000975',\n",
       " 'IgA short lived plasma cell': 'CL:0000976',\n",
       " 'IgG short lived plasma cell': 'CL:0000977',\n",
       " 'IgM short lived plasma cell': 'CL:0000978',\n",
       " 'IgG memory B cell': 'CL:0000979',\n",
       " 'plasmablast': 'CL:0000980',\n",
       " 'double negative memory B cell': 'CL:0000981',\n",
       " 'IgG plasmablast': 'CL:0000982',\n",
       " 'IgM plasmablast': 'CL:0000983',\n",
       " 'IgA plasmablast': 'CL:0000984',\n",
       " 'IgG plasma cell': 'CL:0000985',\n",
       " 'IgM plasma cell': 'CL:0000986',\n",
       " 'IgA plasma cell': 'CL:0000987',\n",
       " 'hematopoietic cell': 'CL:0000988',\n",
       " 'CD11c-low plasmacytoid dendritic cell': 'CL:0000989',\n",
       " 'conventional dendritic cell': 'CL:0000990',\n",
       " 'CD11c-negative plasmacytoid dendritic cell': 'CL:0000991',\n",
       " 'immature CD11c-low plasmacytoid dendritic cell': 'CL:0000992',\n",
       " 'mature CD11c-low plasmacytoid dendritic cell': 'CL:0000993',\n",
       " 'immature CD11c-negative plasmacytoid dendritic cell': 'CL:0000994',\n",
       " 'CD34-positive, CD38-positive common myeloid progenitor OR CD34-positive, CD38-positive common lymphoid progenitor': 'CL:0000995',\n",
       " 'mature CD11c-negative plasmacytoid dendritic cell': 'CL:0000996',\n",
       " 'immature CD8_alpha-negative CD11b-positive dendritic cell': 'CL:0000997',\n",
       " 'CD8_alpha-negative CD11b-negative dendritic cell': 'CL:0000998',\n",
       " 'CD4-positive CD11b-positive dendritic cell': 'CL:0000999',\n",
       " 'CD8_alpha-positive CD11b-negative dendritic cell': 'CL:0001000',\n",
       " 'immature CD8_alpha-negative CD11b-negative dendritic cell': 'CL:0001001',\n",
       " 'mature CD8_alpha-negative CD11b-negative dendritic cell': 'CL:0001002',\n",
       " 'mature CD8_alpha-negative CD11b-positive dendritic cell': 'CL:0001003',\n",
       " 'immature CD8_alpha-positive CD11b-negative dendritic cell': 'CL:0001004',\n",
       " 'mature CD8_alpha-positive CD11b-negative dendritic cell': 'CL:0001005',\n",
       " 'dermal dendritic cell': 'CL:0001006',\n",
       " 'interstitial dendritic cell': 'CL:0001007',\n",
       " 'Kit and Sca1-positive hematopoietic stem cell': 'CL:0001008',\n",
       " 'immature dermal dendritic cell': 'CL:0001009',\n",
       " 'mature dermal dendritic cell': 'CL:0001010',\n",
       " 'immature interstitial dendritic cell': 'CL:0001011',\n",
       " 'CD7-negative lymphoid progenitor OR granulocyte monocyte progenitor': 'CL:0001012',\n",
       " 'mature interstitial dendritic cell': 'CL:0001013',\n",
       " 'CD1a-positive Langerhans cell': 'CL:0001014',\n",
       " 'CD8_alpha-low Langerhans cell': 'CL:0001015',\n",
       " 'immature CD1a-positive Langerhans cell': 'CL:0001016',\n",
       " 'mature CD1a-positive Langerhans cell': 'CL:0001017',\n",
       " 'immature CD8_alpha-low Langerhans cell': 'CL:0001018',\n",
       " 'CD115-positive monocyte OR common dendritic progenitor': 'CL:0001019',\n",
       " 'mature CD8_alpha-low Langerhans cell': 'CL:0001020',\n",
       " 'CD34-positive, CD38-positive common lymphoid progenitor': 'CL:0001021',\n",
       " 'CD115-positive monocyte': 'CL:0001022',\n",
       " 'Kit-positive, CD34-positive common myeloid progenitor': 'CL:0001023',\n",
       " 'CD34-positive, CD38-negative hematopoietic stem cell': 'CL:0001024',\n",
       " 'Kit-positive, Sca1-positive common lymphoid progenitor': 'CL:0001025',\n",
       " 'CD34-positive, CD38-positive common myeloid progenitor': 'CL:0001026',\n",
       " 'CD7-negative lymphoid progenitor cell': 'CL:0001027',\n",
       " 'CD7-positive lymphoid progenitor cell': 'CL:0001028',\n",
       " 'common dendritic progenitor': 'CL:0001029',\n",
       " 'CD117-positive common myeloid progenitor OR CD217-positive common lymphoid progenitor': 'CL:0001030',\n",
       " 'cerebellar granule cell': 'CL:0001031',\n",
       " 'cortical granule cell': 'CL:0001032',\n",
       " 'hippocampal granule cell': 'CL:0001033',\n",
       " 'cell in vitro': 'CL:0001034',\n",
       " 'bone cell': 'CL:0001035',\n",
       " 'terminally differentiated osteoblast': 'CL:0001039',\n",
       " 'non-terminally differentiated osteoblast': 'CL:0001040',\n",
       " 'CD8-positive, CXCR3-positive, alpha-beta regulatory T cell': 'CL:0001041',\n",
       " 'T-helper 22 cell': 'CL:0001042',\n",
       " 'activated CD4-positive, alpha-beta T cell, human': 'CL:0001043',\n",
       " 'effector CD4-positive, alpha-beta T cell': 'CL:0001044',\n",
       " 'naive CCR4-positive regulatory T cell': 'CL:0001045',\n",
       " 'memory CCR4-positive regulatory T cell': 'CL:0001046',\n",
       " 'CD4-positive, CD25-positive, CCR4-positive, alpha-beta regulatory T cell': 'CL:0001047',\n",
       " 'activated CD4-positive, CD25-positive, CCR4-positive, alpha-beta regulatory T cell, human': 'CL:0001048',\n",
       " 'activated CD8-positive, alpha-beta T cell, human': 'CL:0001049',\n",
       " 'effector CD8-positive, alpha-beta T cell': 'CL:0001050',\n",
       " 'CD4-positive, CXCR3-negative, CCR6-negative, alpha-beta T cell': 'CL:0001051',\n",
       " 'CD8-positive, CXCR3-negative, CCR6-negative, alpha-beta T cell': 'CL:0001052',\n",
       " 'IgD-negative memory B cell': 'CL:0001053',\n",
       " 'CD14-positive monocyte': 'CL:0001054',\n",
       " 'CD14-positive, CD16-low monocyte': 'CL:0001055',\n",
       " 'dendritic cell, human': 'CL:0001056',\n",
       " 'myeloid dendritic cell, human': 'CL:0001057',\n",
       " 'plasmacytoid dendritic cell, human': 'CL:0001058',\n",
       " 'common myeloid progenitor, CD34-positive': 'CL:0001059',\n",
       " 'hematopoietic oligopotent progenitor cell, lineage-negative': 'CL:0001060',\n",
       " 'abnormal cell': 'CL:0001061',\n",
       " 'effector memory CD8-positive, alpha-beta T cell, terminally differentiated': 'CL:0001062',\n",
       " 'neoplastic cell': 'CL:0001063',\n",
       " 'malignant cell': 'CL:0001064',\n",
       " 'innate lymphoid cell': 'CL:0001065',\n",
       " 'erythroid progenitor cell, mammalian': 'CL:0001066',\n",
       " 'group 1 innate lymphoid cell': 'CL:0001067',\n",
       " 'ILC1': 'CL:0001068',\n",
       " 'group 2 innate lymphoid cell': 'CL:0001069',\n",
       " 'beige adipocyte': 'CL:0001070',\n",
       " 'group 3 innate lymphoid cell': 'CL:0001071',\n",
       " 'CD34-negative, CD117-positive innate lymphoid cell, human': 'CL:0001072',\n",
       " 'CD34-negative, CD56-positive, CD117-positive innate lymphoid cell, human': 'CL:0001073',\n",
       " 'CD34-positive, CD56-positive, CD117-positive common innate lymphoid precursor, human': 'CL:0001074',\n",
       " 'KLRG1-positive innate lymphoid cell, human': 'CL:0001075',\n",
       " 'NKp46-positive innate lymphoid cell, human': 'CL:0001076',\n",
       " 'ILC1, human': 'CL:0001077',\n",
       " 'group 3 innate lymphoid cell, human': 'CL:0001078',\n",
       " 'NKp44-positive group 3 innate lymphoid cell, human': 'CL:0001079',\n",
       " 'NKp44-negative group 3 innate lymphoid cell, human': 'CL:0001080',\n",
       " 'group 2 innate lymphoid cell, human': 'CL:0001081',\n",
       " 'immature innate lymphoid cell': 'CL:0001082',\n",
       " 'effector memory CD4-positive, alpha-beta T cell, terminally differentiated': 'CL:0001087',\n",
       " 'lymphocyte of B lineage, CD19-positive': 'CL:0001200',\n",
       " 'B cell, CD19-positive': 'CL:0001201',\n",
       " 'CD86-positive plasmablast': 'CL:0001202',\n",
       " 'CD8-positive, alpha-beta memory T cell, CD45RO-positive': 'CL:0001203',\n",
       " 'CD4-positive, alpha-beta memory T cell, CD45RO-positive': 'CL:0001204',\n",
       " 'visual pigment cell (sensu Nematoda and Protostomia)': 'CL:0001658',\n",
       " 'Kit-positive erythroid progenitor cell': 'CL:0002000',\n",
       " 'CD34-positive, CD38-positive granulocyte monocyte progenitor': 'CL:0002001',\n",
       " 'Kit-positive granulocyte monocyte progenitor': 'CL:0002002',\n",
       " 'CD34-positive, GlyA-negative erythroid progenitor cell': 'CL:0002003',\n",
       " 'CD34-negative, GlyA-negative proerythroblast': 'CL:0002004',\n",
       " 'CD34-positive, CD38-positive megakaryocyte erythroid progenitor cell': 'CL:0002005',\n",
       " 'Kit-positive, CD34-negative megakaryocyte erythroid progenitor cell': 'CL:0002006',\n",
       " 'Kit-low, CD34-positive eosinophil progenitor cell': 'CL:0002007',\n",
       " 'CD34-positive, CD38-positive eosinophil progenitor cell': 'CL:0002008',\n",
       " 'macrophage dendritic cell progenitor': 'CL:0002009',\n",
       " 'pre-conventional dendritic cell': 'CL:0002010',\n",
       " 'Kit-positive macrophage dendritic cell progenitor': 'CL:0002011',\n",
       " 'Kit-low proerythroblast': 'CL:0002012',\n",
       " 'GlyA-positive basophilic erythroblast': 'CL:0002013',\n",
       " 'Kit-negative, Ly-76 high basophilic erythroblast': 'CL:0002014',\n",
       " 'Kit-negative, Ly-76 high polychromatophilic erythroblast': 'CL:0002015',\n",
       " 'CD71-low, GlyA-positive polychromatic erythroblast': 'CL:0002016',\n",
       " 'Kit-negative, Ly-76 high orthochromatophilic erythroblasts': 'CL:0002017',\n",
       " 'CD71-negative, GlyA-positive orthochromatic erythroblast': 'CL:0002018',\n",
       " 'Ly-76 high reticulocyte': 'CL:0002019',\n",
       " 'GlyA-positive reticulocytes': 'CL:0002020',\n",
       " 'GlyA-positive erythrocyte': 'CL:0002021',\n",
       " 'Ly-76 high positive erythrocyte': 'CL:0002022',\n",
       " 'CD34-positive, CD41-positive, CD42-positive megakaryocyte progenitor cell': 'CL:0002023',\n",
       " 'Kit-positive megakaryocyte progenitor cell': 'CL:0002024',\n",
       " 'CD34-positive, CD41-positive, CD42-negative megakaryocyte progenitor cell': 'CL:0002025',\n",
       " 'CD34-negative, CD41-positive, CD42-positive megakaryocyte cell': 'CL:0002026',\n",
       " 'CD9-positive, CD41-positive megakaryocyte cell': 'CL:0002027',\n",
       " 'basophil mast progenitor cell': 'CL:0002028',\n",
       " 'Fc-epsilon RIalpha-low mast cell progenitor': 'CL:0002029',\n",
       " 'Fc-epsilon RIalpha-high basophil progenitor cell': 'CL:0002030',\n",
       " 'hematopoietic lineage restricted progenitor cell': 'CL:0002031',\n",
       " 'hematopoietic oligopotent progenitor cell': 'CL:0002032',\n",
       " 'short term hematopoietic stem cell': 'CL:0002033',\n",
       " 'long term hematopoietic stem cell': 'CL:0002034',\n",
       " 'Slamf1-negative multipotent progenitor cell': 'CL:0002035',\n",
       " 'Slamf1-positive multipotent progenitor cell': 'CL:0002036',\n",
       " 'CD2-positive, CD5-positive, CD44-positive alpha-beta intraepithelial T cell': 'CL:0002037',\n",
       " 'T follicular helper cell': 'CL:0002038',\n",
       " 'immature NK T cell stage I': 'CL:0002039',\n",
       " 'immature NK T cell stage II': 'CL:0002040',\n",
       " 'immature NK T cell stage III': 'CL:0002041',\n",
       " 'immature NK T cell stage IV': 'CL:0002042',\n",
       " 'CD34-positive, CD38-negative multipotent progenitor cell': 'CL:0002043',\n",
       " 'Kit-positive, integrin beta7-high basophil mast progenitor cell': 'CL:0002044',\n",
       " 'fraction A pre-pro B cell': 'CL:0002045',\n",
       " 'early pro-B cell': 'CL:0002046',\n",
       " 'fraction B precursor B cell': 'CL:0002047',\n",
       " 'late pro-B cell': 'CL:0002048',\n",
       " 'fraction C precursor B cell': 'CL:0002049',\n",
       " \"fraction C' precursor B cell\": 'CL:0002050',\n",
       " 'CD38-high pre-BCR positive cell': 'CL:0002051',\n",
       " 'fraction D precursor B cell': 'CL:0002052',\n",
       " 'CD22-positive, CD38-low small pre-B cell': 'CL:0002053',\n",
       " 'fraction E immature B cell': 'CL:0002054',\n",
       " 'CD38-negative immature B cell': 'CL:0002055',\n",
       " 'fraction F mature B cell': 'CL:0002056',\n",
       " 'CD14-positive, CD16-negative classical monocyte': 'CL:0002057',\n",
       " 'Gr1-low non-classical monocyte': 'CL:0002058',\n",
       " 'CD8alpha-positive thymic conventional dendritic cell': 'CL:0002059',\n",
       " 'melanophage': 'CL:0002060',\n",
       " 'T-helper 9 cell': 'CL:0002061',\n",
       " 'pulmonary alveolar type 1 cell': 'CL:0002062',\n",
       " 'pulmonary alveolar type 2 cell': 'CL:0002063',\n",
       " 'pancreatic acinar cell': 'CL:0002064',\n",
       " 'Feyrter cell': 'CL:0002066',\n",
       " 'type A enteroendocrine cell': 'CL:0002067',\n",
       " 'Purkinje myocyte': 'CL:0002068',\n",
       " ...}"
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "triplet_selector.name2id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "4e73d092-d44d-4951-aa35-9ec761f46e20",
   "metadata": {},
   "outputs": [],
   "source": [
    "from itertools import combinations\n",
    "import numpy as np\n",
    "import random\n",
    "import torch\n",
    "from typing import List, Union, Optional\n",
    "\n",
    "from scimilarity.ontologies import (\n",
    "    import_cell_ontology,\n",
    "    get_id_mapper,\n",
    "    get_all_ancestors,\n",
    "    get_all_descendants,\n",
    "    get_parents,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "00fdd966-3646-48da-a1ad-a46514a45479",
   "metadata": {},
   "outputs": [],
   "source": [
    "from datetime import datetime\n",
    "import json\n",
    "import os\n",
    "import pandas as pd\n",
    "import pytorch_lightning as pl\n",
    "import torch\n",
    "from torch import nn\n",
    "import torch.nn.functional as F\n",
    "from typing import Optional, List\n",
    "\n",
    "from scimilarity.triplet_selector import TripletSelector\n",
    "from scimilarity.nn_models import Encoder, Decoder\n",
    "\n",
    "class TripletLoss(torch.nn.TripletMarginLoss):\n",
    "    \"\"\"\n",
    "    Wrapper for pytorch TripletMarginLoss.\n",
    "    Triplets are generated using TripletSelector object which take embeddings and labels\n",
    "    then return triplets.\n",
    "    \"\"\"\n",
    "\n",
    "    def __init__(\n",
    "        self,\n",
    "        margin: float,\n",
    "        sample_across_studies: bool = True,\n",
    "        negative_selection: str = \"semihard\",\n",
    "        perturb_labels: bool = True,\n",
    "        perturb_labels_fraction: float = 0.5,\n",
    "    ):\n",
    "        \"\"\"Constructor.\n",
    "\n",
    "        Parameters\n",
    "        ----------\n",
    "        margin: float\n",
    "            Triplet loss margin.\n",
    "        sample_across_studies: bool, default: True\n",
    "            Whether to enforce anchor-positive pairs being from different studies.\n",
    "        negative_selection: str\n",
    "            Method for negative selection: {\"semihard\", \"hardest\", \"random\"}\n",
    "        perturb_labels: bool, default: True\n",
    "            Whether to perturb the ontology labels by coarse graining one level up.\n",
    "        perturb_labels_fraction: float, default: 0.5\n",
    "            The fraction of labels to perturb\n",
    "\n",
    "        Examples\n",
    "        --------\n",
    "        >>> triplet_loss = TripletLoss(margin=0.05)\n",
    "        \"\"\"\n",
    "\n",
    "        super().__init__()\n",
    "        self.margin = margin\n",
    "        self.sample_across_studies = sample_across_studies\n",
    "        self.triplet_selector = TripletSelector(\n",
    "            margin=margin,\n",
    "            negative_selection=negative_selection,\n",
    "            perturb_labels=perturb_labels,\n",
    "            perturb_labels_fraction=perturb_labels_fraction,\n",
    "        )\n",
    "\n",
    "    def forward(\n",
    "        self,\n",
    "        embeddings: torch.Tensor,\n",
    "        labels: torch.Tensor,\n",
    "        int2label: dict,\n",
    "        studies: torch.Tensor,\n",
    "    ):\n",
    "        if self.sample_across_studies is False:\n",
    "            studies = None\n",
    "\n",
    "        (\n",
    "            triplets,\n",
    "            num_violating_triplets,\n",
    "            num_viable_triplets,\n",
    "        ) = self.triplet_selector.get_triplets(embeddings, labels, int2label, studies)\n",
    "\n",
    "        anchor, positive, negative = triplets\n",
    "        return (\n",
    "            F.triplet_margin_loss(\n",
    "                anchor,\n",
    "                positive,\n",
    "                negative,\n",
    "                margin=self.margin,\n",
    "                p=self.p,\n",
    "                eps=self.eps,\n",
    "                swap=self.swap,\n",
    "                reduction=\"none\",\n",
    "            ),\n",
    "            torch.tensor(num_violating_triplets, dtype=torch.float),\n",
    "            torch.tensor(num_viable_triplets, dtype=torch.float),\n",
    "        )\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "7e29f3f6-5c0c-4e9b-bc27-309ae819c7c4",
   "metadata": {},
   "outputs": [],
   "source": [
    "triplet_loss = TripletLoss(margin=0.05)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "791041cc-f1f8-41ce-b8e0-b435b4ffa785",
   "metadata": {},
   "outputs": [],
   "source": [
    "import ssl\n",
    "ssl._create_default_https_context = ssl._create_unverified_context"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "d3ee674f-35ad-4e90-8bbb-3f499e7fc7bb",
   "metadata": {},
   "outputs": [],
   "source": [
    "class MetricLearning(pl.LightningModule):\n",
    "    \"\"\"A class encapsulating the metric learning.\"\"\"\n",
    "\n",
    "    def __init__(\n",
    "        self,\n",
    "        n_genes: int,\n",
    "        latent_dim: int = 128,\n",
    "        hidden_dim: List[int] = [1024, 1024],\n",
    "        dropout: float = 0.5,\n",
    "        input_dropout: float = 0.4,\n",
    "        triplet_loss_weight: float = 0.001,\n",
    "        margin: float = 0.05,\n",
    "        negative_selection: str = \"semihard\",\n",
    "        sample_across_studies: bool = True,\n",
    "        perturb_labels: bool = True,\n",
    "        perturb_labels_fraction: float = 0.5,\n",
    "        lr: float = 5e-3,\n",
    "        l1: float = 1e-4,\n",
    "        l2: float = 0.01,\n",
    "        max_epochs: int = 500,\n",
    "        residual: bool = False,\n",
    "    ):\n",
    "        \"\"\"Constructor.\n",
    "\n",
    "        Parameters (network)\n",
    "        ----------\n",
    "        n_genes: int\n",
    "            The number of genes in the gene space, representing the input dimensions.\n",
    "        latent_dim: int, default: 128\n",
    "            The latent space dimensions\n",
    "        hidden_dim: List[int], default: [1024, 1024]\n",
    "            A list of hidden layer dimensions, describing the number of layers and their dimensions.\n",
    "            Hidden layers are constructed in the order of the list for the encoder and in reverse\n",
    "            for the decoder.\n",
    "        dropout: float, default: 0.5\n",
    "            The dropout rate for hidden layers\n",
    "        input_dropout: float, default: 0.4\n",
    "            The dropout rate for the input layer\n",
    "\n",
    "        Parameters (training)\n",
    "        ----------\n",
    "        triplet_loss_weight: float, default 0.001\n",
    "            The weighting for triplet loss vs reconstruction loss.  This weighting sums to 1\n",
    "            such that triplet loss weight is triplet_loss_weight and reconstruction loss weight is (1 - triplet_loss_weight).\n",
    "        margin: float, default: 0.05\n",
    "            The margin parameter in triplet loss.\n",
    "        negative_selection: {\"semihard\", \"hardest\", \"random\"}, default: \"semihard\"\n",
    "            The negative selection function.\n",
    "        sample_across_studies: bool, default: True\n",
    "            Whether to enforce anchor-positive pairs being from different studies.\n",
    "        perturb_labels: bool, default: True\n",
    "            Whether to perturb celltype labels by coarse graining the label based on cell ontology.\n",
    "        perturb_labels_fraction: float, default: 0.5\n",
    "            The fraction of cells per batch to perform label perturbation.\n",
    "        lr: float, default: 5e-3\n",
    "            The initial learning rate\n",
    "        l1: float, default: 1e-4\n",
    "            The l1 penalty lambda. A value of 0 will disable l1 penalty.\n",
    "        l2: float, default: 1e-2\n",
    "            The l2 penalty lambda (weight decay). A value of 0 will disable l2 penalty.\n",
    "        max_epochs: int, default: 500\n",
    "            The max epochs, used by the scheduler to determine lr annealing rate.\n",
    "\n",
    "        Examples\n",
    "        --------\n",
    "        >>> datamodule = MetricLearningZarrDataModule(\n",
    "                batch_size=1000,\n",
    "                num_workers=1,\n",
    "                obs_field=\"celltype_name\",\n",
    "                train_path=\"train\",\n",
    "                gene_order=\"gene_order.tsv\",\n",
    "            )\n",
    "        >>> model = MetricLearning(datamodule.n_genes)\n",
    "        \"\"\"\n",
    "\n",
    "        super().__init__()\n",
    "        self.save_hyperparameters()\n",
    "        valid_negative_selection = {\"semihard\", \"hardest\", \"random\"}\n",
    "        if negative_selection not in valid_negative_selection:\n",
    "            raise ValueError(\n",
    "                f\"Unknown return_type {negative_selection}.\"\n",
    "                f\"Options are: {valid_negative_selection}.\"\n",
    "            )\n",
    "\n",
    "        # network architecture\n",
    "        self.n_genes = n_genes\n",
    "        self.latent_dim = latent_dim\n",
    "        self.hidden_dim = hidden_dim\n",
    "        self.dropout = dropout\n",
    "        self.input_dropout = input_dropout\n",
    "        self.residual = residual\n",
    "\n",
    "        # networks\n",
    "        self.encoder = Encoder(\n",
    "            self.n_genes,\n",
    "            latent_dim=self.latent_dim,\n",
    "            hidden_dim=self.hidden_dim,\n",
    "            dropout=self.dropout,\n",
    "            input_dropout=self.input_dropout,\n",
    "            residual=self.residual,\n",
    "        )\n",
    "        self.decoder = Decoder(\n",
    "            self.n_genes,\n",
    "            latent_dim=self.latent_dim,\n",
    "            hidden_dim=list(reversed(self.hidden_dim)),\n",
    "            dropout=self.dropout,\n",
    "            residual=self.residual,\n",
    "        )\n",
    "\n",
    "        # save layer sizes\n",
    "        model_state_dict = self.encoder.state_dict()\n",
    "        self.layer_sizes = {\n",
    "            entry: list(model_state_dict[entry].size()) for entry in model_state_dict\n",
    "        }\n",
    "\n",
    "        # mixed loss weight\n",
    "        self.triplet_loss_weight = triplet_loss_weight\n",
    "\n",
    "        # constraints\n",
    "        self.margin = margin\n",
    "        self.negative_selection = negative_selection\n",
    "        self.sample_across_studies = sample_across_studies\n",
    "        self.perturb_labels = perturb_labels\n",
    "        self.perturb_labels_fraction = perturb_labels_fraction\n",
    "\n",
    "        # lr and regularization\n",
    "        self.lr = lr\n",
    "        self.l1 = l1\n",
    "        self.l2 = l2\n",
    "        self.max_epochs = max_epochs\n",
    "\n",
    "        # losses\n",
    "        self.triplet_loss_fn = TripletLoss(\n",
    "            margin=self.margin,\n",
    "            negative_selection=self.negative_selection,\n",
    "            sample_across_studies=self.sample_across_studies,\n",
    "            perturb_labels=self.perturb_labels,\n",
    "            perturb_labels_fraction=self.perturb_labels_fraction,\n",
    "        )\n",
    "        self.mse_loss_fn = nn.MSELoss()\n",
    "\n",
    "        self.scheduler = None\n",
    "\n",
    "        self.val_step_outputs = []\n",
    "        self.test_step_outputs = []\n",
    "\n",
    "    def configure_optimizers(self):\n",
    "        \"\"\"Configure optimizers.\"\"\"\n",
    "\n",
    "        optimizer = torch.optim.AdamW(self.parameters(), self.lr, weight_decay=self.l2)\n",
    "        self.scheduler = {\n",
    "            \"scheduler\": torch.optim.lr_scheduler.CosineAnnealingLR(\n",
    "                optimizer, T_max=self.max_epochs\n",
    "            ),\n",
    "            \"interval\": \"epoch\",\n",
    "            \"frequency\": 1,\n",
    "        }\n",
    "        return {\n",
    "            \"optimizer\": optimizer,\n",
    "            \"lr_scheduler\": self.scheduler,\n",
    "        }  # pytorch-lightning required format\n",
    "\n",
    "\n",
    "    def forward(self, x):\n",
    "        \"\"\"Forward.\n",
    "\n",
    "        Parameters\n",
    "        ----------\n",
    "        x: torch.Tensor\n",
    "            Input tensor corresponding to input layer.\n",
    "\n",
    "        Returns\n",
    "        -------\n",
    "        z: torch.Tensor\n",
    "            Output tensor corresponding to the last encoder layer.\n",
    "        x_hat: torch.Tensor\n",
    "            Output tensor corresponding to the last decoder layer.\n",
    "        \"\"\"\n",
    "\n",
    "        z = self.encoder(x)\n",
    "        x_hat = self.decoder(z)\n",
    "        return z, x_hat\n",
    "\n",
    "\n",
    "    def get_losses(self, batch, use_studies: bool = True):\n",
    "        \"\"\"Calculate the triplet and reconstruction loss.\n",
    "\n",
    "        Parameters\n",
    "        ----------\n",
    "        batch:\n",
    "            A batch as defined by a pytorch DataLoader.\n",
    "        use_studies: bool, default: True\n",
    "            Whether to use studies metadata in mining triplets and calculating triplet loss\n",
    "\n",
    "        Returns\n",
    "        -------\n",
    "        triplet_loss: torch.Tensor\n",
    "            Triplet loss.\n",
    "        mse: torch.Tensor\n",
    "            MSE reconstruction loss\n",
    "        num_hard_triplets: torch.Tensor\n",
    "            Number of hard triplets.\n",
    "        num_viable_triplets: torch.Tensor\n",
    "            Number of viable triplets.\n",
    "        \"\"\"\n",
    "\n",
    "        cells, labels, studies = batch\n",
    "        if not use_studies:\n",
    "            studies = None\n",
    "        embedding, reconstruction = self(cells)\n",
    "        triplet_loss, num_hard_triplets, num_viable_triplets = self.triplet_loss_fn(\n",
    "            embedding, labels, int2label, studies\n",
    "        )\n",
    "        mse = self.mse_loss_fn(cells, reconstruction)\n",
    "        asw = self.triplet_loss_fn.triplet_selector.get_asw(\n",
    "            embedding, labels, int2label\n",
    "        )\n",
    "        nmse = mse / cells.pow(2).mean()\n",
    "        return (\n",
    "            triplet_loss,\n",
    "            mse,\n",
    "            num_hard_triplets,\n",
    "            num_viable_triplets,\n",
    "            asw,\n",
    "            nmse,\n",
    "        )\n",
    "\n",
    "    def get_mixed_loss(self, triplet_loss, mse):\n",
    "        \"\"\"Calculate the mixed loss.\n",
    "\n",
    "        Parameters\n",
    "        ----------\n",
    "        triplet_loss: torch.Tensor\n",
    "            Triplet loss.\n",
    "        mse: torch.Tensor\n",
    "            MSE reconstruction loss\n",
    "\n",
    "        Returns\n",
    "        -------\n",
    "        torch.Tensor\n",
    "            Mixed loss.\n",
    "        \"\"\"\n",
    "\n",
    "        if self.triplet_loss_weight == 0:\n",
    "            return mse\n",
    "        if self.triplet_loss_weight == 1:\n",
    "            return triplet_loss\n",
    "        return (self.triplet_loss_weight * triplet_loss) + (\n",
    "            (1.0 - self.triplet_loss_weight) * mse\n",
    "        )\n",
    "\n",
    "\n",
    "    def training_step(self, batch, batch_idx):\n",
    "        \"\"\"Pytorch-lightning training step.\n",
    "\n",
    "        Parameters\n",
    "        ----------\n",
    "        batch:\n",
    "            A batch as defined by a pytorch DataLoader.\n",
    "        batch_idx:\n",
    "            A batch index as defined by a pytorch-lightning.\n",
    "        \"\"\"\n",
    "\n",
    "        (\n",
    "            triplet_losses,\n",
    "            mse,\n",
    "            num_hard_triplets,\n",
    "            num_viable_triplets,\n",
    "            asw,\n",
    "            nmse,\n",
    "        ) = self.get_losses(batch)\n",
    "\n",
    "        triplet_loss = triplet_losses.mean()\n",
    "        num_nonzero_loss = (triplet_losses > 0).sum(dtype=torch.float).detach()\n",
    "        hard_triplets = num_hard_triplets / num_viable_triplets\n",
    "\n",
    "        loss = self.get_mixed_loss(triplet_loss, mse)\n",
    "\n",
    "        current_lr = self.scheduler[\"scheduler\"].get_last_lr()[0]\n",
    "\n",
    "        if self.l1 > 0:  # use l1 penalty for first layer\n",
    "            for layer in self.encoder.network:\n",
    "                if isinstance(layer, nn.Linear):\n",
    "                    l1_norm = sum(p.abs().sum() for p in layer.parameters())\n",
    "                    l1_penalty = self.l1 * l1_norm * current_lr\n",
    "                    loss += l1_penalty\n",
    "                    self.log(\n",
    "                        \"train l1 penalty\", l1_penalty, prog_bar=False, logger=True\n",
    "                    )\n",
    "                    break\n",
    "\n",
    "        # if self.l2 > 0:  # use l2 penalty\n",
    "        #    l2_regularization = []\n",
    "        #    for layer in self.encoder.network:\n",
    "        #        if isinstance(layer, nn.Linear):\n",
    "        #            l2_norm = sum(p.pow(2).sum() for p in layer.parameters())\n",
    "        #            l2_regularization.append(l2_norm)\n",
    "        #    for layer in self.decoder.network:\n",
    "        #        if isinstance(layer, nn.Linear):\n",
    "        #            l2_norm = sum(p.pow(2).sum() for p in layer.parameters())\n",
    "        #            l2_regularization.append(l2_norm)\n",
    "        #    l2_penalty = (\n",
    "        #        self.l2 * sum(l2_regularization[0:-1]) * current_lr\n",
    "        #    )  # all but reconstruction layer\n",
    "        #    loss += l2_penalty\n",
    "        #    self.log(\"train l2 penalty\", l2_penalty, prog_bar=False, logger=True)\n",
    "\n",
    "        self.log(\"train loss\", loss, prog_bar=False, logger=True)\n",
    "        self.log(\"train triplet loss\", triplet_loss, prog_bar=True, logger=True)\n",
    "        self.log(\"train mse\", mse, prog_bar=True, logger=True)\n",
    "        self.log(\"train hard triplets\", hard_triplets, prog_bar=True, logger=True)\n",
    "        self.log(\n",
    "            \"train num nonzero loss\", num_nonzero_loss, prog_bar=False, logger=True\n",
    "        )\n",
    "        self.log(\n",
    "            \"train num hard triplets\", num_hard_triplets, prog_bar=False, logger=True\n",
    "        )\n",
    "        self.log(\n",
    "            \"train num viable triplets\",\n",
    "            num_viable_triplets,\n",
    "            prog_bar=False,\n",
    "            logger=True,\n",
    "        )\n",
    "        self.log(\"train asw\", asw, prog_bar=True, logger=True)\n",
    "        self.log(\"train nmse\", nmse, prog_bar=True, logger=True)\n",
    "        return {\n",
    "            \"loss\": loss,\n",
    "            \"triplet_loss\": triplet_loss.detach(),\n",
    "            \"mse\": mse.detach(),\n",
    "            \"hard_triplets\": hard_triplets,\n",
    "            \"num_nonzeros_loss\": num_nonzero_loss,\n",
    "            \"num_hard_triplets\": num_hard_triplets,\n",
    "            \"num_viable_triplets\": num_viable_triplets,\n",
    "            \"asw\": asw,\n",
    "            \"nmse\": nmse,\n",
    "        }\n",
    "\n",
    "\n",
    "    def on_validation_epoch_start(self):\n",
    "        \"\"\"Pytorch-lightning validation epoch start.\"\"\"\n",
    "        super().on_validation_epoch_start()\n",
    "        self.val_step_outputs = []\n",
    "\n",
    "\n",
    "    def validation_step(self, batch, batch_idx):\n",
    "        \"\"\"Pytorch-lightning validation step.\n",
    "\n",
    "        Parameters\n",
    "        ----------\n",
    "        batch:\n",
    "            A batch as defined by a pytorch DataLoader.\n",
    "        batch_idx:\n",
    "            A batch index as defined by a pytorch-lightning.\n",
    "        \"\"\"\n",
    "\n",
    "        if self.trainer.datamodule.val_dataset is None:\n",
    "            return {}\n",
    "        return self._eval_step(batch, prefix=\"val\")\n",
    "\n",
    "\n",
    "    def on_validation_epoch_end(self):\n",
    "        \"\"\"Pytorch-lightning validation epoch end evaluation.\"\"\"\n",
    "\n",
    "        if self.trainer.datamodule.val_dataset is None:\n",
    "            return {}\n",
    "        return self._eval_epoch(prefix=\"val\")\n",
    "\n",
    "\n",
    "    def on_test_epoch_start(self):\n",
    "        \"\"\"Pytorch-lightning test epoch start.\"\"\"\n",
    "        super().on_test_epoch_start()\n",
    "        self.test_step_outputs = []\n",
    "\n",
    "\n",
    "    def test_step(self, batch, batch_idx):\n",
    "        \"\"\"Pytorch-lightning test step.\n",
    "\n",
    "        Parameters\n",
    "        ----------\n",
    "        batch:\n",
    "            A batch as defined by a pytorch DataLoader.\n",
    "        batch_idx:\n",
    "            A batch index as defined by a pytorch-lightning.\n",
    "        \"\"\"\n",
    "\n",
    "        if self.trainer.datamodule.val_dataset is None:\n",
    "            return {}\n",
    "        return self._eval_step(batch, prefix=\"test\")\n",
    "\n",
    "\n",
    "    def on_test_epoch_end(self):\n",
    "        \"\"\"Pytorch-lightning test epoch end evaluation.\"\"\"\n",
    "\n",
    "        if self.trainer.datamodule.val_dataset is None:\n",
    "            return {}\n",
    "        return self._eval_epoch(prefix=\"test\")\n",
    "\n",
    "\n",
    "    def _eval_step(self, batch, prefix: str):\n",
    "        \"\"\"Evaluation of validation or test step.\n",
    "\n",
    "        Parameters\n",
    "        ----------\n",
    "        batch:\n",
    "            A batch as defined by a pytorch DataLoader.\n",
    "        prefix: str\n",
    "            A string prefix to label validation versus test evaluation.\n",
    "\n",
    "        Returns\n",
    "        -------\n",
    "        dict\n",
    "            A dictionary containing step evaluation metrics.\n",
    "        \"\"\"\n",
    "\n",
    "        (\n",
    "            triplet_losses,\n",
    "            mse,\n",
    "            num_hard_triplets,\n",
    "            num_viable_triplets,\n",
    "            asw,\n",
    "            nmse,\n",
    "        ) = self.get_losses(batch, use_studies=False)\n",
    "\n",
    "        triplet_loss = triplet_losses.mean()\n",
    "        num_nonzero_loss = (triplet_losses > 0).sum()\n",
    "        hard_triplets = num_hard_triplets / num_viable_triplets\n",
    "\n",
    "        loss = self.get_mixed_loss(triplet_loss, mse)\n",
    "\n",
    "        evaluation_metric = (1 - asw) / 2 + nmse\n",
    "\n",
    "        losses = {\n",
    "            f\"{prefix}_loss\": loss,\n",
    "            f\"{prefix}_triplet_loss\": triplet_loss,\n",
    "            f\"{prefix}_mse\": mse,\n",
    "            f\"{prefix}_hard_triplets\": hard_triplets,\n",
    "            f\"{prefix}_num_nonzero_loss\": num_nonzero_loss,\n",
    "            f\"{prefix}_num_hard_triplets\": num_hard_triplets,\n",
    "            f\"{prefix}_num_viable_triplets\": num_viable_triplets,\n",
    "            f\"{prefix}_num_viable_triplets\": num_viable_triplets,\n",
    "            f\"{prefix}_asw\": asw,\n",
    "            f\"{prefix}_nmse\": nmse,\n",
    "            f\"{prefix}_evaluation_metric\": evaluation_metric,\n",
    "        }\n",
    "\n",
    "        if prefix == \"val\":\n",
    "            self.val_step_outputs.append(losses)\n",
    "        elif prefix == \"test\":\n",
    "            self.test_step_outputs.append(losses)\n",
    "        return losses\n",
    "\n",
    "    def _eval_epoch(self, prefix: str):\n",
    "        \"\"\"Evaluation of validation or test epoch.\n",
    "\n",
    "        Parameters\n",
    "        ----------\n",
    "        prefix: str\n",
    "            A string prefix to label validation versus test evaluation.\n",
    "\n",
    "        Returns\n",
    "        -------\n",
    "        dict\n",
    "            A dictionary containing epoch evaluation metrics.\n",
    "        \"\"\"\n",
    "\n",
    "        if prefix == \"val\":\n",
    "            step_outputs = self.val_step_outputs\n",
    "        elif prefix == \"test\":\n",
    "            step_outputs = self.test_step_outputs\n",
    "\n",
    "        loss = torch.Tensor([step[f\"{prefix}_loss\"] for step in step_outputs]).mean()\n",
    "        triplet_loss = torch.Tensor(\n",
    "            [step[f\"{prefix}_triplet_loss\"] for step in step_outputs]\n",
    "        ).mean()\n",
    "        mse = torch.Tensor([step[f\"{prefix}_mse\"] for step in step_outputs]).mean()\n",
    "        hard_triplets = torch.Tensor(\n",
    "            [step[f\"{prefix}_hard_triplets\"] for step in step_outputs]\n",
    "        ).mean()\n",
    "        num_nonzero_loss = torch.Tensor(\n",
    "            [step[f\"{prefix}_num_nonzero_loss\"] for step in step_outputs]\n",
    "        ).mean()\n",
    "        num_hard_triplets = torch.Tensor(\n",
    "            [step[f\"{prefix}_num_hard_triplets\"] for step in step_outputs]\n",
    "        ).mean()\n",
    "        num_viable_triplets = torch.Tensor(\n",
    "            [step[f\"{prefix}_num_viable_triplets\"] for step in step_outputs]\n",
    "        ).mean()\n",
    "        asw = torch.Tensor([step[f\"{prefix}_asw\"] for step in step_outputs]).mean()\n",
    "        nmse = torch.Tensor([step[f\"{prefix}_nmse\"] for step in step_outputs]).mean()\n",
    "        evaluation_metric = torch.Tensor(\n",
    "            [step[f\"{prefix}_evaluation_metric\"] for step in step_outputs]\n",
    "        ).mean()\n",
    "\n",
    "        self.log(f\"{prefix} loss\", loss, logger=True)\n",
    "        self.log(f\"{prefix} triplet loss\", triplet_loss, logger=True)\n",
    "        self.log(f\"{prefix} mse\", mse, logger=True)\n",
    "        self.log(f\"{prefix} hard triplets\", hard_triplets, logger=True)\n",
    "        self.log(f\"{prefix} num nonzero loss\", num_nonzero_loss, logger=True)\n",
    "        self.log(f\"{prefix} num hard triplets\", num_hard_triplets, logger=True)\n",
    "        self.log(f\"{prefix} num viable triplets\", num_viable_triplets, logger=True)\n",
    "        self.log(f\"{prefix} asw\", asw, logger=True)\n",
    "        self.log(f\"{prefix} nmse\", nmse, logger=True)\n",
    "        self.log(f\"{prefix} evaluation_metric\", evaluation_metric, logger=True)\n",
    "\n",
    "        losses = {\n",
    "            f\"{prefix}_loss\": loss,\n",
    "            f\"{prefix}_triplet_loss\": triplet_loss,\n",
    "            f\"{prefix}_mse\": mse,\n",
    "            f\"{prefix}_hard_triplets\": hard_triplets,\n",
    "            f\"{prefix}_num_nonzero_loss\": num_nonzero_loss,\n",
    "            f\"{prefix}_num_hard_triplets\": num_hard_triplets,\n",
    "            f\"{prefix}_num_viable_triplets\": num_viable_triplets,\n",
    "            f\"{prefix}_asw\": asw,\n",
    "            f\"{prefix}_nmse\": nmse,\n",
    "            f\"{prefix}_evaluation_metric\": evaluation_metric,\n",
    "        }\n",
    "        return losses\n",
    "\n",
    "    def save_all(\n",
    "        self,\n",
    "        model_path: str,\n",
    "    ):\n",
    "        if not os.path.isdir(model_path):\n",
    "            os.makedirs(model_path)\n",
    "\n",
    "        # save NN model\n",
    "        self.encoder.save_state(os.path.join(model_path, \"encoder.ckpt\"))\n",
    "        self.decoder.save_state(os.path.join(model_path, \"decoder.ckpt\"))\n",
    "\n",
    "        # save layer sizes as json, useful to infer model architecture\n",
    "        with open(os.path.join(model_path, \"layer_sizes.json\"), \"w\") as f:\n",
    "            f.write(json.dumps(self.layer_sizes))\n",
    "\n",
    "        # save hyperparameters as json\n",
    "        hyperparameters = {\n",
    "            \"latent_dim\": self.latent_dim,\n",
    "            \"hidden_dim\": self.hidden_dim,\n",
    "            \"dropout\": self.dropout,\n",
    "            \"input_dropout\": self.input_dropout,\n",
    "            \"margin\": self.margin,\n",
    "            \"triplet_loss_weight\": self.triplet_loss_weight,\n",
    "            \"negative_selection\": self.negative_selection,\n",
    "            \"sample_across_studies\": self.sample_across_studies,\n",
    "            \"perturb_labels\": self.perturb_labels,\n",
    "            \"perturb_labels_fraction\": self.perturb_labels_fraction,\n",
    "            \"lr\": self.lr,\n",
    "            \"l1_lambda\": self.l1,\n",
    "            \"l2_lambda\": self.l2,\n",
    "            \"batch_size\": self.trainer.datamodule.batch_size,\n",
    "            \"max_epochs\": self.max_epochs,\n",
    "            \"residual\": self.residual,\n",
    "        }\n",
    "        with open(os.path.join(model_path, \"hyperparameters.json\"), \"w\") as f:\n",
    "            f.write(json.dumps(hyperparameters))\n",
    "\n",
    "        # write gene order\n",
    "        with open(os.path.join(model_path, \"gene_order.tsv\"), \"w\") as f:\n",
    "            f.write(\"\\n\".join(self.trainer.datamodule.gene_order))\n",
    "\n",
    "        # write dictionary to map label_ints to labels\n",
    "        pd.Series(self.trainer.datamodule.int2label).to_csv(\n",
    "            os.path.join(model_path, \"label_ints.csv\")\n",
    "        )\n",
    "\n",
    "        # write metadata: data paths, timestamp\n",
    "        meta_data = {\n",
    "            \"date\": datetime.now().strftime(\"%Y-%m-%d %H:%M:%S\"),\n",
    "        }\n",
    "        if \"train_path\" in dir(self.trainer.datamodule):\n",
    "            meta_data[\"train_path\"] = self.trainer.datamodule.train_path\n",
    "            meta_data[\"val_path\"] = self.trainer.datamodule.val_path\n",
    "            # write reference labels\n",
    "            with open(os.path.join(model_path, \"reference_labels.tsv\"), \"w\") as f:\n",
    "                f.write(\"\\n\".join(self.trainer.datamodule.train_Y))\n",
    "        elif \"cell_tdb_uri\" in dir(self.trainer.datamodule):\n",
    "            meta_data[\"cell_tdb_uri\"] = self.trainer.datamodule.cell_tdb_uri\n",
    "            meta_data[\"gene_tdb_uri\"] = self.trainer.datamodule.gene_tdb_uri\n",
    "            meta_data[\"counts_tdb_uri\"] = self.trainer.datamodule.counts_tdb_uri\n",
    "            self.trainer.datamodule.train_df.to_csv(\n",
    "                os.path.join(model_path, \"train_cells.csv\")\n",
    "            )\n",
    "            if self.trainer.datamodule.val_df is not None:\n",
    "                self.trainer.datamodule.val_df.to_csv(\n",
    "                    os.path.join(model_path, \"val_cells.csv\")\n",
    "                )\n",
    "            # write reference labels\n",
    "            with open(os.path.join(model_path, \"reference_labels.tsv\"), \"w\") as f:\n",
    "                f.write(\n",
    "                    \"\\n\".join(\n",
    "                        self.trainer.datamodule.train_df[\"cellTypeName\"].values.tolist()\n",
    "                    )\n",
    "                )\n",
    "        with open(os.path.join(model_path, \"metadata.json\"), \"w\") as f:\n",
    "            f.write(json.dumps(meta_data))\n",
    "\n",
    "    def load_state(\n",
    "        self,\n",
    "        encoder_filename: str,\n",
    "        decoder_filename: str,\n",
    "        use_gpu: bool = False,\n",
    "        freeze: bool = False,\n",
    "    ):\n",
    "        \"\"\"Load model state.\n",
    "\n",
    "        Parameters\n",
    "        ----------\n",
    "        encoder_filename: str\n",
    "            Filename containing the encoder model state.\n",
    "        decoder_filename: str\n",
    "            Filename containing the decoder model state.\n",
    "        use_gpu: bool, default: False\n",
    "            Boolean indicating whether or not to use GPUs.\n",
    "        freeze: bool, default: False\n",
    "            Freeze all but bottleneck layer, used if pretraining the encoder.\n",
    "        \"\"\"\n",
    "\n",
    "        self.encoder.load_state(encoder_filename, use_gpu)\n",
    "        self.decoder.load_state(decoder_filename, use_gpu)\n",
    "\n",
    "        if freeze:\n",
    "            # encoder batchnorm freeze\n",
    "            for i in range(len(self.encoder.network)):\n",
    "                if isinstance(self.encoder.network[i], nn.BatchNorm1d):\n",
    "                    for param in self.encoder.network[i].parameters():\n",
    "                        param.requires_grad = False  # freeze\n",
    "\n",
    "            # encoder linear freeze\n",
    "            encoder_linear_idx = []\n",
    "            for i in range(len(self.encoder.network)):\n",
    "                if isinstance(self.encoder.network[i], nn.Linear):\n",
    "                    encoder_linear_idx.append(i)\n",
    "            for i in range(len(encoder_linear_idx)):\n",
    "                if i < len(encoder_linear_idx) - 1:  # freeze all but bottleneck\n",
    "                    for param in self.encoder.network[\n",
    "                        encoder_linear_idx[i]\n",
    "                    ].parameters():\n",
    "                        param.requires_grad = False  # freeze\n",
    "                else:\n",
    "                    for param in self.encoder.network[\n",
    "                        encoder_linear_idx[i]\n",
    "                    ].parameters():\n",
    "                        param.requires_grad = True  # unfreeze"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "04707b05-0118-45a1-8b30-1fba06ad85a7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "34667"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "datamodule.n_genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "7b8d087d-202c-428c-9976-4dc218ae611a",
   "metadata": {},
   "outputs": [],
   "source": [
    "from pytorch_lightning import Trainer\n",
    "\n",
    "# 初始化你的模型\n",
    "model = MetricLearning(datamodule.n_genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "d57f522d-08ba-40bd-bbda-d5b6328ebc1b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "MetricLearning(\n",
       "  (encoder): Encoder(\n",
       "    (network): ModuleList(\n",
       "      (0): Sequential(\n",
       "        (0): Dropout(p=0.4, inplace=False)\n",
       "        (1): Linear(in_features=34667, out_features=1024, bias=True)\n",
       "        (2): BatchNorm1d(1024, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)\n",
       "        (3): PReLU(num_parameters=1)\n",
       "      )\n",
       "      (1): Sequential(\n",
       "        (0): Dropout(p=0.5, inplace=False)\n",
       "        (1): Linear(in_features=1024, out_features=1024, bias=True)\n",
       "        (2): BatchNorm1d(1024, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)\n",
       "        (3): PReLU(num_parameters=1)\n",
       "      )\n",
       "      (2): Linear(in_features=1024, out_features=128, bias=True)\n",
       "    )\n",
       "  )\n",
       "  (decoder): Decoder(\n",
       "    (network): ModuleList(\n",
       "      (0): Sequential(\n",
       "        (0): Linear(in_features=128, out_features=1024, bias=True)\n",
       "        (1): BatchNorm1d(1024, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)\n",
       "        (2): PReLU(num_parameters=1)\n",
       "      )\n",
       "      (1): Sequential(\n",
       "        (0): Dropout(p=0.5, inplace=False)\n",
       "        (1): Linear(in_features=1024, out_features=1024, bias=True)\n",
       "        (2): BatchNorm1d(1024, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)\n",
       "        (3): PReLU(num_parameters=1)\n",
       "      )\n",
       "      (2): Linear(in_features=1024, out_features=34667, bias=True)\n",
       "    )\n",
       "  )\n",
       "  (triplet_loss_fn): TripletLoss()\n",
       "  (mse_loss_fn): MSELoss()\n",
       ")"
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "6e42f668-db05-40ca-ada7-fc0f03d4063b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(tensor([[ 0.0706,  0.2073,  0.0109,  ..., -0.0017, -0.0100, -0.0760],\n",
       "         [-0.0128, -0.0461, -0.1440,  ...,  0.0762,  0.0803, -0.1886],\n",
       "         [ 0.1080,  0.0219, -0.0311,  ...,  0.1358, -0.0655,  0.0179],\n",
       "         ...,\n",
       "         [ 0.0171,  0.0853, -0.0644,  ...,  0.0506,  0.0892, -0.0411],\n",
       "         [ 0.1707,  0.1254, -0.0641,  ...,  0.1234, -0.0903, -0.1283],\n",
       "         [-0.0066, -0.0200,  0.0060,  ...,  0.0780, -0.0463, -0.0759]],\n",
       "        grad_fn=<DivBackward0>),\n",
       " tensor([[-0.2675,  0.4704,  0.2071,  ...,  0.2816, -0.1630,  0.3185],\n",
       "         [-0.4967,  0.4436,  0.2416,  ...,  0.1844,  0.3107, -0.0721],\n",
       "         [-0.6210, -0.3960,  0.1267,  ...,  0.0835,  0.2555,  0.1837],\n",
       "         ...,\n",
       "         [-0.1788,  0.6497,  0.0054,  ..., -0.0246,  0.8695, -0.1197],\n",
       "         [-0.0763, -0.6518, -0.6226,  ..., -0.2271,  0.1449, -0.1271],\n",
       "         [ 0.3007,  0.3907, -0.1019,  ..., -0.1534,  0.0281, -0.2759]],\n",
       "        grad_fn=<AddmmBackward0>))"
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.forward(batch[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "e2d2352e-7e13-466a-8d68-5342313482e3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "torch.Size([1000, 128])"
      ]
     },
     "execution_count": 64,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.forward(batch[0])[0].shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "902eea16-3dd7-452a-a77b-a48fb07342fb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "torch.Size([1000, 34667])"
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.forward(batch[0])[1].shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "85d07717-6204-40d0-98df-ab785e0b1e87",
   "metadata": {},
   "outputs": [],
   "source": [
    "embeddings, _ = model.forward(batch[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "e754ac3f-977b-4466-89eb-998bcc7fc22c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "tensor([[ 0.0106,  0.1331,  0.0747,  ...,  0.2205,  0.0040, -0.1336],\n",
       "        [ 0.0485,  0.0374, -0.0858,  ...,  0.1137,  0.0259, -0.1913],\n",
       "        [ 0.0031,  0.0510,  0.0277,  ..., -0.0363, -0.0523, -0.2076],\n",
       "        ...,\n",
       "        [ 0.0592, -0.0116,  0.1152,  ..., -0.0365,  0.0745, -0.0992],\n",
       "        [ 0.0536, -0.1053,  0.0034,  ...,  0.0659, -0.0403, -0.4034],\n",
       "        [-0.1027, -0.0217,  0.0662,  ..., -0.0620, -0.1144, -0.1382]],\n",
       "       grad_fn=<DivBackward0>)"
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "embeddings"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "dbf9e798-3f0f-4595-b1a1-beffc1cfd2b0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "((tensor([[ 1.0635e-02,  1.3310e-01,  7.4704e-02, -3.6707e-03,  7.4565e-04,\n",
       "            1.0046e-01,  5.5280e-02, -1.7478e-01, -7.4276e-02,  9.0263e-02,\n",
       "            3.8132e-02,  5.7911e-02,  1.8540e-01, -1.8250e-01, -8.2295e-02,\n",
       "           -8.1861e-02,  2.2857e-02, -2.3162e-02,  5.9862e-02,  8.3447e-02,\n",
       "            2.6331e-02, -3.8658e-02,  4.2412e-02, -3.5144e-02,  1.9559e-02,\n",
       "           -1.3284e-01, -1.0458e-01,  1.1118e-02, -4.1271e-02,  1.4218e-01,\n",
       "            6.7288e-02,  1.6084e-02,  9.6247e-02,  5.7376e-02,  5.7281e-02,\n",
       "           -3.6550e-02,  8.1002e-02,  3.3958e-02,  3.3214e-02, -3.9106e-02,\n",
       "           -1.8417e-01,  1.3401e-01, -4.0572e-02,  6.1037e-03,  1.3350e-01,\n",
       "           -1.5692e-02, -2.4100e-01, -2.4445e-02, -5.1360e-02,  1.9058e-02,\n",
       "           -5.9606e-02, -7.3374e-03,  1.2492e-01, -8.0148e-02, -3.2777e-02,\n",
       "           -9.3750e-02,  1.1271e-01,  1.0915e-01,  3.5869e-02, -1.4058e-02,\n",
       "           -2.4243e-02, -9.5141e-02, -1.3024e-01, -1.1425e-01, -8.4133e-02,\n",
       "            5.9834e-03,  5.4315e-02,  1.1227e-02,  1.4714e-01, -8.6306e-02,\n",
       "           -5.4453e-02, -5.1231e-02,  2.9135e-02,  9.4511e-02,  2.9038e-02,\n",
       "           -4.2200e-02,  1.2358e-01,  8.3657e-02, -1.6764e-01,  1.1786e-01,\n",
       "            1.0295e-01, -5.7675e-02, -2.3164e-02,  1.6573e-01, -1.3397e-01,\n",
       "            2.1283e-02,  7.6464e-02, -1.5246e-02, -9.6587e-02, -1.3601e-02,\n",
       "            1.4358e-01, -3.2768e-02,  7.4598e-02,  5.3292e-02,  9.7911e-02,\n",
       "            6.2422e-02, -2.3049e-02, -3.2100e-02,  2.1893e-01,  1.9637e-01,\n",
       "           -3.1502e-02,  1.2310e-01,  3.9759e-05,  2.3438e-02, -4.6445e-02,\n",
       "           -3.1591e-03, -3.3060e-02, -1.5163e-01, -9.7036e-03,  1.1962e-01,\n",
       "           -2.5254e-02, -4.9936e-02,  1.0126e-01, -5.0881e-02,  4.9948e-02,\n",
       "           -7.5357e-02, -2.7763e-02, -1.1555e-01, -3.0064e-02, -3.2516e-02,\n",
       "           -1.9582e-02,  1.4210e-02, -6.8309e-02, -1.5545e-03,  1.1296e-02,\n",
       "            2.2050e-01,  3.9908e-03, -1.3361e-01]], grad_fn=<IndexBackward0>),\n",
       "  tensor([[ 1.0635e-02,  1.3310e-01,  7.4704e-02, -3.6707e-03,  7.4565e-04,\n",
       "            1.0046e-01,  5.5280e-02, -1.7478e-01, -7.4276e-02,  9.0263e-02,\n",
       "            3.8132e-02,  5.7911e-02,  1.8540e-01, -1.8250e-01, -8.2295e-02,\n",
       "           -8.1861e-02,  2.2857e-02, -2.3162e-02,  5.9862e-02,  8.3447e-02,\n",
       "            2.6331e-02, -3.8658e-02,  4.2412e-02, -3.5144e-02,  1.9559e-02,\n",
       "           -1.3284e-01, -1.0458e-01,  1.1118e-02, -4.1271e-02,  1.4218e-01,\n",
       "            6.7288e-02,  1.6084e-02,  9.6247e-02,  5.7376e-02,  5.7281e-02,\n",
       "           -3.6550e-02,  8.1002e-02,  3.3958e-02,  3.3214e-02, -3.9106e-02,\n",
       "           -1.8417e-01,  1.3401e-01, -4.0572e-02,  6.1037e-03,  1.3350e-01,\n",
       "           -1.5692e-02, -2.4100e-01, -2.4445e-02, -5.1360e-02,  1.9058e-02,\n",
       "           -5.9606e-02, -7.3374e-03,  1.2492e-01, -8.0148e-02, -3.2777e-02,\n",
       "           -9.3750e-02,  1.1271e-01,  1.0915e-01,  3.5869e-02, -1.4058e-02,\n",
       "           -2.4243e-02, -9.5141e-02, -1.3024e-01, -1.1425e-01, -8.4133e-02,\n",
       "            5.9834e-03,  5.4315e-02,  1.1227e-02,  1.4714e-01, -8.6306e-02,\n",
       "           -5.4453e-02, -5.1231e-02,  2.9135e-02,  9.4511e-02,  2.9038e-02,\n",
       "           -4.2200e-02,  1.2358e-01,  8.3657e-02, -1.6764e-01,  1.1786e-01,\n",
       "            1.0295e-01, -5.7675e-02, -2.3164e-02,  1.6573e-01, -1.3397e-01,\n",
       "            2.1283e-02,  7.6464e-02, -1.5246e-02, -9.6587e-02, -1.3601e-02,\n",
       "            1.4358e-01, -3.2768e-02,  7.4598e-02,  5.3292e-02,  9.7911e-02,\n",
       "            6.2422e-02, -2.3049e-02, -3.2100e-02,  2.1893e-01,  1.9637e-01,\n",
       "           -3.1502e-02,  1.2310e-01,  3.9759e-05,  2.3438e-02, -4.6445e-02,\n",
       "           -3.1591e-03, -3.3060e-02, -1.5163e-01, -9.7036e-03,  1.1962e-01,\n",
       "           -2.5254e-02, -4.9936e-02,  1.0126e-01, -5.0881e-02,  4.9948e-02,\n",
       "           -7.5357e-02, -2.7763e-02, -1.1555e-01, -3.0064e-02, -3.2516e-02,\n",
       "           -1.9582e-02,  1.4210e-02, -6.8309e-02, -1.5545e-03,  1.1296e-02,\n",
       "            2.2050e-01,  3.9908e-03, -1.3361e-01]], grad_fn=<IndexBackward0>),\n",
       "  tensor([[ 1.0635e-02,  1.3310e-01,  7.4704e-02, -3.6707e-03,  7.4565e-04,\n",
       "            1.0046e-01,  5.5280e-02, -1.7478e-01, -7.4276e-02,  9.0263e-02,\n",
       "            3.8132e-02,  5.7911e-02,  1.8540e-01, -1.8250e-01, -8.2295e-02,\n",
       "           -8.1861e-02,  2.2857e-02, -2.3162e-02,  5.9862e-02,  8.3447e-02,\n",
       "            2.6331e-02, -3.8658e-02,  4.2412e-02, -3.5144e-02,  1.9559e-02,\n",
       "           -1.3284e-01, -1.0458e-01,  1.1118e-02, -4.1271e-02,  1.4218e-01,\n",
       "            6.7288e-02,  1.6084e-02,  9.6247e-02,  5.7376e-02,  5.7281e-02,\n",
       "           -3.6550e-02,  8.1002e-02,  3.3958e-02,  3.3214e-02, -3.9106e-02,\n",
       "           -1.8417e-01,  1.3401e-01, -4.0572e-02,  6.1037e-03,  1.3350e-01,\n",
       "           -1.5692e-02, -2.4100e-01, -2.4445e-02, -5.1360e-02,  1.9058e-02,\n",
       "           -5.9606e-02, -7.3374e-03,  1.2492e-01, -8.0148e-02, -3.2777e-02,\n",
       "           -9.3750e-02,  1.1271e-01,  1.0915e-01,  3.5869e-02, -1.4058e-02,\n",
       "           -2.4243e-02, -9.5141e-02, -1.3024e-01, -1.1425e-01, -8.4133e-02,\n",
       "            5.9834e-03,  5.4315e-02,  1.1227e-02,  1.4714e-01, -8.6306e-02,\n",
       "           -5.4453e-02, -5.1231e-02,  2.9135e-02,  9.4511e-02,  2.9038e-02,\n",
       "           -4.2200e-02,  1.2358e-01,  8.3657e-02, -1.6764e-01,  1.1786e-01,\n",
       "            1.0295e-01, -5.7675e-02, -2.3164e-02,  1.6573e-01, -1.3397e-01,\n",
       "            2.1283e-02,  7.6464e-02, -1.5246e-02, -9.6587e-02, -1.3601e-02,\n",
       "            1.4358e-01, -3.2768e-02,  7.4598e-02,  5.3292e-02,  9.7911e-02,\n",
       "            6.2422e-02, -2.3049e-02, -3.2100e-02,  2.1893e-01,  1.9637e-01,\n",
       "           -3.1502e-02,  1.2310e-01,  3.9759e-05,  2.3438e-02, -4.6445e-02,\n",
       "           -3.1591e-03, -3.3060e-02, -1.5163e-01, -9.7036e-03,  1.1962e-01,\n",
       "           -2.5254e-02, -4.9936e-02,  1.0126e-01, -5.0881e-02,  4.9948e-02,\n",
       "           -7.5357e-02, -2.7763e-02, -1.1555e-01, -3.0064e-02, -3.2516e-02,\n",
       "           -1.9582e-02,  1.4210e-02, -6.8309e-02, -1.5545e-03,  1.1296e-02,\n",
       "            2.2050e-01,  3.9908e-03, -1.3361e-01]], grad_fn=<IndexBackward0>)),\n",
       " 0,\n",
       " 6)"
      ]
     },
     "execution_count": 68,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "triplet_selector.get_triplets(embeddings,labels,int2label)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "80d6d278-4b56-4967-bf1f-5fc8efb07c7f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(tensor([[ 1.0635e-02,  1.3310e-01,  7.4704e-02, -3.6707e-03,  7.4565e-04,\n",
       "           1.0046e-01,  5.5280e-02, -1.7478e-01, -7.4276e-02,  9.0263e-02,\n",
       "           3.8132e-02,  5.7911e-02,  1.8540e-01, -1.8250e-01, -8.2295e-02,\n",
       "          -8.1861e-02,  2.2857e-02, -2.3162e-02,  5.9862e-02,  8.3447e-02,\n",
       "           2.6331e-02, -3.8658e-02,  4.2412e-02, -3.5144e-02,  1.9559e-02,\n",
       "          -1.3284e-01, -1.0458e-01,  1.1118e-02, -4.1271e-02,  1.4218e-01,\n",
       "           6.7288e-02,  1.6084e-02,  9.6247e-02,  5.7376e-02,  5.7281e-02,\n",
       "          -3.6550e-02,  8.1002e-02,  3.3958e-02,  3.3214e-02, -3.9106e-02,\n",
       "          -1.8417e-01,  1.3401e-01, -4.0572e-02,  6.1037e-03,  1.3350e-01,\n",
       "          -1.5692e-02, -2.4100e-01, -2.4445e-02, -5.1360e-02,  1.9058e-02,\n",
       "          -5.9606e-02, -7.3374e-03,  1.2492e-01, -8.0148e-02, -3.2777e-02,\n",
       "          -9.3750e-02,  1.1271e-01,  1.0915e-01,  3.5869e-02, -1.4058e-02,\n",
       "          -2.4243e-02, -9.5141e-02, -1.3024e-01, -1.1425e-01, -8.4133e-02,\n",
       "           5.9834e-03,  5.4315e-02,  1.1227e-02,  1.4714e-01, -8.6306e-02,\n",
       "          -5.4453e-02, -5.1231e-02,  2.9135e-02,  9.4511e-02,  2.9038e-02,\n",
       "          -4.2200e-02,  1.2358e-01,  8.3657e-02, -1.6764e-01,  1.1786e-01,\n",
       "           1.0295e-01, -5.7675e-02, -2.3164e-02,  1.6573e-01, -1.3397e-01,\n",
       "           2.1283e-02,  7.6464e-02, -1.5246e-02, -9.6587e-02, -1.3601e-02,\n",
       "           1.4358e-01, -3.2768e-02,  7.4598e-02,  5.3292e-02,  9.7911e-02,\n",
       "           6.2422e-02, -2.3049e-02, -3.2100e-02,  2.1893e-01,  1.9637e-01,\n",
       "          -3.1502e-02,  1.2310e-01,  3.9759e-05,  2.3438e-02, -4.6445e-02,\n",
       "          -3.1591e-03, -3.3060e-02, -1.5163e-01, -9.7036e-03,  1.1962e-01,\n",
       "          -2.5254e-02, -4.9936e-02,  1.0126e-01, -5.0881e-02,  4.9948e-02,\n",
       "          -7.5357e-02, -2.7763e-02, -1.1555e-01, -3.0064e-02, -3.2516e-02,\n",
       "          -1.9582e-02,  1.4210e-02, -6.8309e-02, -1.5545e-03,  1.1296e-02,\n",
       "           2.2050e-01,  3.9908e-03, -1.3361e-01]], grad_fn=<IndexBackward0>),\n",
       " tensor([[ 1.0635e-02,  1.3310e-01,  7.4704e-02, -3.6707e-03,  7.4565e-04,\n",
       "           1.0046e-01,  5.5280e-02, -1.7478e-01, -7.4276e-02,  9.0263e-02,\n",
       "           3.8132e-02,  5.7911e-02,  1.8540e-01, -1.8250e-01, -8.2295e-02,\n",
       "          -8.1861e-02,  2.2857e-02, -2.3162e-02,  5.9862e-02,  8.3447e-02,\n",
       "           2.6331e-02, -3.8658e-02,  4.2412e-02, -3.5144e-02,  1.9559e-02,\n",
       "          -1.3284e-01, -1.0458e-01,  1.1118e-02, -4.1271e-02,  1.4218e-01,\n",
       "           6.7288e-02,  1.6084e-02,  9.6247e-02,  5.7376e-02,  5.7281e-02,\n",
       "          -3.6550e-02,  8.1002e-02,  3.3958e-02,  3.3214e-02, -3.9106e-02,\n",
       "          -1.8417e-01,  1.3401e-01, -4.0572e-02,  6.1037e-03,  1.3350e-01,\n",
       "          -1.5692e-02, -2.4100e-01, -2.4445e-02, -5.1360e-02,  1.9058e-02,\n",
       "          -5.9606e-02, -7.3374e-03,  1.2492e-01, -8.0148e-02, -3.2777e-02,\n",
       "          -9.3750e-02,  1.1271e-01,  1.0915e-01,  3.5869e-02, -1.4058e-02,\n",
       "          -2.4243e-02, -9.5141e-02, -1.3024e-01, -1.1425e-01, -8.4133e-02,\n",
       "           5.9834e-03,  5.4315e-02,  1.1227e-02,  1.4714e-01, -8.6306e-02,\n",
       "          -5.4453e-02, -5.1231e-02,  2.9135e-02,  9.4511e-02,  2.9038e-02,\n",
       "          -4.2200e-02,  1.2358e-01,  8.3657e-02, -1.6764e-01,  1.1786e-01,\n",
       "           1.0295e-01, -5.7675e-02, -2.3164e-02,  1.6573e-01, -1.3397e-01,\n",
       "           2.1283e-02,  7.6464e-02, -1.5246e-02, -9.6587e-02, -1.3601e-02,\n",
       "           1.4358e-01, -3.2768e-02,  7.4598e-02,  5.3292e-02,  9.7911e-02,\n",
       "           6.2422e-02, -2.3049e-02, -3.2100e-02,  2.1893e-01,  1.9637e-01,\n",
       "          -3.1502e-02,  1.2310e-01,  3.9759e-05,  2.3438e-02, -4.6445e-02,\n",
       "          -3.1591e-03, -3.3060e-02, -1.5163e-01, -9.7036e-03,  1.1962e-01,\n",
       "          -2.5254e-02, -4.9936e-02,  1.0126e-01, -5.0881e-02,  4.9948e-02,\n",
       "          -7.5357e-02, -2.7763e-02, -1.1555e-01, -3.0064e-02, -3.2516e-02,\n",
       "          -1.9582e-02,  1.4210e-02, -6.8309e-02, -1.5545e-03,  1.1296e-02,\n",
       "           2.2050e-01,  3.9908e-03, -1.3361e-01]], grad_fn=<IndexBackward0>),\n",
       " tensor([[ 1.0635e-02,  1.3310e-01,  7.4704e-02, -3.6707e-03,  7.4565e-04,\n",
       "           1.0046e-01,  5.5280e-02, -1.7478e-01, -7.4276e-02,  9.0263e-02,\n",
       "           3.8132e-02,  5.7911e-02,  1.8540e-01, -1.8250e-01, -8.2295e-02,\n",
       "          -8.1861e-02,  2.2857e-02, -2.3162e-02,  5.9862e-02,  8.3447e-02,\n",
       "           2.6331e-02, -3.8658e-02,  4.2412e-02, -3.5144e-02,  1.9559e-02,\n",
       "          -1.3284e-01, -1.0458e-01,  1.1118e-02, -4.1271e-02,  1.4218e-01,\n",
       "           6.7288e-02,  1.6084e-02,  9.6247e-02,  5.7376e-02,  5.7281e-02,\n",
       "          -3.6550e-02,  8.1002e-02,  3.3958e-02,  3.3214e-02, -3.9106e-02,\n",
       "          -1.8417e-01,  1.3401e-01, -4.0572e-02,  6.1037e-03,  1.3350e-01,\n",
       "          -1.5692e-02, -2.4100e-01, -2.4445e-02, -5.1360e-02,  1.9058e-02,\n",
       "          -5.9606e-02, -7.3374e-03,  1.2492e-01, -8.0148e-02, -3.2777e-02,\n",
       "          -9.3750e-02,  1.1271e-01,  1.0915e-01,  3.5869e-02, -1.4058e-02,\n",
       "          -2.4243e-02, -9.5141e-02, -1.3024e-01, -1.1425e-01, -8.4133e-02,\n",
       "           5.9834e-03,  5.4315e-02,  1.1227e-02,  1.4714e-01, -8.6306e-02,\n",
       "          -5.4453e-02, -5.1231e-02,  2.9135e-02,  9.4511e-02,  2.9038e-02,\n",
       "          -4.2200e-02,  1.2358e-01,  8.3657e-02, -1.6764e-01,  1.1786e-01,\n",
       "           1.0295e-01, -5.7675e-02, -2.3164e-02,  1.6573e-01, -1.3397e-01,\n",
       "           2.1283e-02,  7.6464e-02, -1.5246e-02, -9.6587e-02, -1.3601e-02,\n",
       "           1.4358e-01, -3.2768e-02,  7.4598e-02,  5.3292e-02,  9.7911e-02,\n",
       "           6.2422e-02, -2.3049e-02, -3.2100e-02,  2.1893e-01,  1.9637e-01,\n",
       "          -3.1502e-02,  1.2310e-01,  3.9759e-05,  2.3438e-02, -4.6445e-02,\n",
       "          -3.1591e-03, -3.3060e-02, -1.5163e-01, -9.7036e-03,  1.1962e-01,\n",
       "          -2.5254e-02, -4.9936e-02,  1.0126e-01, -5.0881e-02,  4.9948e-02,\n",
       "          -7.5357e-02, -2.7763e-02, -1.1555e-01, -3.0064e-02, -3.2516e-02,\n",
       "          -1.9582e-02,  1.4210e-02, -6.8309e-02, -1.5545e-03,  1.1296e-02,\n",
       "           2.2050e-01,  3.9908e-03, -1.3361e-01]], grad_fn=<IndexBackward0>))"
      ]
     },
     "execution_count": 69,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "triplet_selector.get_triplets(embeddings,labels,int2label)[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "b2fef0cc-9926-493d-8b48-9c760a015cf4",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "GPU available: True (cuda), used: True\n",
      "TPU available: False, using: 0 TPU cores\n",
      "HPU available: False, using: 0 HPUs\n"
     ]
    }
   ],
   "source": [
    "trainer = pl.Trainer()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "6516df7e-98ab-443f-97de-84a36aa7bd3c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<pytorch_lightning.trainer.trainer.Trainer at 0x7eff726af760>"
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "trainer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "id": "98c16e25-aa8e-413b-bb7a-a86584f907e3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['__bool__',\n",
       " '__class__',\n",
       " '__delattr__',\n",
       " '__dir__',\n",
       " '__doc__',\n",
       " '__eq__',\n",
       " '__format__',\n",
       " '__ge__',\n",
       " '__getattribute__',\n",
       " '__gt__',\n",
       " '__hash__',\n",
       " '__init__',\n",
       " '__init_subclass__',\n",
       " '__le__',\n",
       " '__lt__',\n",
       " '__ne__',\n",
       " '__new__',\n",
       " '__reduce__',\n",
       " '__reduce_ex__',\n",
       " '__repr__',\n",
       " '__setattr__',\n",
       " '__sizeof__',\n",
       " '__str__',\n",
       " '__subclasshook__']"
      ]
     },
     "execution_count": 72,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dir(trainer.train_dataloader)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "id": "c62bb67c-5fcd-4d7a-bde1-945930ba402d",
   "metadata": {},
   "outputs": [],
   "source": [
    "cells, labels, studies = batch"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "be5851fa-0f1f-464f-98cc-7dbfc56b34f0",
   "metadata": {},
   "outputs": [],
   "source": [
    "trainer.datamodule"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "ceed1130-86f8-471d-a0c8-e7e4a53faaf6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'oligodendrocyte': 0,\n",
       " 'lymphocyte': 1,\n",
       " 'oligodendrocyte precursor cell': 2,\n",
       " 'phagocyte': 3,\n",
       " 'endothelial cell': 4,\n",
       " 'microglial cell': 5,\n",
       " 'neuron': 6,\n",
       " 'astrocyte': 7}"
      ]
     },
     "execution_count": 75,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "datamodule.label2int"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "33474c6d-18e7-4cb4-813c-78d8f0a48417",
   "metadata": {},
   "outputs": [],
   "source": [
    "# !pip install numpy==1.24.0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "b3eeb299-52fb-48ae-9e93-f040f131cf8c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# import numpy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "id": "463d4eac-9126-44ce-a40d-417213f94aa2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# numpy.__version__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "id": "a44dda7a-03bf-4f94-a90e-20bbe26fb79d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 开始训练\n",
    "# trainer.fit(model, train_dataloader)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "32a3e5a3-2694-4990-849c-1b9f1b7b67f5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['__bool__',\n",
       " '__class__',\n",
       " '__delattr__',\n",
       " '__dir__',\n",
       " '__doc__',\n",
       " '__eq__',\n",
       " '__format__',\n",
       " '__ge__',\n",
       " '__getattribute__',\n",
       " '__gt__',\n",
       " '__hash__',\n",
       " '__init__',\n",
       " '__init_subclass__',\n",
       " '__le__',\n",
       " '__lt__',\n",
       " '__ne__',\n",
       " '__new__',\n",
       " '__reduce__',\n",
       " '__reduce_ex__',\n",
       " '__repr__',\n",
       " '__setattr__',\n",
       " '__sizeof__',\n",
       " '__str__',\n",
       " '__subclasshook__']"
      ]
     },
     "execution_count": 80,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dir(trainer.datamodule)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "id": "9d8d4632-db5a-426c-9292-1fd04a79c6ff",
   "metadata": {},
   "outputs": [],
   "source": [
    "torch.save(model,\"model.pth\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "id": "a7821ad3-2383-400a-8dec-48e1829c24d1",
   "metadata": {},
   "outputs": [],
   "source": [
    "#####functions######"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "id": "06629030-1e8d-4628-a57d-282b07d74469",
   "metadata": {},
   "outputs": [],
   "source": [
    "from typing import Optional, Tuple, Union"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "id": "601a5be7-9109-440f-83aa-ee65908946ec",
   "metadata": {},
   "outputs": [],
   "source": [
    "class CellEmbedding:\n",
    "    \"\"\"A class that embeds cell gene expression data using a ML model.\"\"\"\n",
    "\n",
    "    def __init__(\n",
    "        self,\n",
    "        model_path: str,\n",
    "        use_gpu: bool = False,\n",
    "    ):\n",
    "        \"\"\"Constructor.\n",
    "\n",
    "        Parameters\n",
    "        ----------\n",
    "        model_path: str\n",
    "            Path to the directory containing model files.\n",
    "        use_gpu: bool, default: False\n",
    "            Use GPU instead of CPU.\n",
    "\n",
    "        Examples\n",
    "        --------\n",
    "        >>> ce = CellEmbedding(model_path=\"/opt/data/model\")\n",
    "        \"\"\"\n",
    "\n",
    "        import json\n",
    "        import os\n",
    "        import pandas as pd\n",
    "        from .nn_models import Encoder\n",
    "\n",
    "        self.model_path = model_path\n",
    "        self.use_gpu = use_gpu\n",
    "\n",
    "        self.filenames = {\n",
    "            \"model\": os.path.join(self.model_path, \"encoder.ckpt\"),\n",
    "            \"gene_order\": os.path.join(self.model_path, \"gene_order.tsv\"),\n",
    "        }\n",
    "\n",
    "        # get gene order\n",
    "        with open(self.filenames[\"gene_order\"], \"r\") as fh:\n",
    "            self.gene_order = [line.strip() for line in fh]\n",
    "\n",
    "        # get neural network model and infer network size\n",
    "        with open(os.path.join(self.model_path, \"layer_sizes.json\"), \"r\") as fh:\n",
    "            layer_sizes = json.load(fh)\n",
    "        # keys: network.1.weight, network.2.weight, ..., network.n.weight\n",
    "        layers = [\n",
    "            (key, layer_sizes[key])\n",
    "            for key in sorted(list(layer_sizes.keys()))\n",
    "            if \"weight\" in key and len(layer_sizes[key]) > 1\n",
    "        ]\n",
    "        parameters = {\n",
    "            \"latent_dim\": layers[-1][1][0],  # last\n",
    "            \"hidden_dim\": [layer[1][0] for layer in layers][0:-1],  # all but last\n",
    "        }\n",
    "\n",
    "        self.n_genes = len(self.gene_order)\n",
    "        self.latent_dim = parameters[\"latent_dim\"]\n",
    "        self.model = Encoder(\n",
    "            n_genes=self.n_genes,\n",
    "            latent_dim=parameters[\"latent_dim\"],\n",
    "            hidden_dim=parameters[\"hidden_dim\"],\n",
    "            residual=False,\n",
    "        )\n",
    "        if self.use_gpu is True:\n",
    "            self.model.cuda()\n",
    "        self.model.load_state(self.filenames[\"model\"])\n",
    "        self.model.eval()\n",
    "\n",
    "        self.int2label = pd.read_csv(\n",
    "            os.path.join(self.model_path, \"label_ints.csv\"), index_col=0\n",
    "        )[\"0\"].to_dict()\n",
    "        self.label2int = {value: key for key, value in self.int2label.items()}\n",
    "\n",
    "    def get_embeddings(\n",
    "        self,\n",
    "        X: Union[\"scipy.sparse.csr_matrix\", \"scipy.sparse.csc_matrix\", \"numpy.ndarray\"],\n",
    "        num_cells: int = -1,\n",
    "        buffer_size: int = 10000,\n",
    "    ) -> \"numpy.ndarray\":\n",
    "        \"\"\"Calculate embeddings for lognormed gene expression matrix.\n",
    "\n",
    "        Parameters\n",
    "        ----------\n",
    "        X: scipy.sparse.csr_matrix, scipy.sparse.csc_matrix, numpy.ndarray\n",
    "            Gene space aligned and log normalized (tp10k) gene expression matrix.\n",
    "        num_cells: int, default: -1\n",
    "            The number of cells to embed, starting from index 0.\n",
    "            A value of -1 will embed all cells.\n",
    "        buffer_size: int, default: 10000\n",
    "            The number of cells to embed in one batch.\n",
    "\n",
    "        Returns\n",
    "        -------\n",
    "        numpy.ndarray\n",
    "            A 2D numpy array of embeddings [num_cells x latent_space_dimensions].\n",
    "\n",
    "        Examples\n",
    "        --------\n",
    "        >>> from scimilarity.utils import align_dataset, lognorm_counts\n",
    "        >>> ce = CellEmbedding(model_path=\"/opt/data/model\")\n",
    "        >>> data = align_dataset(data, ce.gene_order)\n",
    "        >>> data = lognorm_counts(data)\n",
    "        >>> embeddings = ce.get_embeddings(data.X)\n",
    "        \"\"\"\n",
    "\n",
    "        import numpy as np\n",
    "        from scipy.sparse import csr_matrix, csc_matrix\n",
    "        import torch\n",
    "        import zarr\n",
    "\n",
    "        if num_cells == -1:\n",
    "            num_cells = X.shape[0]\n",
    "\n",
    "        if (\n",
    "            (isinstance(X, csr_matrix) or isinstance(X, csc_matrix))\n",
    "            and (\n",
    "                isinstance(X.data, zarr.core.Array)\n",
    "                or isinstance(X.indices, zarr.core.Array)\n",
    "                or isinstance(X.indptr, zarr.core.Array)\n",
    "            )\n",
    "            and num_cells <= buffer_size\n",
    "        ):\n",
    "            X.data = X.data[...]\n",
    "            X.indices = X.indices[...]\n",
    "            X.indptr = X.indptr[...]\n",
    "\n",
    "        embedding_parts = []\n",
    "        with torch.inference_mode():  # disable gradients, not needed for inference\n",
    "            for i in range(0, num_cells, buffer_size):\n",
    "                profiles = None\n",
    "                if isinstance(X, np.ndarray):\n",
    "                    profiles = torch.Tensor(X[i : i + buffer_size])\n",
    "                elif isinstance(X, torch.Tensor):\n",
    "                    profiles = X[i : i + buffer_size]\n",
    "                elif isinstance(X, csr_matrix) or isinstance(X, csc_matrix):\n",
    "                    profiles = torch.Tensor(X[i : i + buffer_size].toarray())\n",
    "\n",
    "                if profiles is None:\n",
    "                    raise RuntimeError(f\"Unknown data type {type(X)}.\")\n",
    "\n",
    "                if self.use_gpu is True:\n",
    "                    profiles = profiles.cuda()\n",
    "                embedding_parts.append(self.model(profiles))\n",
    "\n",
    "        if not embedding_parts:\n",
    "            raise RuntimeError(f\"No valid cells detected.\")\n",
    "\n",
    "        if self.use_gpu:\n",
    "            # detach, move from gpu into cpu, return as numpy array\n",
    "            embedding = torch.vstack(embedding_parts).detach().cpu().numpy()\n",
    "        else:\n",
    "            # detach, return as numpy array\n",
    "            embedding = torch.vstack(embedding_parts).detach().numpy()\n",
    "\n",
    "        if np.isnan(embedding).any():\n",
    "            raise RuntimeError(f\"NaN detected in embeddings.\")\n",
    "\n",
    "        return embedding"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "254aa701-5663-4198-9567-92dc5bd0a6f9",
   "metadata": {},
   "outputs": [],
   "source": [
    "import torch\n",
    "\n",
    "# 加载 .pth 文件\n",
    "file_path = \"/root/autodl-tmp/model.pth\"\n",
    "model_data = torch.load(file_path, map_location=torch.device('cpu'))\n",
    "\n",
    "# 查看内容\n",
    "print(model_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b5319241-4307-4ba9-84ec-f8f6edd7fabf",
   "metadata": {},
   "outputs": [],
   "source": [
    "decoder = model.decoder  \n",
    "torch.save(decoder.state_dict(), 'decoder.ckpt')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "47f22e32-3a69-469d-b26b-93cbeab8b00b",
   "metadata": {},
   "outputs": [],
   "source": [
    "model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "036d2900-d985-429f-8e0d-695644d8137b",
   "metadata": {},
   "outputs": [],
   "source": [
    "model = torch.load('model.pth', weights_only=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fe158bad-87c2-41c3-872c-85ef61621fe9",
   "metadata": {},
   "outputs": [],
   "source": [
    "decoder = model.decoder  \n",
    "torch.save(decoder.state_dict(), 'decoder.ckpt')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "12212b84-213d-4c1f-a675-b6eccc8d5ab5",
   "metadata": {},
   "outputs": [],
   "source": [
    "encoder = model.encoder  \n",
    "torch.save(encoder.state_dict(), 'encoder.ckpt')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "591754ea-a91b-4797-bb4f-2f6fa43e4531",
   "metadata": {},
   "outputs": [],
   "source": [
    "import json\n",
    "\n",
    "# 假设你在训练模型时定义了超参数\n",
    "hyperparameters = {\n",
    "    \"latent_dim\": 128,  # 隐层维度\n",
    "    \"hidden_dim\": [1024, 1024, 1024],  # 隐藏层的大小\n",
    "    \"dropout\": 0.5,  # Dropout 概率\n",
    "    \"input_dropout\": 0.4,  # 输入层的 Dropout 概率\n",
    "    \"margin\": 0.05,  # 边际值\n",
    "    \"alpha\": 0.001,  # 学习率相关的超参数\n",
    "    \"negative_selection\": \"semihard\",  # 负选择方法\n",
    "    \"sample_across_studies\": True,  # 是否跨研究进行采样\n",
    "    \"perturb_labels\": True,  # 是否扰动标签\n",
    "    \"perturb_labels_fraction\": 0.5,  # 扰动标签的比例\n",
    "    \"lr\": 0.005,  # 学习率\n",
    "    \"l1_lambda\": 0.0001,  # L1 正则化系数\n",
    "    \"l2_lambda\": 0.01,  # L2 正则化系数\n",
    "    \"batch_size\": 1000,  # 批次大小\n",
    "    \"iterations\": 50000,  # 迭代次数\n",
    "    \"residual\": False  # 残差是否使用\n",
    "}\n",
    "\n",
    "# 保存到超参数 JSON 文件\n",
    "with open('hyperparameter.json', 'w') as f:\n",
    "    json.dump(hyperparameters, f, indent=4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a4b0fcae-cb60-4401-8f43-bbc939ade9df",
   "metadata": {},
   "outputs": [],
   "source": [
    "model.hparams"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "947ae6ae-76f5-43c7-972d-5aa61ac18f7b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 保存到超参数 JSON 文件\n",
    "with open('hyperparameter.json', 'w') as f:\n",
    "    json.dump(model.hparams, f, indent=4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "id": "42693762-819f-4c91-b129-d7434bf5381d",
   "metadata": {},
   "outputs": [],
   "source": [
    "class CellSearchKNN(CellEmbedding):\n",
    "    \"\"\"A class for searching similar cells using cell embeddings kNN.\"\"\"\n",
    "\n",
    "    def __init__(\n",
    "        self,\n",
    "        model_path: str,\n",
    "        use_gpu: bool = False,\n",
    "    ):\n",
    "        \"\"\"Constructor.\n",
    "\n",
    "        Parameters\n",
    "        ----------\n",
    "        model_path: str\n",
    "            Path to the directory containing model files.\n",
    "        use_gpu: bool, default: False\n",
    "            Use GPU instead of CPU.\n",
    "\n",
    "        Examples\n",
    "        --------\n",
    "        >>> cs = CellSearchKNN(model_path=\"/opt/data/model\")\n",
    "        \"\"\"\n",
    "\n",
    "        super().__init__(\n",
    "            model_path=model_path,\n",
    "            use_gpu=use_gpu,\n",
    "        )\n",
    "\n",
    "        self.knn = None\n",
    "        self.safelist = None\n",
    "        self.blocklist = None\n",
    "\n",
    "    def load_knn_index(self, knn_file: str):\n",
    "        \"\"\"Load the kNN index file\n",
    "\n",
    "        Parameters\n",
    "        ----------\n",
    "        knn_file: str\n",
    "            Filename of the kNN index.\n",
    "        \"\"\"\n",
    "\n",
    "        import hnswlib\n",
    "        import os\n",
    "\n",
    "        if os.path.isfile(knn_file):\n",
    "            self.knn = hnswlib.Index(space=\"cosine\", dim=self.model.latent_dim)\n",
    "            self.knn.load_index(knn_file)\n",
    "        else:\n",
    "            print(f\"Warning: No KNN index found at {knn_file}\")\n",
    "            self.knn = None\n",
    "\n",
    "\n",
    "    def get_nearest_neighbors(\n",
    "        self, embeddings: \"numpy.ndarray\", k: int = 50, ef: int = 100\n",
    "    ) -> Tuple[\"numpy.ndarray\", \"numpy.ndarray\"]:\n",
    "        \"\"\"Get nearest neighbors.\n",
    "        Used by classes that inherit from CellEmbedding and have an instantiated kNN.\n",
    "\n",
    "        Parameters\n",
    "        ----------\n",
    "        embeddings: numpy.ndarray\n",
    "            Embeddings as a numpy array.\n",
    "        k: int, default: 50\n",
    "            The number of nearest neighbors.\n",
    "        ef: int, default: 100\n",
    "            The size of the dynamic list for the nearest neighbors.\n",
    "            See https://github.com/nmslib/hnswlib/blob/master/ALGO_PARAMS.md\n",
    "\n",
    "        Returns\n",
    "        -------\n",
    "        nn_idxs: numpy.ndarray\n",
    "            A 2D numpy array of nearest neighbor indices [num_cells x k].\n",
    "        nn_dists: numpy.ndarray\n",
    "            A 2D numpy array of nearest neighbor distances [num_cells x k].\n",
    "\n",
    "        Examples\n",
    "        --------\n",
    "        >>> nn_idxs, nn_dists = get_nearest_neighbors(embeddings)\n",
    "        \"\"\"\n",
    "\n",
    "        if self.knn is None:\n",
    "            raise RuntimeError(\"kNN is not initialized.\")\n",
    "        self.knn.set_ef(ef)\n",
    "        return self.knn.knn_query(embeddings, k=k)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "71cba576-647e-45e0-9342-020057a0dc2d",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
